<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.6.40">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Jelmer Poelstra">
<meta name="dcterms.date" content="2026-01-30">

<title>Working with genome and HTS data ‚Äì ENTMLGY6703-SP26 - Jelmer‚Äôs lectures</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
  margin-bottom: 0em;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<link href="../img/favicon.png" rel="icon" type="image/png">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting-57476271c5f0467c372e20ff5d25630e.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap-315fdf77af9df3389e506d97b91a2191.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<link href="../site_libs/quarto-contrib/fontawesome6-6.7.2/all.min.css" rel="stylesheet">
<link href="../site_libs/quarto-contrib/fontawesome6-6.7.2/latex-fontsize.css" rel="stylesheet">
<script src="../site_libs/quarto-contrib/glightbox/glightbox.min.js"></script>
<link href="../site_libs/quarto-contrib/glightbox/glightbox.min.css" rel="stylesheet">
<link href="../site_libs/quarto-contrib/glightbox/lightbox.css" rel="stylesheet">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>

  <script src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

<link rel="stylesheet" href="../styles.css">
</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../index.html">
    <span class="navbar-title">ENTMLGY6703-SP26 - Jelmer‚Äôs lectures</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../index.html"> <i class="bi bi-house" role="img" aria-label="This site's homepage">
</i> 
<span class="menu-text"> </span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-week-3-hts" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Week 3: HTS</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-week-3-hts">    
        <li>
    <a class="dropdown-item" href="../pages/sequencing.html">
 <span class="dropdown-text">Lecture: High-throughput Sequencing &amp; Genomes</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../pages/infrastructure.html">
 <span class="dropdown-text">Lab part 1: Computing Environment</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../pages/data.html">
 <span class="dropdown-text">Lab part 2: Working with genome and HTS Data</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-week-4-rna-seq" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Week 4: RNA-seq</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-week-4-rna-seq">    
        <li>
    <a class="dropdown-item" href="../pages/r_homework.html">
 <span class="dropdown-text">R Homework</span></a>
  </li>  
        <li class="dropdown-header">Lecture: RNA-seq</li>
        <li class="dropdown-header">Lab: Differential expression analysis</li>
    </ul>
  </li>
</ul>
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="../about.html"> 
<span class="menu-text">About</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="https://ondemand.osc.edu"> 
<span class="menu-text"><i class="fa-solid fa-desktop" aria-label="desktop"></i></span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
    <div class="dropdown">
      <a href="" title="" id="quarto-navigation-tool-dropdown-0" class="quarto-navigation-tool dropdown-toggle px-1" data-bs-toggle="dropdown" aria-expanded="false" role="link" aria-label="The GitHub repo for this website"><i class="bi bi-github"></i></a>
      <ul class="dropdown-menu" aria-labelledby="quarto-navigation-tool-dropdown-0">
          <li>
            <a class="dropdown-item quarto-navbar-tools-item" href="https://github.com/jelmerp/ENTMLGY6703-SP26">
            Source Code
            </a>
          </li>
          <li>
            <a class="dropdown-item quarto-navbar-tools-item" href="https://github.com/jelmerp/ENTMLGY6703-SP26/issues/new">
            Report an Issue
            </a>
          </li>
      </ul>
    </div>
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#introduction" id="toc-introduction" class="nav-link active" data-scroll-target="#introduction"><span class="header-section-number">1</span> Introduction</a>
  <ul class="collapse">
  <li><a href="#goals-for-this-part-of-the-lab" id="toc-goals-for-this-part-of-the-lab" class="nav-link" data-scroll-target="#goals-for-this-part-of-the-lab"><span class="header-section-number">1.1</span> Goals for this part of the lab</a></li>
  <li><a href="#the-dataset" id="toc-the-dataset" class="nav-link" data-scroll-target="#the-dataset"><span class="header-section-number">1.2</span> The dataset</a></li>
  </ul></li>
  <li><a href="#the-files-youll-work-with" id="toc-the-files-youll-work-with" class="nav-link" data-scroll-target="#the-files-youll-work-with"><span class="header-section-number">2</span> The files you‚Äôll work with</a>
  <ul class="collapse">
  <li><a href="#the-sample-metadata" id="toc-the-sample-metadata" class="nav-link" data-scroll-target="#the-sample-metadata"><span class="header-section-number">2.1</span> The sample metadata</a></li>
  </ul></li>
  <li><a href="#reference-genome-files" id="toc-reference-genome-files" class="nav-link" data-scroll-target="#reference-genome-files"><span class="header-section-number">3</span> Reference genome files</a>
  <ul class="collapse">
  <li><a href="#finding-a-reference-genome" id="toc-finding-a-reference-genome" class="nav-link" data-scroll-target="#finding-a-reference-genome"><span class="header-section-number">3.1</span> Finding a reference genome</a></li>
  <li><a href="#the-genome-assembly-fasta-file" id="toc-the-genome-assembly-fasta-file" class="nav-link" data-scroll-target="#the-genome-assembly-fasta-file"><span class="header-section-number">3.2</span> The genome assembly FASTA file</a></li>
  <li><a href="#annotation-files" id="toc-annotation-files" class="nav-link" data-scroll-target="#annotation-files"><span class="header-section-number">3.3</span> Annotation files</a></li>
  </ul></li>
  <li><a href="#fastq-files" id="toc-fastq-files" class="nav-link" data-scroll-target="#fastq-files"><span class="header-section-number">4</span> FASTQ files</a>
  <ul class="collapse">
  <li><a href="#the-fastq-format" id="toc-the-fastq-format" class="nav-link" data-scroll-target="#the-fastq-format"><span class="header-section-number">4.1</span> The FASTQ format</a></li>
  <li><a href="#fastq-file-sizes-and-organization" id="toc-fastq-file-sizes-and-organization" class="nav-link" data-scroll-target="#fastq-file-sizes-and-organization"><span class="header-section-number">4.2</span> FASTQ file sizes and organization</a></li>
  <li><a href="#viewing-fastq-files" id="toc-viewing-fastq-files" class="nav-link" data-scroll-target="#viewing-fastq-files"><span class="header-section-number">4.3</span> Viewing FASTQ files</a></li>
  </ul></li>
  <li><a href="#fastqc-for-quality-control-of-hts-reads" id="toc-fastqc-for-quality-control-of-hts-reads" class="nav-link" data-scroll-target="#fastqc-for-quality-control-of-hts-reads"><span class="header-section-number">5</span> FastQC for quality control of HTS reads</a>
  <ul class="collapse">
  <li><a href="#running-fastqc" id="toc-running-fastqc" class="nav-link" data-scroll-target="#running-fastqc"><span class="header-section-number">5.1</span> Running FastQC</a></li>
  <li><a href="#fastqc-output-files" id="toc-fastqc-output-files" class="nav-link" data-scroll-target="#fastqc-output-files"><span class="header-section-number">5.2</span> FastQC output files</a></li>
  </ul></li>
  <li><a href="#interpreting-fastqcs-output" id="toc-interpreting-fastqcs-output" class="nav-link" data-scroll-target="#interpreting-fastqcs-output"><span class="header-section-number">6</span> Interpreting FastQC‚Äôs output</a>
  <ul class="collapse">
  <li><a href="#overview-of-module-results" id="toc-overview-of-module-results" class="nav-link" data-scroll-target="#overview-of-module-results"><span class="header-section-number">6.1</span> Overview of module results</a></li>
  <li><a href="#basic-statistics-module" id="toc-basic-statistics-module" class="nav-link" data-scroll-target="#basic-statistics-module"><span class="header-section-number">6.2</span> Basic statistics module</a></li>
  <li><a href="#per-base-quality-graph" id="toc-per-base-quality-graph" class="nav-link" data-scroll-target="#per-base-quality-graph"><span class="header-section-number">6.3</span> Per base quality graph</a></li>
  <li><a href="#adapter-content-graph" id="toc-adapter-content-graph" class="nav-link" data-scroll-target="#adapter-content-graph"><span class="header-section-number">6.4</span> Adapter content graph</a></li>
  </ul></li>
  <li><a href="#prep-for-next-weeks-lab" id="toc-prep-for-next-weeks-lab" class="nav-link" data-scroll-target="#prep-for-next-weeks-lab"><span class="header-section-number">7</span> Prep for next week‚Äôs lab</a></li>
  <li><a href="#in-closing" id="toc-in-closing" class="nav-link" data-scroll-target="#in-closing"><span class="header-section-number">8</span> In closing</a></li>
  <li><a href="#appendix-gtf-files" id="toc-appendix-gtf-files" class="nav-link" data-scroll-target="#appendix-gtf-files"><span class="header-section-number">9</span> Appendix: GTF files</a>
  <ul class="collapse">
  <li><a href="#the-gtf-format" id="toc-the-gtf-format" class="nav-link" data-scroll-target="#the-gtf-format"><span class="header-section-number">9.1</span> The GTF format</a></li>
  <li><a href="#the-culex-pipiens-gtf-file" id="toc-the-culex-pipiens-gtf-file" class="nav-link" data-scroll-target="#the-culex-pipiens-gtf-file"><span class="header-section-number">9.2</span> The <em>Culex pipiens</em> GTF file</a></li>
  </ul></li>
  </ul>
<div class="quarto-other-links"><h2>Other Links</h2><ul><li><a href="sequencing.html"><i class="bi bi-bookmarks"></i>Week 3 lecture</a></li><li><a href="infrastructure.html"><i class="bi bi-pencil"></i>Week 3 lab - part 1</a></li></ul></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Working with genome and HTS data</h1>
<p class="subtitle lead">Week 3 lab ‚Äì part 2</p>
</div>


<div class="quarto-title-meta-author">
  <div class="quarto-title-meta-heading">Author</div>
  <div class="quarto-title-meta-heading">Affiliation</div>
  
    <div class="quarto-title-meta-contents">
    <p class="author">Jelmer Poelstra <a href="mailto:poelstra.1@osu.edu" class="quarto-title-author-email"><i class="bi bi-envelope"></i></a> <a href="https://orcid.org/0000-0002-3514-7462" class="quarto-title-author-orcid"> <img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg=="></a></p>
  </div>
  <div class="quarto-title-meta-contents">
        <p class="affiliation">
            <a href="https://mcic.osu.edu/bioinformatics">
            CFAES Bioinformatics Core, Ohio State University
            </a>
          </p>
      </div>
  </div>

<div class="quarto-title-meta">

      
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">January 30, 2026</p>
    </div>
  </div>
  
    
  </div>
  


</header>


<hr>
<p><br></p>
<section id="introduction" class="level2" data-number="1">
<h2 data-number="1" class="anchored" data-anchor-id="introduction"><span class="header-section-number">1</span> Introduction</h2>
<section id="goals-for-this-part-of-the-lab" class="level3" data-number="1.1">
<h3 data-number="1.1" class="anchored" data-anchor-id="goals-for-this-part-of-the-lab"><span class="header-section-number">1.1</span> Goals for this part of the lab</h3>
<p>In this second part of the lab, you will learn how to:</p>
<ol type="1">
<li>Look at files in the Unix shell</li>
<li>Find a reference genome and the associated files</li>
<li>Explore and understand reference genome files</li>
<li>Explore and understand HTS files</li>
<li>Perform quality-control on HTS reads with a command-line tool, FastQC</li>
</ol>
</section>
<section id="the-dataset" class="level3" data-number="1.2">
<h3 data-number="1.2" class="anchored" data-anchor-id="the-dataset"><span class="header-section-number">1.2</span> The dataset</h3>
<p>As discussed in the lecture, you‚Äôll work with RNA-Seq data from <span class="citation" data-cites="garrig√≥s2025">Garrig√≥s et al. (<a href="#ref-garrig√≥s2025" role="doc-biblioref">2025</a>)</span> in today‚Äôs and next week‚Äôs labs:</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="../img/paper.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:75.0%"></p>
</figure>
</div>
<p>This paper used RNA-Seq data to study gene expression in <em>Culex pipiens</em> mosquitos infected with malaria-causing <em>Plasmodium</em> protozoans. Next week, you will learn about RNA-Seq methodology and experimental design, and we‚Äôll discuss the paper‚Äôs experimental design in more detail. Here, we‚Äôll explore the paper‚Äôs HTS data files.</p>
</section>
</section>
<section id="the-files-youll-work-with" class="level2" data-number="2">
<h2 data-number="2" class="anchored" data-anchor-id="the-files-youll-work-with"><span class="header-section-number">2</span> The files you‚Äôll work with</h2>
<p>The main data files in a reference-based HTS project<a href="#fn1" class="footnote-ref" id="fnref1" role="doc-noteref"><sup>1</sup></a> are sequence reads (usually produced for the focal project) and reference genome files (usually already available).</p>
<p>For this project ‚Äìour re-analysis of the <span class="citation" data-cites="garrig√≥s2025">Garrig√≥s et al. (<a href="#ref-garrig√≥s2025" role="doc-biblioref">2025</a>)</span> data‚Äì I have downloaded those files for you and placed them in <code>/fs/scratch/PAS2880/ENT6703/data</code> on OSC. You‚Äôve seen some of these files in the first part of this lab, but we didn‚Äôt pay much attention to what they represent.</p>
<p>First, we‚Äôll actually move one level up to the <code>ENT6703</code> dir:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> /fs/scratch/PAS2880/ENT6703</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Let‚Äôs get an overview of the full set of files using the <code>tree</code> command (with the option <code>-C</code> to show colors):</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Unlike ls, tree will show the full directory structure</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="ex">tree</span> <span class="at">-C</span> data</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre class="bash-out"><code># (The colors you should see in the terminal aren't shown here)
data
‚îú‚îÄ‚îÄ fastq
‚îÇ   ‚îú‚îÄ‚îÄ ERR10802863_R1.fastq.gz
‚îÇ   ‚îú‚îÄ‚îÄ ERR10802863_R2.fastq.gz
‚îÇ   ‚îú‚îÄ‚îÄ ERR10802864_R1.fastq.gz
‚îÇ   ‚îú‚îÄ‚îÄ ERR10802864_R2.fastq.gz
‚îÇ   ‚îú‚îÄ‚îÄ ERR10802865_R1.fastq.gz
‚îÇ   ‚îú‚îÄ‚îÄ ERR10802865_R2.fastq.gz
‚îÇ   ‚îú‚îÄ‚îÄ ERR10802866_R1.fastq.gz
‚îÇ   ‚îú‚îÄ‚îÄ ERR10802866_R2.fastq.gz
‚îÇ   ‚îú‚îÄ‚îÄ ERR10802867_R1.fastq.gz
‚îÇ   ‚îú‚îÄ‚îÄ ERR10802867_R2.fastq.gz
|   ‚îú‚îÄ‚îÄ [...truncated - more FASTQ files...]
‚îú‚îÄ‚îÄ meta
‚îÇ   ‚îî‚îÄ‚îÄ metadata.tsv
‚îú‚îÄ‚îÄ README.md
‚îî‚îÄ‚îÄ ref
    ‚îú‚îÄ‚îÄ GCF_016801865.2.fna
    ‚îî‚îÄ‚îÄ GCF_016801865.2.gtf
3 directories, 48 files</code></pre>
<p>So, we have:</p>
<ul>
<li>A whole bunch of FASTQ files (<code>.fastq.gz</code> extension) which contain the HTS reads</li>
<li>Two reference genomes files representing the assembly/sequence (<code>.fna</code>) and the annotation table (<code>.gtf</code>)</li>
<li>A sample metadata file with treatment information for each sample</li>
</ul>
<p><strong>All these files are plain-text files</strong>, which is one reason why the Unix shell excells at working with them. You‚Äôll take a closer look at each below.</p>
<details>
<summary>
<i class="fa-solid fa-comments" aria-label="comments"></i> Looking at the <code>tree</code> output, what is the (full) path to the file <code>metadata.tsv</code>? <em>(Click to expand)</em>
</summary>
The path is <code>/fs/scratch/PAS2880/ENT6703/data/meta/metadata.tsv</code>.
</details>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-1-contents" aria-controls="callout-1" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Side note: How you could retrieve this data yourself <em>(Click to expand)</em>
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-1" class="callout-1-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>At the end of the <span class="citation" data-cites="garrig√≥s2025">Garrig√≥s et al. (<a href="#ref-garrig√≥s2025" role="doc-biblioref">2025</a>)</span> paper, there is a section called ‚ÄúOpen Research‚Äù with a ‚ÄúData Availability Statement‚Äù, which reads:</p>
<blockquote class="blockquote">
<p>Raw sequences generated in this study have been submitted to the European Nucleotide Archive ENA database (<a href="https://www.ebi.ac.uk/ena/browser/home" class="uri">https://www.ebi.ac.uk/ena/browser/home</a>) under project accession number PRJEB41609, Study ERP125411. Sample metadata are available at <a href="https://doi.org/10.20350/digitalCSIC/15708" class="uri">https://doi.org/10.20350/digitalCSIC/15708</a>.</p>
</blockquote>
<p>The project accession number <code>PRJEB41609</code> can be used to download the reads in various ways. Because of the large amount of data involved, it‚Äôs convenient to download them directly to OSC rather than first to your own computer. One way to do this is to use the <a href="https://sra-explorer.info/">SRA Explorer</a> website, where you can enter the project accession number and get download links so you can download the files with simple commands.</p>
<p>The second link can be used to download the sample metadata.</p>
</div>
</div>
</div>
<section id="the-sample-metadata" class="level3" data-number="2.1">
<h3 data-number="2.1" class="anchored" data-anchor-id="the-sample-metadata"><span class="header-section-number">2.1</span> The sample metadata</h3>
<p>Let‚Äôs start by taking a quick look at a table with information about the samples analyzed in this paper, which we‚Äôll call the sample metadata. This file, <code>metadata.tsv</code>, is a small tabular plain-text file, with columns separated by Tabs (TSV/<code>.tsv</code> is short for ‚ÄúTab-Separated Values‚Äù):</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ls</span> <span class="at">-lh</span> data/meta</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre class="bash-out"><code>-rw-r--r-- 1 jelmer PAS0471 644 Jan 21 09:15 metadata.tsv</code></pre>
<p>Use the <code>cat</code> command (short for catenate/concatenate) to show the file contents in the shell:</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span> meta/metadata.tsv</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre class="bash-out"><code>sample_id       dpi     treatment
ERR10802864     1       cathemerium
ERR10802867     1       cathemerium
ERR10802870     1       cathemerium
ERR10802866     1       control
ERR10802869     1       control
ERR10802863     1       control
ERR10802871     1       relictum
ERR10802874     1       relictum
ERR10802865     1       relictum
ERR10802868     1       relictum
ERR10802882     10      cathemerium
ERR10802875     10      cathemerium
ERR10802879     10      cathemerium
ERR10802883     10      cathemerium
ERR10802878     10      control
ERR10802884     10      control
ERR10802877     10      control
ERR10802881     10      control
ERR10802876     10      relictum
ERR10802880     10      relictum
ERR10802885     10      relictum
ERR10802886     10      relictum</code></pre>
<p>There are 22 samples in total, with two variables of interest:</p>
<ul>
<li><code>dpi</code> ‚Äî time of sampling in Days Post-Infection</li>
<li><code>treatment</code> ‚Äî treatment group: infection status</li>
</ul>
<p>Each <code>treatment</code>-<code>dpi</code> combination has replicates to allow for statistical comparisons. As mentioned, we‚Äôll discuss these aspects of the experimental design in more detail next week.</p>
</section>
</section>
<section id="reference-genome-files" class="level2" data-number="3">
<h2 data-number="3" class="anchored" data-anchor-id="reference-genome-files"><span class="header-section-number">3</span> Reference genome files</h2>
<p>There are two main types of reference genome files needed for a HTS project like reference-based RNA-Seq:</p>
<ol type="1">
<li><p>The <strong>genome assembly</strong>, i.e.&nbsp;the nucleotide sequence of each chromosome/scaffold. These are stored in the simple <strong>FASTA</strong> format.</p></li>
<li><p>The <strong>genome annotation</strong>, i.e.&nbsp;a table with the positions of genes and other so-called genomic features on each chromosome/scaffold. These are commonly stored in <strong>GTF</strong> or similar formats.</p></li>
</ol>
<p>Below, after seeing how you can obtain such files, we‚Äôll focus on exploring the genome assembly FASTA file.</p>
<section id="finding-a-reference-genome" class="level3" data-number="3.1">
<h3 data-number="3.1" class="anchored" data-anchor-id="finding-a-reference-genome"><span class="header-section-number">3.1</span> Finding a reference genome</h3>
<p>Imagine you are one of the researchers involved in the <em>Culex</em> study. You‚Äôd want to find out if a <em>Culex pipiens</em> reference genome is available, and if so, download the relevant files. The authors state the following in the paper:</p>
<blockquote class="blockquote">
<p>Because the reference genome and annotations of <em>Cx. pipiens</em> are not published yet, we used the reference genome and annotations of phylogenetically closest species that were available in Ensembl, <em>Cx. quinquefasciatus</em>.</p>
</blockquote>
<p>As discussed in the lecture, reference genomes of closely related species can generally be used if no genome is available for the focal species, but it‚Äôs always best to use a genome of the exact species if possible. We also saw that new reference genomes are rapidly being published. So perhaps a reference genome of <em>Culex pipiens</em> has become available in the meantime?</p>
<p>Generally, the first place to look for reference genome data is at NCBI, <a href="https://ncbi.nlm.nih.gov" class="uri">https://ncbi.nlm.nih.gov</a>, where you can start by simply typing the name of your organism in the search box at the home page. By means of example, let‚Äôs first search for the hoverfly <em>Episyrphus balteatus</em>:</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="../img/refgenome_search.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:75.0%"></p>
</figure>
</div>
<p>You should get the following box at the top of the results:</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="../img/refgenome_result1b.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:80.0%"></p>
</figure>
</div>
<p>If you then click on ‚ÄúGenomes‚Äù, you should see the following table<a href="#fn2" class="footnote-ref" id="fnref2" role="doc-noteref"><sup>2</sup></a>:</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="../img/refgenome_result2b.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:95.0%"></p>
</figure>
</div>
<p>So, NCBI has 5 genomes assemblies for <em>Episyrphus balteatus</em>. The top one:</p>
<ul>
<li>Has a green check mark next to it, which means that this genome has been designated the primary reference genome for the focal organism.</li>
<li>Is the only one with an entry in the <code>RefSeq</code> column, which means it has an (NCBI) annotation available.</li>
</ul>
<p>That top assembly, <code>idEpiBalt1.1</code>, would therefore be the one to choose, unless you have project-specific reasons to use one of the other assemblies (e.g., a genome from a specific population/lineage).</p>
<section id="exercise-a-culex-pipiens-reference-genome" class="level4 exercise">
<h4 class="anchored" data-anchor-id="exercise-a-culex-pipiens-reference-genome"><i class="fa-solid fa-user-edit" aria-label="user-edit"></i> Exercise: A <em>Culex pipiens</em> reference genome</h4>
<ol type="1">
<li><p>Now search for <em>Culex pipiens</em>. How many <em>Culex pipiens</em> genomes are on NCBI, and if so, which would you pick?</p>
<details>
<summary>
<p><em>Click for the solution</em></p>
</summary>
<p>Go through the same process as shown above for the fruitfly, now entering <code>Culex pipiens</code> in the search box.</p>
<p>You should find that there are 5 <em>Culex pipiens</em> assemblies. Very much like for the fruitfly, as it happens, the first one (<code>TS_CPP_V2</code>) is the only with a reference checkmark next to it and an entry in the <code>RefSeq</code> column:</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="../img/refgenome_culex_pipiens.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:95.0%"></p>
</figure>
</div>
<p>So, there <em>is</em> currently a reference genome for <em>Cx. pipiens</em> available, and we‚Äôll be using that one: the files you‚Äôve seen in the <code>ref</code> dir are for that genome.</p>
</details></li>
<li><p>You can click on each assembly to get more information, including statistics like the number of scaffolds. Take a closer look at our focal genome on the NCBI website. What is the size of the genome assembly? How many chromosomes and scaffolds does it contain? And how many protein-coding genes?</p>
<details>
<summary>
<p><em>Click for the solution</em></p>
</summary>
<p>On the <a href="https://www.ncbi.nlm.nih.gov/datasets/genome/GCF_016801865.2/">genome‚Äôs page at NCBI</a>, it says that the genome assembly has:</p>
<ul>
<li>A size of 566.3 Mb (Megabases)</li>
<li>3 chromosomes and 289 unplaced scaffolds</li>
<li>16,297 protein-coding genes annotated</li>
</ul>
<p>See these screenshots:</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="../img/refgenome_stats.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:40.0%"></p>
</figure>
</div>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="../img/refannot_stats.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:50.0%"></p>
</figure>
</div></details></li>
</ol>

</section>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-2-contents" aria-controls="callout-2" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Side note: How you can download reference genome files <em>(Click to expand)</em>
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-2" class="callout-2-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>If you wanted to download the reference genome files from the NCBI website, you could either select an assembly in the overview table and click the <code>Download</code> button, or click the <code>Download</code> button at the top of the page for a specific assembly. That should get you the following pop-up window:</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="../img/refgenome_download.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:55.0%"></p>
</figure>
</div>
<p>You‚Äôll want to select at least the ‚Äú<em>Genome sequences (FASTA)</em>‚Äù and one or both of the ‚Äú<em>Annotation features</em>‚Äù files (GTF is often preferred with RNA-Seq).</p>
<p>This allows you to download the data to your computer, and you could then upload it OSC. Though a faster and more reproducible solution would be to use a <em>command</em> in your OSC shell to directly download these ‚Äî the <code>datasets</code> and <code>curl</code> buttons next to the <code>Download</code> one a genome‚Äôs page help with that.</p>
</div>
</div>
</div>
<div class="callout callout-style-default callout-warning callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-3-contents" aria-controls="callout-3" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Side note: Reference genome complications <em>(Click to expand)</em>
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-3" class="callout-3-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>Other useful database for reference genomes are <a href="https://ensembl.org">Ensembl</a> and specialized databases that exist for certain organisms, like <a href="https://flybase.org/">FlyBase</a> for <em>Drosophila</em>, <a href="https://vectorbase.org/">VectorBase</a> mostly for mosquitos, and <a href="https://phytozome-next.jgi.doe.gov/">JGI Phytozome</a> for plants.</p>
<p>In many cases, these databases <em>don‚Äôt contain the exact same reference genome files than NCBI</em>. Often, at least the annotation is different, but even the assembly may have small differences including in chromosome/scaffold names, which can make these files completely incompatible. And to make matters even more complicated, it is not always clear which database is the best source for your genome.</p>
</div>
</div>
</div>
</section>
<section id="the-genome-assembly-fasta-file" class="level3" data-number="3.2">
<h3 data-number="3.2" class="anchored" data-anchor-id="the-genome-assembly-fasta-file"><span class="header-section-number">3.2</span> The genome assembly FASTA file</h3>
<section id="the-fasta-format" class="level4">
<h4 class="anchored" data-anchor-id="the-fasta-format">The FASTA format</h4>
<p>FASTA files contain one or more DNA or amino acid sequences, with no limits on the number of sequences or the sequence lengths. FASTA is a ubiquitous format in genetics and genomics ‚Äì it is the standard format for, e.g.:</p>
<ul>
<li>Genome assembly sequences</li>
<li>Transcriptomes and proteomes (all of an organism‚Äôs transcripts &amp; amino acid sequences, respectively)</li>
<li>Sequence downloads from NCBI such as a single gene/protein or other GenBank entry</li>
</ul>
<p>The following example FASTA file contains two entries:</p>
<pre class="bash-in-nocolor"><code>&gt;unique_sequence_ID Optional description
ATTCATTAAAGCAGTTTATTGGCTTAATGTACATCAGTGAAATCATAAATGCTAAAAA
&gt;unique_sequence_ID2
ATTCATTAAAGCAGTTTATTGGCTTAATGTACATCAGTGAAATCATAAATGCTAAATG</code></pre>
<p>Each FASTA entry consists of:</p>
<ol type="1">
<li>A <strong>header</strong> line that starts with a <code>&gt;</code> (greater-than sign) and provides an identifier and optionally additional information about the sequence<a href="#fn3" class="footnote-ref" id="fnref3" role="doc-noteref"><sup>3</sup></a></li>
<li>A nucleotide or amino acid <strong>sequence</strong> (which can span 1 or several/many lines)</li>
</ol>
</section>
<section id="the-culex-genome-assembly-fasta-file" class="level4">
<h4 class="anchored" data-anchor-id="the-culex-genome-assembly-fasta-file">The <em>Culex</em> genome assembly FASTA file</h4>
<p>The <em>Culex pipiens</em> reference genome files are in the <code>ref</code> directory:</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ls</span> <span class="at">-lh</span> data/ref</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre class="bash-out"><code>-rw------- 1 jelmer PAS0471 547M Jan 22 12:34 GCF_016801865.2.fna
-rw------- 1 jelmer PAS0471 123M Jan 22 12:34 GCF_016801865.2.gtf</code></pre>
<p>The FASTA file is <code>GCF_016801865.2.fna</code> and is quite large (547 Mb). <code>cat</code> isn‚Äôt a great command to take a peek at such a large file, because it will print all of it to screen üò≥.</p>
<p>Yet the Unix shell more broadly is actually excellent for exploring such files, because visual text editors (say, Notepad or TextEdit) would have lots of trouble opening these large files<a href="#fn4" class="footnote-ref" id="fnref4" role="doc-noteref"><sup>4</sup></a>.</p>
<p>A great command to easily view text files of any size is <strong><code>less</code></strong>, which opens them up in a ‚Äúpager‚Äù within your shell ‚Äì you‚Äôll see what that means if you try it with one of the assembly FASTA file:</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="fu">less</span> data/ref/GCF_016801865.2.fna</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre class="bash-out"><code>&gt;NC_068937.1 Culex pipiens pallens isolate TS chromosome 1, TS_CPP_V2, whole genome shotgun sequence
aagcccttttatggtcaaaaatatcgtttaacttgaatatttttccttaaaaaataaataaatttaagcaaacagctgag
tagatgtcatctactcaaatctacccataagcacacccctgttcaatttttttttcagccataagggcgcctccagtcaa
attttcatattgagaatttcaatacaattttttaagtcgtaggggcgcctccagtcaaattttcatattgagaatttcaa
tacatttttttatgtcgtaggggcgcctccagtcaaattttcatattgagaatttcaatacattttttttaagtcgtagg
ggcgcctccagtcaaattttcatattgagaatttcaatacatttttttaagtcttaggggcgcctccagtcaaattttca
tattgagaatttcaatacatttttttaagtcgtaggggcgcctccagtcaaattttcatattgagaattttaatacaatt
ttttaaatcctaggggcgccttcagacaaacttaatttaaaaaatatcgctcctcgacttggcgactttgcgactgactg
cgacagcactaccttggaacactgaaatgtttggttgactttccagaaagagtgcatatgacttgaaaaaaaaagagcgc
ttcaaaattgagtcaagaaattggtgaaacttggtgcaagcccttttatggttaaaaatatcgtttaacttgaatatttt
tccttaaaaaataaataaatttaagcaaacagctgagtagatgtcatctactcaaatctacccataagcacacccctgga
CCTAATTCATGGAGGTGAATAGAGCATACGTAAATACAAAACTCATGACATTAGCCTGTAAGGATTGTGTaattaatgca
aaaatattgaTAGAATGAAAGATGCAAGTCccaaaaattttaagtaaatgaATAGTAATCATAAAGATAActgatgatga</code></pre>
</section>
<section id="exercise-explore-the-file-with-less" class="level4 exercise">
<h4 class="anchored" data-anchor-id="exercise-explore-the-file-with-less"><i class="fa-solid fa-user-edit" aria-label="user-edit"></i> Exercise: Explore the file with <code>less</code></h4>
<p>After running the above command, the file should have ‚Äúopened‚Äù inside the <code>less</code> pager.</p>
<p>Move around in the file and explore it a bit, which you can do in several ways: by scrolling with your mouse, with up and down arrows, or, if you have them, <kbd>PgUp</kbd> and <kbd>PgDn</kbd> keys (also, <kbd>u</kbd> will move up half a page and <kbd>d</kbd> will move down half a page).</p>
<p>How many FASTA entries are you seeing in the part of the file you explored?</p>
<div class="callout callout-style-simple callout-tip">
<div class="callout-body d-flex">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-body-container">
<p>Notice that you are ‚Äúinside‚Äù this pager and won‚Äôt have your shell prompt back until you <strong>press</strong> <kbd>q</kbd> to quit <code>less</code>.</p>
</div>
</div>
</div>
</section>
<div class="callout callout-style-default callout-warning callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-4-contents" aria-controls="callout-4" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Side note: Lowercase vs.&nbsp;uppercase nucleotide letters <em>(Click to expand)</em>
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-4" class="callout-4-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>As you have probably noticed, nucleotide bases are typically typed in uppercase (<code>A</code>, <code>C</code>, <code>G</code>, <code>T</code>). What does the mixture of lowercase and uppercase bases in the <em>Cx. pipiens</em> assembly FASTA mean, then?</p>
<p>Lowercase bases are ‚Äúsoft-masked‚Äù: they are repetitive sequences, and bioinformatics programs can treat them differently than non-repetitive sequences, which are in uppercase.</p>
</div>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-5-contents" aria-controls="callout-5" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Side note: Several file extensions are used for FASTA files (<em>Click to expand</em>)
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-5" class="callout-5-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<ul>
<li>You may see files with any of the following file extensions: <code>.fa</code>, <code>.fasta</code>, <code>.fna</code>, <code>.faa</code></li>
<li>Of these, ‚Äúgeneric‚Äù extensions are <code>.fasta</code> and <code>.fa</code> (e.g: <code>culex_assembly.fasta</code>)</li>
<li>The other extensions explicitly indicate whether sequences are <em>nucleotides</em> (<code>.fna</code>) or <em>amino acids</em> (<code>.faa</code>)</li>
</ul>
<p>These extensions are mainly used to help users quickly identify the sequence type: the underlying format is the same in all cases.</p>
</div>
</div>
</div>
</section>
<section id="annotation-files" class="level3" data-number="3.3">
<h3 data-number="3.3" class="anchored" data-anchor-id="annotation-files"><span class="header-section-number">3.3</span> Annotation files</h3>
<p>As you‚Äôve seen, the genome assembly FASTA file merely contains the sequences of the chromosomes/scaffolds. How do you know where on these sequences genes are located? That‚Äôs where the <em>annotation</em> comes in, which in this case is stored in a GTF file (<code>GCF_016801865.2.gtf</code>).</p>
<p>We won‚Äôt explore the GTF annotation file (but check the <a href="#appendix">Appendix</a> in your own time if you‚Äôre interested), but it is basically a large table with information about the positions of genes and other genomic features. While the actual format isn‚Äôt quite as simple as this, the basic idea is as follows:</p>
<pre class="bash-out-solo"><code>chromosome      feature  start  end   strand
--------------------------------------------
chromosome1     gene1    17     1356   +
chromosome1     gene2    4902   5041   +</code></pre>
<p>Among other things, this information allows you (and tools that can do it for you) to <em>extract</em> gene sequences from the genome assembly FASTA file.</p>
</section>
</section>
<section id="fastq-files" class="level2" data-number="4">
<h2 data-number="4" class="anchored" data-anchor-id="fastq-files"><span class="header-section-number">4</span> FASTQ files</h2>
<section id="the-fastq-format" class="level3" data-number="4.1">
<h3 data-number="4.1" class="anchored" data-anchor-id="the-fastq-format"><span class="header-section-number">4.1</span> The FASTQ format</h3>
<p>FASTQ is the standard HTS read data file format. Just like the FASTA format, it contains sequences, but in FASTQ, these are accompanied by <em>quality scores</em> (hence FASTQ as in ‚ÄúQ‚Äù for ‚Äúquality‚Äù).</p>
<p>Each read is represented by <strong>four lines</strong>:</p>
<ol type="1">
<li>A <strong>header</strong> that starts with <strong><code>@</code></strong> and e.g.&nbsp;uniquely identifies the read</li>
<li>The <strong>sequence</strong> itself</li>
<li>A plus sign, <strong><code>+</code></strong></li>
<li>One-character ASCII <strong>quality scores</strong> for each base</li>
</ol>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><a href="../img/fastq_dark.png" class="lightbox" data-gallery="quarto-lightbox-gallery-1" title="The FASTQ format."><img src="../img/fastq_dark.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:92.0%" alt="An annotated screenshot of two reads in a FASTQ file."></a></p>
</figure>
</div>
<figcaption>The FASTQ format.</figcaption>
</figure>
</div>
<p>Each ASCII character in the quality line corresponds to a numeric ‚ÄúPhred‚Äù quality score (<span class="math inline">\(Q\)</span>) that is a function of the estimated <em>error</em> probability for a base call (<span class="math inline">\(P\)</span>):</p>
<p><span class="math display">\[
Q = -10 * log_{10}(P)
\]</span></p>
<p>Some specific probabilities and their rough qualitative interpretation for Illumina data:</p>
<table class="table-striped table-hover caption-top table">
<colgroup>
<col style="width: 25%">
<col style="width: 25%">
<col style="width: 22%">
<col style="width: 25%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: center;">Phred quality score</th>
<th>Error probability</th>
<th>Rough interpretation</th>
<th style="text-align: center;">ASCII in FASTQ</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: center;"><strong>10</strong></td>
<td>1 in 10</td>
<td>terrible</td>
<td style="text-align: center;"><code>+</code></td>
</tr>
<tr class="even">
<td style="text-align: center;"><strong>20</strong></td>
<td>1 in 100</td>
<td>bad</td>
<td style="text-align: center;"><code>5</code></td>
</tr>
<tr class="odd">
<td style="text-align: center;"><strong>30</strong></td>
<td>1 in 1,000</td>
<td>good</td>
<td style="text-align: center;"><code>?</code></td>
</tr>
<tr class="even">
<td style="text-align: center;"><strong>40</strong></td>
<td>1 in 10,000</td>
<td>excellent</td>
<td style="text-align: center;"><code>I</code></td>
</tr>
</tbody>
</table>
<hr style="height:1pt; visibility:hidden;">
<details>
<summary>
<i class="fa-solid fa-comments" aria-label="comments"></i> But why use those ASCII characters and not just numbers in the quality line? <em>(Click to see the solution)</em>
</summary>
<p>First, to be clear, the question that arises is this: the numeric Phred quality score is represented in FASTQ files by a ‚ÄúASCII character‚Äù corresponding to the number rather than the number itself. That seems roundabout and harder to understand. So why is this done?</p>
<p>This is done because it allows for a <strong>single-character</strong> representation of each possible score ‚Äî such that <em>each quality score character can correspond to (&amp; line up with)</em> <em>a base character in the read.</em></p>
For your reference, <a href="https://www.drive5.com/usearch/manual/quality_score.html">here is a complete lookup table</a> to convert the ASCII characters to Phred quality scores (look at the top table, <code>BASE=33</code>).
<hr style="height:1pt; visibility:hidden;">
</details>
</section>
<section id="fastq-file-sizes-and-organization" class="level3" data-number="4.2">
<h3 data-number="4.2" class="anchored" data-anchor-id="fastq-file-sizes-and-organization"><span class="header-section-number">4.2</span> FASTQ file sizes and organization</h3>
<p>Let‚Äôs take another look at the list of FASTQ files:</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ls</span> <span class="at">-lh</span> data/fastq</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre class="bash-out"><code>-rw-r--r-- 1 jelmer PAS0471 21M Jan 21 13:36 ERR10802863_R1.fastq.gz
-rw-r--r-- 1 jelmer PAS0471 22M Jan 21 13:36 ERR10802863_R2.fastq.gz
-rw-r--r-- 1 jelmer PAS0471 21M Jan 21 13:36 ERR10802864_R1.fastq.gz
-rw-r--r-- 1 jelmer PAS0471 22M Jan 21 13:36 ERR10802864_R2.fastq.gz
-rw-r--r-- 1 jelmer PAS0471 22M Jan 21 13:36 ERR10802865_R1.fastq.gz
-rw-r--r-- 1 jelmer PAS0471 22M Jan 21 13:36 ERR10802865_R2.fastq.gz
[...truncated...]</code></pre>
<p>In the file listing above:</p>
<ul>
<li><p>First, take note of the file sizes. They are merely about 22 Mb in size, and all have a very similar size. This is because I ‚Äúsubsampled‚Äù the FASTQ files, only retaining 500,000 reads per file. <em>The original files were on average over 1 Gb in size with about 30 million reads.</em></p></li>
<li><p>Second, if you look closely at the file names, it looks like there are two FASTQ files per sample: one with <code>_R1</code> at the end of the file name, and one with <code>_R2</code>.</p></li>
</ul>
<details>
<summary>
<i class="fa-solid fa-comments" aria-label="comments"></i> What might each of these two files per sample represent/contain? <em>(Click for the solution)</em>
</summary>
These contain the forward reads (<code>_R1.fastq.gz</code>) vs.&nbsp;the reverse reads (<code>_R2.fastq.gz</code>).
</details>
<details>
<summary>
<i class="fa-solid fa-comments" aria-label="comments"></i> Do you have any idea why the file extension ends in <code>.gz</code>? <em>(Click for the solution)</em>
</summary>
This means the file is <strong>compressed</strong>, specifically with a program called ‚Äúgzip‚Äù. This saves a lot of space: compressed files can be up to 10 times smaller than uncompressed files. Most bioinformatics tools, including the FastQC program we‚Äôll run in a bit, can work with gzipped files directly, so there is no need to decompress them.
</details>
</section>
<section id="viewing-fastq-files" class="level3" data-number="4.3">
<h3 data-number="4.3" class="anchored" data-anchor-id="viewing-fastq-files"><span class="header-section-number">4.3</span> Viewing FASTQ files</h3>
<p>Despite the file compression, you can simply use the <code>less</code> command as before to view the FASTQ files:</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="fu">less</span> data/fastq/ERR10802863_R1.fastq.gz</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre class="bash-out"><code>@ERR10802863.8435456 8435456 length=74
CAACGAATACATCATGTTTGCGAAACTACTCCTCCTCGCCTTGGTGGGGATCAGTACTGCGTACCAGTATGAGT
+
AAAAAEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEE
@ERR10802863.27637245 27637245 length=74
GCCACACTTTTGAAGAACAGCGTCATTGTTCTTAATTTTGTCGGCAACGCCTGCACGAGCCTTCCACGTAAGTT
+
AAAAAEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEAEE&lt;EEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEE</code></pre>
<section id="exercise-spotting-problematic-reads" class="level4 exercise">
<h4 class="anchored" data-anchor-id="exercise-spotting-problematic-reads"><i class="fa-solid fa-user-edit" aria-label="user-edit"></i> Exercise: Spotting problematic reads</h4>
<p>Explore the FASTQ file a bit inside <code>less</code>. Can you spot any unusual-looking reads? How do you interpret this?</p>
<details>
<summary>
<em>Click for the solution</em>
</summary>
<p>The 5<sup>th</sup> read looks as follows ‚Äî and if you scroll down you should quickly see several more reads just like this:</p>
<pre class="bash-out-solo"><code>@ERR10802863.11918285 11918285 length=35
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
+
###################################</code></pre>
<p>These only consist of <code>N</code> bases (and they are also shorter than the other reads), and are therefore completely useless!</p>
<em>(Note that it is generally <em>not</em> common to run into these kinds of failed reads as easily!)</em>
</details>
</section>
<hr style="height:1pt; visibility:hidden;">
<p>While the above problematic reads were easy to spot, you wouldn‚Äôt typically perform quality control by manually inspecting FASTQ files. There are simply too many reads to make sense of! Instead, you can use <strong>specialized tools like FastQC to summarize</strong> FASTQ files ‚Äì that‚Äôs what we‚Äôll do now.</p>
</section>
</section>
<section id="fastqc-for-quality-control-of-hts-reads" class="level2" data-number="5">
<h2 data-number="5" class="anchored" data-anchor-id="fastqc-for-quality-control-of-hts-reads"><span class="header-section-number">5</span> FastQC for quality control of HTS reads</h2>
<p>FastQC is a very useful and widely used tool to perform <strong>quality control of HTS reads</strong>. For the budding computational biologist, it is also a good first example of a bioinformatics/HTS analysis tool with a command-line interface.</p>
<section id="running-fastqc" class="level3" data-number="5.1">
<h3 data-number="5.1" class="anchored" data-anchor-id="running-fastqc"><span class="header-section-number">5.1</span> Running FastQC</h3>
<p>Command-line programs like FastQC are operated quite similarly to the Unix shell commands you‚Äôve seen. To run FastQC, you use the command <code>fastqc</code>.</p>
<p>To analyze a FASTQ files with default FastQC settings, a complete FastQC command would simply be <code>fastqc</code>, followed by the name of the file as the argument:</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="ex">fastqc</span> data/fastq/ERR10802863_R1.fastq.gz</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre class="bash-out"><code>bash: fastqc: command not found...</code></pre>
<section id="loading-the-fastqc-module" class="level4">
<h4 class="anchored" data-anchor-id="loading-the-fastqc-module">Loading the FastQC ‚Äúmodule‚Äù</h4>
<p>But of course, it can‚Äôt be quite that simple ‚Äì we get an error! While FastQC <em>is</em> installed at OSC<a href="#fn5" class="footnote-ref" id="fnref5" role="doc-noteref"><sup>5</sup></a>, we have to <em>‚Äúload it‚Äù</em> before use. The finer details of using software at OSC are outside of the scope of this lab, so here, please accept that we can load FastQC as follows:</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="ex">module</span> load fastqc/0.12.1</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="changing-the-output-dir" class="level4">
<h4 class="anchored" data-anchor-id="changing-the-output-dir">Changing the output dir</h4>
<p>Before trying again, you‚Äôll also make a chance to the command. An ill-advised default behavior of FastQC is that it writes output files in the dir where the input files are ‚Äì it‚Äôs not good practice to directly mix data and results like that!</p>
<p>To figure out how we can change that behavior, note that many commands and programs have an <strong>option <code>-h</code> and/or <code>--help</code></strong> to print usage information to the screen.</p>
</section>
<section id="exercise-finding-the-output-directory-option" class="level4 exercise">
<h4 class="anchored" data-anchor-id="exercise-finding-the-output-directory-option"><i class="fa-solid fa-user-edit" aria-label="user-edit"></i> Exercise: Finding the output directory option</h4>
<p>Print FastQC‚Äôs help info, and figure out which option you can use to specify an output dir of your choice.</p>
<details>
<summary>
<em>Click for the solution</em>
</summary>
<p>Both <code>fastqc -h</code> and <code>fastqc --help</code> will show the help info. You‚Äôll get quite a bit of output printed to screen, including this snippet about output directories:</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="ex">fastqc</span> <span class="at">-h</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre class="bash-out"><code>  -o --outdir  Create all output files in the specified output directory.
               Please note that this directory must exist as the program
               will not create it.  If this option is not set then the 
               output file for each sequence file is created in the same
               directory as the sequence file which was processed.</code></pre>
This means that you can use <code>-o</code> or equivalently, <code>--outdir</code> to specify an output dir. (Commands often have both short- and long-form options just for convenience.)
</details>
</section>
<hr style="height:1pt; visibility:hidden;">
<p>You‚Äôll want to have FastQC write output to a deicated dir within the dir you created in the first part of the lab. For me, for example, a good choice would be <code>jelmer/fastqc</code>:</p>
<div class="sourceCode" id="cb24"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Replace 'jelmer' with your own dir name </span></span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a><span class="ex">fastqc</span> <span class="at">--outdir</span> jelmer/fastqc data/fastq/ERR10802863_R1.fastq.gz</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre class="bash-out"><code>Specified output directory 'jelmer/fastqc' does not exist</code></pre>
<p>But we get an error! üò°</p>
<section id="exercise-fixing-the-output-directory-problem" class="level4 exercise">
<h4 class="anchored" data-anchor-id="exercise-fixing-the-output-directory-problem"><i class="fa-solid fa-user-edit" aria-label="user-edit"></i> Exercise: Fixing the output directory problem</h4>
<p>What is going on this time? Can you try to fix the problem?</p>
<details>
<summary>
<em>If you know what the problem is, click here for a hint on how to fix it</em>
</summary>
<p>You can create a new directory in the OnDemand file explorer, or use a command to create dirs.</p>
<p>For the latter option: the command <code>mkdir</code> can be used to create new dirs, where you give the name/path of the dir you want to create as the argument:</p>
<div class="sourceCode" id="cb26"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Replace 'DESIRED-OUTPUT-DIR' with the actual name we want to give the new dir</span></span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> DESIRED-OUTPUT-DIR</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<details>
<summary>
<em>Click here for the solution</em>
</summary>
<ul>
<li><p>The problem, as the error message indicates, is that the output directory that we specified with <code>--outdir</code> does not exist. We might have expected FastQC to be smart/flexible enough to create this dir for us, but alas. To be fair, it did warn us about this in the Help info!</p></li>
<li><p>You can create this dir in the OnDemand file explorer, but here I will show how to do it with a command: <code>mkdir</code> can be used to create new dirs.</p>
<div class="sourceCode" id="cb27"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> jelmer/fastqc</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div></li>
</ul>
</details>
</section>
<p>Now that you‚Äôve created the output directory, let‚Äôs give it a final try before throwing our laptops out the window:</p>
<div class="sourceCode" id="cb28"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="ex">fastqc</span> <span class="at">--outdir</span> jelmer/fastqc data/fastq/ERR10802863_R1.fastq.gz</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre class="bash-out"><code>application/gzip
Started analysis of ERR10802863_R1.fastq.gz
Approx 5% complete for ERR10802863_R1.fastq.gz
Approx 10% complete for ERR10802863_R1.fastq.gz
Approx 15% complete for ERR10802863_R1.fastq.gz
Approx 20% complete for ERR10802863_R1.fastq.gz
Approx 25% complete for ERR10802863_R1.fastq.gz
[...truncated...]
Approx 100% complete for ERR10802863_R1.fastq.gz
Analysis complete for ERR10802863_R1.fastq.gz</code></pre>
<p><strong>Success!!</strong> ü•≥</p>
</section>
<section id="fastqc-output-files" class="level3" data-number="5.2">
<h3 data-number="5.2" class="anchored" data-anchor-id="fastqc-output-files"><span class="header-section-number">5.2</span> FastQC output files</h3>
<p>Let‚Äôs take a look at the files in the output dir you specified:</p>
<div class="sourceCode" id="cb30"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="co"># [Replace 'jelmer' with your own dir name]</span></span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ls</span> <span class="at">-lh</span> jelmer/fastqc</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre class="bash-out"><code>total 1.1M
-rw-r--r-- 1 jelmer PAS0471 718K Jan 27 13:12 ERR10802863_R1_fastqc.html
-rw-r--r-- 1 jelmer PAS0471 364K Jan 27 13:12 ERR10802863_R1_fastqc.zip</code></pre>
<ul>
<li>There is a <code>.zip</code> file ‚Äî this contains <strong>tables</strong> with FastQC‚Äôs data summaries</li>
<li>There is an <code>.html</code> (HTML) file ‚Äì this contains QC plots and is what we‚Äôll look at next</li>
</ul>
<section id="exercise-run-fastqc-for-the-r2-file" class="level4 exercise">
<h4 class="anchored" data-anchor-id="exercise-run-fastqc-for-the-r2-file"><i class="fa-solid fa-user-edit" aria-label="user-edit"></i> Exercise: Run FastQC for the R2 file</h4>
<p>Run FastQC for the corresponding R2 FASTQ file. Would you use the same output dir?</p>
<details>
<summary>
<em>Click here for the solution</em>
</summary>
<p>Yes, it makes sense to use the same output dir, since as you could see above, the output file names have the input file identifiers in them. As such, you don‚Äôt need to worry about overwriting files, and it will be easier to have all the results in a single dir.</p>
<p>To run FastQC for the R2 (=reverse-read) FASTQ file:</p>
<div class="sourceCode" id="cb32"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="ex">fastqc</span> <span class="at">--outdir</span> jelmer/fastqc data/fastq/ERR10802863_R2.fastq.gz</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre class="bash-out"><code>Started analysis of ERR10802863_R2.fastq.gz
Approx 5% complete for ERR10802863_R2.fastq.gz
Approx 10% complete for ERR10802863_R2.fastq.gz
Approx 15% complete for ERR10802863_R2.fastq.gz
[...truncated...]
Analysis complete for ERR10802863_2.fastq.gz</code></pre>
<div class="sourceCode" id="cb34"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ls</span> <span class="at">-lh</span> jelmer/fastqc</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre class="bash-out"><code>-rw-r--r-- 1 jelmer PAS0471 241K Jan 21 21:50 ERR10802863_R1_fastqc.html
-rw-r--r-- 1 jelmer PAS0471 256K Jan 21 21:50 ERR10802863_R1_fastqc.zip
-rw-r--r-- 1 jelmer PAS0471 234K Jan 21 21:53 ERR10802863_R2_fastqc.html
-rw-r--r-- 1 jelmer PAS0471 244K Jan 21 21:53 ERR10802863_R2_fastqc.zip</code></pre>
Now, you have four FastQC output files: two for each of your preceding successful FastQC runs.
</details>
</section>
</section>
</section>
<section id="interpreting-fastqcs-output" class="level2" data-number="6">
<h2 data-number="6" class="anchored" data-anchor-id="interpreting-fastqcs-output"><span class="header-section-number">6</span> Interpreting FastQC‚Äôs output</h2>
<p>First, you‚Äôll have to download FastQC‚Äôs output HTML files, because you can‚Äôt view them in the Unix shell<a href="#fn6" class="footnote-ref" id="fnref6" role="doc-noteref"><sup>6</sup></a>.</p>
<ol type="1">
<li>Find the FastQC HTML files in the OnDemand Files menu file explorer.</li>
<li>Select the two files by clicking the check-box at the right, and click the Download bottom towards the top.</li>
<li>Download the file to somewhere on your computer (doesn‚Äôt matter where), and your browser should give you a pop-up with the file you can click on to open it.</li>
<li>Since the file is in HTML format, it should <strong>open in your web browser</strong>.</li>
</ol>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><a href="../img/ondemand-fastqc-download.png" class="lightbox" data-gallery="quarto-lightbox-gallery-2" title="With the two HTML files selected, you can click the Download button at the top"><img src="../img/ondemand-fastqc-download.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:80.0%" alt="Screenshot of the OnDemand file explorer with FastQC HTML files selected for download."></a></p>
</figure>
</div>
<figcaption>With the two HTML files selected, you can click the Download button at the top</figcaption>
</figure>
</div>
<p>We‚Äôll now go through a couple of the FastQC plots/modules, with first some example plots<a href="#fn7" class="footnote-ref" id="fnref7" role="doc-noteref"><sup>7</sup></a> with good/bad results for reference.</p>
<section id="overview-of-module-results" class="level3" data-number="6.1">
<h3 data-number="6.1" class="anchored" data-anchor-id="overview-of-module-results"><span class="header-section-number">6.1</span> Overview of module results</h3>
<p>FastQC has ‚Äú<strong>pass</strong>‚Äù (checkmark in green), ‚Äú<strong>warning</strong>‚Äù (exclamation mark in orange), and ‚Äú<strong>fail</strong>‚Äù (cross in red) assessments for each module, as you can see below.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="../img/fastqc_summary2.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:30.0%"></p>
</figure>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-6-contents" aria-controls="callout-6" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
More about the FastQC module assessments <em>(Click to expand)</em>
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-6" class="callout-6-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>These pass/warning/fail assessments are handy and typically at least somewhat meaningful, but you should realize that <em>a ‚Äúwarning‚Äù or a ‚Äúfail‚Äù is not necessarily the bad news that it may appear to be</em>. That‚Äôs because, for example:</p>
<ul>
<li>Some of these modules are rather strict, perhaps overly strict.</li>
<li>Some warnings and fails are easily remedied or simply not a very big deal.</li>
<li>FastQC assumes that your data is derived from whole-genome shotgun sequencing ‚Äî some other types of data like RNA-Seq data will always trigger a couple of warnings and files based on expected differences.</li>
</ul>
</div>
</div>
</div>
</section>
<section id="basic-statistics-module" class="level3" data-number="6.2">
<h3 data-number="6.2" class="anchored" data-anchor-id="basic-statistics-module"><span class="header-section-number">6.2</span> Basic statistics module</h3>
<p>This section shows, for example, the number of sequences (reads) and the read length range for your file:</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="../img/fastqc_mod01_basic-stats.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:85.0%"></p>
</figure>
</div>
</section>
<section id="per-base-quality-graph" class="level3" data-number="6.3">
<h3 data-number="6.3" class="anchored" data-anchor-id="per-base-quality-graph"><span class="header-section-number">6.3</span> Per base quality graph</h3>
<p>This figure visualize the mean per-base Phred quality score (as discussed above; y-axis) along the length of the reads (x-axis). Note that:</p>
<ul>
<li>A decrease in sequence quality along the reads is normal</li>
<li>R2 (reverse) reads are usually worse than R1 (forward) reads</li>
</ul>
<div class="columns">
<div class="column" style="width:50%;">
<p><strong>Good / acceptable:</strong></p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="../img/fastqc_good.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:100.0%"></p>
</figure>
</div>
</div><div class="column" style="width:50%;">
<p><strong>Bad:</strong></p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="../img/fastqc_bad.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:100.0%"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="adapter-content-graph" class="level3" data-number="6.4">
<h3 data-number="6.4" class="anchored" data-anchor-id="adapter-content-graph"><span class="header-section-number">6.4</span> Adapter content graph</h3>
<p>Checks for known adapter sequences. As you‚Äôve learned, when an insert size is shorter than the read length, adapters will end up in the sequence ‚Äì these should be removed!</p>
<div class="columns">
<div class="column" style="width:50%;">
<p><strong>Good:</strong></p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="../img/fastqc_mod11_adapter_good.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:100.0%"></p>
</figure>
</div>
</div><div class="column" style="width:50%;">
<p><strong>Bad:</strong></p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="../img/fastqc_mod11_adapter_bad.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:100.0%"></p>
</figure>
</div>
</div>
</div>
<section id="exercise-interpreting-the-fastqc-output" class="level4 exercise">
<h4 class="anchored" data-anchor-id="exercise-interpreting-the-fastqc-output"><i class="fa-solid fa-user-edit" aria-label="user-edit"></i> Exercise: Interpreting the FastQC output</h4>
<p>Open the HTML file for the R1 FASTQ file and go through the modules we discussed above. Can you make sense of it? Does the data look good to you, overall?</p>
<p>Now open the HTML file for the R2 FASTQ file and take a look just at the quality score graph. Does it look any worse than the R1?</p>
</section>
</section>
</section>
<section id="prep-for-next-weeks-lab" class="level2" data-number="7">
<h2 data-number="7" class="anchored" data-anchor-id="prep-for-next-weeks-lab"><span class="header-section-number">7</span> Prep for next week‚Äôs lab</h2>
<p>In next week‚Äôs lab, you will be working in R, using RStudio, to analyze the gene expression counts resulting from this RNA-Seq data. You will have <a href="../pages/r_homework.html">homework</a> to prepare for that lab. To make that a bit easier for you, I‚Äôll now demonstrate how to start an RStudio session at OSC:</p>
<div class="callout callout-style-simple callout-tip">
<div class="callout-body d-flex">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-body-container">
<p>It‚Äôs necessary to fill out this form (instead of just opening RStudio) because RStudio runs on a compute node rather than a login node. Whenever you want to access a compute node, you need to request resources for it, like this form does.</p>
</div>
</div>
</div>
<ol type="1">
<li><p>Go to the <a href="https://ondemand.osc.edu">OnDemand homepage</a> as before</p></li>
<li><p>Click on <strong><code>Interactive Apps</code></strong> (top bar) and then <strong><code>RStudio Server</code></strong> (all the way at the bottom)</p></li>
<li><p>Fill out the form as follows:</p>
<ul>
<li>Cluster: <strong><code>cardinal</code></strong></li>
<li>R version: <strong><code>4.4.0</code></strong></li>
<li>Project: <strong><code>PAS2880</code></strong></li>
<li>Number of hours: <strong><code>4</code></strong></li>
<li>Node type: <strong><code>any</code></strong></li>
<li>Number of cores: <strong><code>2</code></strong></li>
</ul></li>
<li><p>Click the big blue <strong><code>Launch</code></strong> button at the bottom</p></li>
<li><p>Now, you should be sent to a new page with a box at the top for your RStudio Server ‚Äújob‚Äù, which should initially be ‚ÄúQueued‚Äù (waiting to start).</p></li>
<li><p>Your job should start running very soon, with the top bar of the box turning green and saying ‚ÄúRunning‚Äù.</p>
<details>
<summary>
<p><em>Click to see a screenshot</em></p>
</summary>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="../img/rstudio_running.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:70.0%"></p>
</figure>
</div>
</details></li>
<li><p>Click <strong><code>Connect to RStudio Server</code></strong> at the bottom of the box, and an RStudio Server instance will open in a new browser tab. You‚Äôre ready to go!</p></li>
</ol>
</section>
<section id="in-closing" class="level2" data-number="8">
<h2 data-number="8" class="anchored" data-anchor-id="in-closing"><span class="header-section-number">8</span> In closing</h2>
<p>In today‚Äôs lab, you were introduced to:</p>
<ul>
<li>Working at the Ohio Supercomputer Center</li>
<li>Using the Unix shell</li>
<li>Reference genome files &amp; where to find these</li>
<li>HTS read FASTQ files and how to quality-control these</li>
<li>How to run a command-line bioinformatics tool</li>
</ul>
<p>Taking a step back, I‚Äôve shown you the main pieces of the computational infrastructure for what we may call ‚Äú<strong>command-line genomics</strong>‚Äù: genomics analysis using command-line tools. And you‚Äôve seen a basic example of loading and running a command-line tool at OSC.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-7-contents" aria-controls="callout-7" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Side note: scaling the analysis &amp; next steps for real projects (Click to expand)
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-7" class="callout-7-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>For this brief intro to HTS data analysis, we‚Äôve kept things simple. In a typical HTS data analysis project, however, there are some more pieces to the puzzle that we didn‚Äôt cover today:</p>
<ul>
<li>Using a text editor with an integrated terminal (e.g., VS Code) to be able to write and edit scripts more easily</li>
<li>Putting commands to run programs FastQC in a <strong>‚Äúshell script‚Äù</strong>.</li>
<li><strong>Submitting the script</strong> as a ‚Äúbatch job‚Äù.</li>
<li>To make speed things, using the OSC‚Äôs capabilities, we can <strong>submit multiple jobs <em>in parallel</em> using a loop</strong>.</li>
</ul>
<p>If it seems that speed and computing power may not be an issue, given how fast FastQC ran, keep in mind that:</p>
<ul>
<li>We here worked with subsampled (much smaller than usual) FASTQ files</li>
<li>We only ran FastQC for one of our 23 samples, and your experiment may have 50+ samples</li>
<li>We need to run a bunch more tools, and some of those take much longer to run or need lots of computer resources.</li>
</ul>
<p>All that said, those missing pieces mentioned above are outside the scope of this short introduction ‚Äî but if you managed today, it should not be hard to learn those skills either.</p>
</div>
</div>
</div>
<p><br></p>
<hr>
<hr>
</section>
<section id="appendix-gtf-files" class="level2" data-number="9">
<h2 data-number="9" class="anchored" data-anchor-id="appendix-gtf-files"><span class="header-section-number">9</span> Appendix: GTF files</h2>
<section id="the-gtf-format" class="level3" data-number="9.1">
<h3 data-number="9.1" class="anchored" data-anchor-id="the-gtf-format"><span class="header-section-number">9.1</span> The GTF format</h3>
<p>The GTF format is a tab-delimited tabular file format that contains <strong>genome annotations</strong>, with:</p>
<ul>
<li>One row for each annotated ‚Äú<strong>genomic feature</strong>‚Äù (gene, exon, etc.)</li>
<li>One column for each piece of information about a feature, like its genomic coordinates</li>
</ul>
<p>See the sample below, with an added header line (<em>not normally present</em>) with column names:</p>
<pre class="bash-out-solo"><code>seqname     source  feature start   end     score  strand  frame    attributes
NC_000001   RefSeq  gene    11874   14409   .       +       .       gene_id "DDX11L1"; transcript_id ""; db_xref "GeneID:100287102"; db_xref "HGNC:HGNC:37102"; description "DEAD/H-box helicase 11 like 1 (pseudogene)"; gbkey "Gene"; gene "DDX11L1"; gene_biotype "transcribed_pseudogene"; pseudo "true"; 
NC_000001   RefSeq  exon    11874   12227   .       +       .       gene_id "DDX11L1"; transcript_id "NR_046018.2"; db_xref "GeneID:100287102"; gene "DDX11L1"; product "DEAD/H-box helicase 11 like 1 (pseudogene)"; pseudo "true"; </code></pre>
<p>Some details on the more important/interesting columns:</p>
<ul>
<li><code>seqname</code> ‚Äî Name of the chromosome, scaffold, or contig</li>
<li><code>feature</code> ‚Äî Name of the feature type, e.g.&nbsp;‚Äúgene‚Äù, ‚Äúexon‚Äù, ‚Äúintron‚Äù, ‚ÄúCDS‚Äù</li>
<li><code>start</code> &amp; <code>end</code> ‚Äî Start &amp; end position of the feature</li>
<li><code>strand</code> ‚Äî Whether the feature is on the <code>+</code> (forward) or <code>-</code> (reverse) strand</li>
<li><code>attribute</code> ‚Äî A semicolon-separated list of tag-value pairs with additional information</li>
</ul>
</section>
<section id="the-culex-pipiens-gtf-file" class="level3" data-number="9.2">
<h3 data-number="9.2" class="anchored" data-anchor-id="the-culex-pipiens-gtf-file"><span class="header-section-number">9.2</span> The <em>Culex pipiens</em> GTF file</h3>
<p>Take a look at the GTF file for the <em>Culex pipiens</em> reference genome ‚Äì like before, use <code>less</code>, but now with the <code>-S</code> option:</p>
<div class="sourceCode" id="cb37"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a><span class="fu">less</span> <span class="at">-S</span> data/ref/GCF_016801865.2.gtf</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre class="bash-out"><code>#gtf-version 2.2
#!genome-build TS_CPP_V2
#!genome-build-accession NCBI_Assembly:GCF_016801865.2
#!annotation-source NCBI RefSeq GCF_016801865.2-RS_2022_12
NC_068937.1     Gnomon  gene    2046    110808  .       +       .       gene_id "LOC120427725"; transcript_id ""; db_xref "GeneID:120427725"; description "homeotic protein deformed"; gbkey "Gene"; gene "LOC120427725"; gene_biotype "protein_coding"; 
NC_068937.1     Gnomon  transcript      2046    110808  .       +       .       gene_id "LOC120427725"; transcript_id "XM_052707445.1"; db_xref "GeneID:120427725"; gbkey "mRNA"; gene "LOC120427725"; model_evidence "Supporting evidence includes similarity to: 25 Proteins"; product "homeotic protein deformed, transcript variant X3"; transcript_biotype "mRNA"; 
NC_068937.1     Gnomon  exon    2046    2531    .       +       .       gene_id "LOC120427725"; transcript_id "XM_052707445.1"; db_xref "GeneID:120427725"; gene "LOC120427725"; model_evidence "Supporting evidence includes similarity to: 25 Proteins"; product "homeotic protein deformed, transcript variant X3"; transcript_biotype "mRNA"; exon_number "1"; 
NC_068937.1     Gnomon  exon    52113   52136   .       +       .       gene_id "LOC120427725"; transcript_id "XM_052707445.1"; db_xref "GeneID:120427725"; gene "LOC120427725"; model_evidence "Supporting evidence includes similarity to: 25 Proteins"; product "homeotic protein deformed, transcript variant X3"; transcript_biotype "mRNA"; exon_number "2"; 
NC_068937.1     Gnomon  exon    70113   70962   .       +       .       gene_id "LOC120427725"; transcript_id "XM_052707445.1"; db_xref "GeneID:120427725"; gene "LOC120427725"; model_evidence "Supporting evidence includes similarity to: 25 Proteins"; product "homeotic protein deformed, transcript variant X3"; transcript_biotype "mRNA"; exon_number "3"; 
NC_068937.1     Gnomon  exon    105987  106087  .       +       .       gene_id "LOC120427725"; transcript_id "XM_052707445.1"; db_xref "GeneID:120427725"; gene "LOC120427725"; model_evidence "Supporting evidence includes similarity to: 25 Proteins"; product "homeotic protein deformed, transcript variant X3"; transcript_biotype "mRNA"; exon_number "4"; 
NC_068937.1     Gnomon  exon    106551  106734  .       +       .       gene_id "LOC120427725"; transcript_id "XM_052707445.1"; db_xref "GeneID:120427725"; gene "LOC120427725"; model_evidence "Supporting evidence includes similarity to: 25 Proteins"; product "homeotic protein deformed, transcript variant X3"; transcript_biotype "mRNA"; exon_number "5"; </code></pre>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Avoid line-wrapping with <code>less -S</code>
</div>
</div>
<div class="callout-body-container callout-body">
<p>Lines in a file may contain too many characters to fit on your screen, as will be the case for this GTF file. <code>less</code> will by default ‚Äúwrap‚Äù such lines onto the next line on your screen, but this is often confusing for files like FASTQ and tabular files like GTF. Therefore, we turned off line-wrapping above by using the <code>-S</code> option to <code>less</code>.</p>
</div>
</div>
<section id="exercise-understanding-gtf-entries" class="level4 exercise">
<h4 class="anchored" data-anchor-id="exercise-understanding-gtf-entries"><i class="fa-solid fa-user-edit" aria-label="user-edit"></i> Exercise: Understanding GTF entries</h4>
<p>The GTF file is sorted: all entries from the first line of the table, until the next time you a ‚Äúgene‚Äù entry in the third column, list features are part of the first gene.</p>
<p>Can you make sense of all the entries for the first gene in the GTF, given what you know of gene structures? How many transcripts does the gene have?</p>
<details>
<summary>
<em>Click to see some pointers</em>
</summary>
<ul>
<li>The first gene (<code>LOC120427725</code>) has 3 transcripts.</li>
<li>Each transcript has 6-7 exons, 5 CDSs, and a start and stop codon.</li>
</ul>
<p>Below, I‚Äôve printed all lines belonging to the first gene:</p>
<pre class="bash-out-solo"><code>NC_068937.1 Gnomon  gene    2046    110808  .   +   .   gene_id "LOC120427725"; transcript_id ""; db_xref "GeneID:120427725"; description "homeotic protein deformed"; gbkey "Gene"; gene "LOC120427725"; gene_biotype "protein_coding"; 
NC_068937.1 Gnomon  transcript  2046    110808  .   +   .   gene_id "LOC120427725"; transcript_id "XM_052707445.1"; db_xref "GeneID:120427725"; gbkey "mRNA"; gene "LOC120427725"; model_evidence "Supporting evidence includes similarity to: 25 Proteins"; product "homeotic protein deformed, transcript variant X3"; transcript_biotype "mRNA"; 
NC_068937.1 Gnomon  exon    2046    2531    .   +   .   gene_id "LOC120427725"; transcript_id "XM_052707445.1"; db_xref "GeneID:120427725"; gene "LOC120427725"; model_evidence "Supporting evidence includes similarity to: 25 Proteins"; product "homeotic protein deformed, transcript variant X3"; transcript_biotype "mRNA"; exon_number "1"; 
NC_068937.1 Gnomon  exon    52113   52136   .   +   .   gene_id "LOC120427725"; transcript_id "XM_052707445.1"; db_xref "GeneID:120427725"; gene "LOC120427725"; model_evidence "Supporting evidence includes similarity to: 25 Proteins"; product "homeotic protein deformed, transcript variant X3"; transcript_biotype "mRNA"; exon_number "2"; 
NC_068937.1 Gnomon  exon    70113   70962   .   +   .   gene_id "LOC120427725"; transcript_id "XM_052707445.1"; db_xref "GeneID:120427725"; gene "LOC120427725"; model_evidence "Supporting evidence includes similarity to: 25 Proteins"; product "homeotic protein deformed, transcript variant X3"; transcript_biotype "mRNA"; exon_number "3"; 
NC_068937.1 Gnomon  exon    105987  106087  .   +   .   gene_id "LOC120427725"; transcript_id "XM_052707445.1"; db_xref "GeneID:120427725"; gene "LOC120427725"; model_evidence "Supporting evidence includes similarity to: 25 Proteins"; product "homeotic protein deformed, transcript variant X3"; transcript_biotype "mRNA"; exon_number "4"; 
NC_068937.1 Gnomon  exon    106551  106734  .   +   .   gene_id "LOC120427725"; transcript_id "XM_052707445.1"; db_xref "GeneID:120427725"; gene "LOC120427725"; model_evidence "Supporting evidence includes similarity to: 25 Proteins"; product "homeotic protein deformed, transcript variant X3"; transcript_biotype "mRNA"; exon_number "5"; 
NC_068937.1 Gnomon  exon    109296  109660  .   +   .   gene_id "LOC120427725"; transcript_id "XM_052707445.1"; db_xref "GeneID:120427725"; gene "LOC120427725"; model_evidence "Supporting evidence includes similarity to: 25 Proteins"; product "homeotic protein deformed, transcript variant X3"; transcript_biotype "mRNA"; exon_number "6"; 
NC_068937.1 Gnomon  exon    109726  110808  .   +   .   gene_id "LOC120427725"; transcript_id "XM_052707445.1"; db_xref "GeneID:120427725"; gene "LOC120427725"; model_evidence "Supporting evidence includes similarity to: 25 Proteins"; product "homeotic protein deformed, transcript variant X3"; transcript_biotype "mRNA"; exon_number "7"; 
NC_068937.1 Gnomon  CDS 70143   70962   .   +   0   gene_id "LOC120427725"; transcript_id "XM_052707445.1"; db_xref "GeneID:120427725"; gbkey "CDS"; gene "LOC120427725"; product "homeotic protein deformed"; protein_id "XP_052563405.1"; exon_number "3"; 
NC_068937.1 Gnomon  CDS 105987  106087  .   +   2   gene_id "LOC120427725"; transcript_id "XM_052707445.1"; db_xref "GeneID:120427725"; gbkey "CDS"; gene "LOC120427725"; product "homeotic protein deformed"; protein_id "XP_052563405.1"; exon_number "4"; 
NC_068937.1 Gnomon  CDS 106551  106734  .   +   0   gene_id "LOC120427725"; transcript_id "XM_052707445.1"; db_xref "GeneID:120427725"; gbkey "CDS"; gene "LOC120427725"; product "homeotic protein deformed"; protein_id "XP_052563405.1"; exon_number "5"; 
NC_068937.1 Gnomon  CDS 109296  109660  .   +   2   gene_id "LOC120427725"; transcript_id "XM_052707445.1"; db_xref "GeneID:120427725"; gbkey "CDS"; gene "LOC120427725"; product "homeotic protein deformed"; protein_id "XP_052563405.1"; exon_number "6"; 
NC_068937.1 Gnomon  CDS 109726  110025  .   +   0   gene_id "LOC120427725"; transcript_id "XM_052707445.1"; db_xref "GeneID:120427725"; gbkey "CDS"; gene "LOC120427725"; product "homeotic protein deformed"; protein_id "XP_052563405.1"; exon_number "7"; 
NC_068937.1 Gnomon  start_codon 70143   70145   .   +   0   gene_id "LOC120427725"; transcript_id "XM_052707445.1"; db_xref "GeneID:120427725"; gbkey "CDS"; gene "LOC120427725"; product "homeotic protein deformed"; protein_id "XP_052563405.1"; exon_number "3"; 
NC_068937.1 Gnomon  stop_codon  110026  110028  .   +   0   gene_id "LOC120427725"; transcript_id "XM_052707445.1"; db_xref "GeneID:120427725"; gbkey "CDS"; gene "LOC120427725"; product "homeotic protein deformed"; protein_id "XP_052563405.1"; exon_number "7"; 
NC_068937.1 Gnomon  transcript  5979    110808  .   +   .   gene_id "LOC120427725"; transcript_id "XM_039592629.2"; db_xref "GeneID:120427725"; gbkey "mRNA"; gene "LOC120427725"; model_evidence "Supporting evidence includes similarity to: 24 Proteins"; product "homeotic protein deformed, transcript variant X2"; transcript_biotype "mRNA"; 
NC_068937.1 Gnomon  exon    5979    6083    .   +   .   gene_id "LOC120427725"; transcript_id "XM_039592629.2"; db_xref "GeneID:120427725"; gene "LOC120427725"; model_evidence "Supporting evidence includes similarity to: 24 Proteins"; product "homeotic protein deformed, transcript variant X2"; transcript_biotype "mRNA"; exon_number "1"; 
NC_068937.1 Gnomon  exon    52113   52136   .   +   .   gene_id "LOC120427725"; transcript_id "XM_039592629.2"; db_xref "GeneID:120427725"; gene "LOC120427725"; model_evidence "Supporting evidence includes similarity to: 24 Proteins"; product "homeotic protein deformed, transcript variant X2"; transcript_biotype "mRNA"; exon_number "2"; 
NC_068937.1 Gnomon  exon    70113   70962   .   +   .   gene_id "LOC120427725"; transcript_id "XM_039592629.2"; db_xref "GeneID:120427725"; gene "LOC120427725"; model_evidence "Supporting evidence includes similarity to: 24 Proteins"; product "homeotic protein deformed, transcript variant X2"; transcript_biotype "mRNA"; exon_number "3"; 
NC_068937.1 Gnomon  exon    105987  106087  .   +   .   gene_id "LOC120427725"; transcript_id "XM_039592629.2"; db_xref "GeneID:120427725"; gene "LOC120427725"; model_evidence "Supporting evidence includes similarity to: 24 Proteins"; product "homeotic protein deformed, transcript variant X2"; transcript_biotype "mRNA"; exon_number "4"; 
NC_068937.1 Gnomon  exon    106551  106734  .   +   .   gene_id "LOC120427725"; transcript_id "XM_039592629.2"; db_xref "GeneID:120427725"; gene "LOC120427725"; model_evidence "Supporting evidence includes similarity to: 24 Proteins"; product "homeotic protein deformed, transcript variant X2"; transcript_biotype "mRNA"; exon_number "5"; 
NC_068937.1 Gnomon  exon    109296  109660  .   +   .   gene_id "LOC120427725"; transcript_id "XM_039592629.2"; db_xref "GeneID:120427725"; gene "LOC120427725"; model_evidence "Supporting evidence includes similarity to: 24 Proteins"; product "homeotic protein deformed, transcript variant X2"; transcript_biotype "mRNA"; exon_number "6"; 
NC_068937.1 Gnomon  exon    109726  110808  .   +   .   gene_id "LOC120427725"; transcript_id "XM_039592629.2"; db_xref "GeneID:120427725"; gene "LOC120427725"; model_evidence "Supporting evidence includes similarity to: 24 Proteins"; product "homeotic protein deformed, transcript variant X2"; transcript_biotype "mRNA"; exon_number "7"; 
NC_068937.1 Gnomon  CDS 70143   70962   .   +   0   gene_id "LOC120427725"; transcript_id "XM_039592629.2"; db_xref "GeneID:120427725"; gbkey "CDS"; gene "LOC120427725"; product "homeotic protein deformed"; protein_id "XP_039448563.1"; exon_number "3"; 
NC_068937.1 Gnomon  CDS 105987  106087  .   +   2   gene_id "LOC120427725"; transcript_id "XM_039592629.2"; db_xref "GeneID:120427725"; gbkey "CDS"; gene "LOC120427725"; product "homeotic protein deformed"; protein_id "XP_039448563.1"; exon_number "4"; 
NC_068937.1 Gnomon  CDS 106551  106734  .   +   0   gene_id "LOC120427725"; transcript_id "XM_039592629.2"; db_xref "GeneID:120427725"; gbkey "CDS"; gene "LOC120427725"; product "homeotic protein deformed"; protein_id "XP_039448563.1"; exon_number "5"; 
NC_068937.1 Gnomon  CDS 109296  109660  .   +   2   gene_id "LOC120427725"; transcript_id "XM_039592629.2"; db_xref "GeneID:120427725"; gbkey "CDS"; gene "LOC120427725"; product "homeotic protein deformed"; protein_id "XP_039448563.1"; exon_number "6"; 
NC_068937.1 Gnomon  CDS 109726  110025  .   +   0   gene_id "LOC120427725"; transcript_id "XM_039592629.2"; db_xref "GeneID:120427725"; gbkey "CDS"; gene "LOC120427725"; product "homeotic protein deformed"; protein_id "XP_039448563.1"; exon_number "7"; 
NC_068937.1 Gnomon  start_codon 70143   70145   .   +   0   gene_id "LOC120427725"; transcript_id "XM_039592629.2"; db_xref "GeneID:120427725"; gbkey "CDS"; gene "LOC120427725"; product "homeotic protein deformed"; protein_id "XP_039448563.1"; exon_number "3"; 
NC_068937.1 Gnomon  stop_codon  110026  110028  .   +   0   gene_id "LOC120427725"; transcript_id "XM_039592629.2"; db_xref "GeneID:120427725"; gbkey "CDS"; gene "LOC120427725"; product "homeotic protein deformed"; protein_id "XP_039448563.1"; exon_number "7"; 
NC_068937.1 Gnomon  transcript  60854   110807  .   +   .   gene_id "LOC120427725"; transcript_id "XM_039592628.2"; db_xref "GeneID:120427725"; gbkey "mRNA"; gene "LOC120427725"; model_evidence "Supporting evidence includes similarity to: 24 Proteins"; product "homeotic protein deformed, transcript variant X1"; transcript_biotype "mRNA"; 
NC_068937.1 Gnomon  exon    60854   61525   .   +   .   gene_id "LOC120427725"; transcript_id "XM_039592628.2"; db_xref "GeneID:120427725"; gene "LOC120427725"; model_evidence "Supporting evidence includes similarity to: 24 Proteins"; product "homeotic protein deformed, transcript variant X1"; transcript_biotype "mRNA"; exon_number "1"; 
NC_068937.1 Gnomon  exon    70113   70962   .   +   .   gene_id "LOC120427725"; transcript_id "XM_039592628.2"; db_xref "GeneID:120427725"; gene "LOC120427725"; model_evidence "Supporting evidence includes similarity to: 24 Proteins"; product "homeotic protein deformed, transcript variant X1"; transcript_biotype "mRNA"; exon_number "2"; 
NC_068937.1 Gnomon  exon    105987  106087  .   +   .   gene_id "LOC120427725"; transcript_id "XM_039592628.2"; db_xref "GeneID:120427725"; gene "LOC120427725"; model_evidence "Supporting evidence includes similarity to: 24 Proteins"; product "homeotic protein deformed, transcript variant X1"; transcript_biotype "mRNA"; exon_number "3"; 
NC_068937.1 Gnomon  exon    106551  106734  .   +   .   gene_id "LOC120427725"; transcript_id "XM_039592628.2"; db_xref "GeneID:120427725"; gene "LOC120427725"; model_evidence "Supporting evidence includes similarity to: 24 Proteins"; product "homeotic protein deformed, transcript variant X1"; transcript_biotype "mRNA"; exon_number "4"; 
NC_068937.1 Gnomon  exon    109296  109660  .   +   .   gene_id "LOC120427725"; transcript_id "XM_039592628.2"; db_xref "GeneID:120427725"; gene "LOC120427725"; model_evidence "Supporting evidence includes similarity to: 24 Proteins"; product "homeotic protein deformed, transcript variant X1"; transcript_biotype "mRNA"; exon_number "5"; 
NC_068937.1 Gnomon  exon    109726  110807  .   +   .   gene_id "LOC120427725"; transcript_id "XM_039592628.2"; db_xref "GeneID:120427725"; gene "LOC120427725"; model_evidence "Supporting evidence includes similarity to: 24 Proteins"; product "homeotic protein deformed, transcript variant X1"; transcript_biotype "mRNA"; exon_number "6"; 
NC_068937.1 Gnomon  CDS 70143   70962   .   +   0   gene_id "LOC120427725"; transcript_id "XM_039592628.2"; db_xref "GeneID:120427725"; gbkey "CDS"; gene "LOC120427725"; product "homeotic protein deformed"; protein_id "XP_039448562.1"; exon_number "2"; 
NC_068937.1 Gnomon  CDS 105987  106087  .   +   2   gene_id "LOC120427725"; transcript_id "XM_039592628.2"; db_xref "GeneID:120427725"; gbkey "CDS"; gene "LOC120427725"; product "homeotic protein deformed"; protein_id "XP_039448562.1"; exon_number "3"; 
NC_068937.1 Gnomon  CDS 106551  106734  .   +   0   gene_id "LOC120427725"; transcript_id "XM_039592628.2"; db_xref "GeneID:120427725"; gbkey "CDS"; gene "LOC120427725"; product "homeotic protein deformed"; protein_id "XP_039448562.1"; exon_number "4"; 
NC_068937.1 Gnomon  CDS 109296  109660  .   +   2   gene_id "LOC120427725"; transcript_id "XM_039592628.2"; db_xref "GeneID:120427725"; gbkey "CDS"; gene "LOC120427725"; product "homeotic protein deformed"; protein_id "XP_039448562.1"; exon_number "5"; 
NC_068937.1 Gnomon  CDS 109726  110025  .   +   0   gene_id "LOC120427725"; transcript_id "XM_039592628.2"; db_xref "GeneID:120427725"; gbkey "CDS"; gene "LOC120427725"; product "homeotic protein deformed"; protein_id "XP_039448562.1"; exon_number "6"; 
NC_068937.1 Gnomon  start_codon 70143   70145   .   +   0   gene_id "LOC120427725"; transcript_id "XM_039592628.2"; db_xref "GeneID:120427725"; gbkey "CDS"; gene "LOC120427725"; product "homeotic protein deformed"; protein_id "XP_039448562.1"; exon_number "2"; 
NC_068937.1 Gnomon  stop_codon  110026  110028  .   +   0   gene_id "LOC120427725"; transcript_id "XM_039592628.2"; db_xref "GeneID:120427725"; gbkey "CDS"; gene "LOC120427725"; product "homeotic protein deformed"; protein_id "XP_039448562.1"; exon_number "6"; </code></pre>
</details>
</section>



</section>
</section>


<a onclick="window.scrollTo(0, 0); return false;" role="button" id="quarto-back-to-top"><i class="bi bi-arrow-up"></i> Back to top</a><div id="quarto-appendix" class="default"><section class="quarto-appendix-contents" role="doc-bibliography" id="quarto-bibliography"><h2 class="anchored quarto-appendix-heading">References</h2><div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" role="list">
<div id="ref-garrig√≥s2025" class="csl-entry" role="listitem">
Garrig√≥s, Marta, Guillem Ylla, Josu√© Mart√≠nez-de la Puente, Jordi Figuerola, and Mar√≠a Jos√© Ruiz-L√≥pez. 2025. <span>‚ÄúTwo Avian <em>Plasmodium</em> Species Trigger Different Transcriptional Responses on Their Vector <em>Culex <span>p</span>ipiens</em>.‚Äù</span> <em>Molecular Ecology</em> 34 (15): e17240. <a href="https://doi.org/10.1111/mec.17240">https://doi.org/10.1111/mec.17240</a>.
</div>
</div></section><section id="footnotes" class="footnotes footnotes-end-of-document" role="doc-endnotes"><h2 class="anchored quarto-appendix-heading">Footnotes</h2>

<ol>
<li id="fn1"><p> That is, an HTS project where reads can be mapped to a reference genome.<a href="#fnref1" class="footnote-back" role="doc-backlink">‚Ü©Ô∏é</a></p></li>
<li id="fn2"><p>As of January, 2026.<a href="#fnref2" class="footnote-back" role="doc-backlink">‚Ü©Ô∏é</a></p></li>
<li id="fn3"><p> There are no universal standards for the header line format, but individual programs and databases may expect specific formats.<a href="#fnref3" class="footnote-back" role="doc-backlink">‚Ü©Ô∏é</a></p></li>
<li id="fn4"><p>The reason is that they try to load them into RAM memory entirely, even if you just want to see the first few lines.<a href="#fnref4" class="footnote-back" role="doc-backlink">‚Ü©Ô∏é</a></p></li>
<li id="fn5"><p>For a full list of installed software at OSC: <a href="https://www.osc.edu/resources/available_software/software_list" class="uri">https://www.osc.edu/resources/available_software/software_list</a><a href="#fnref5" class="footnote-back" role="doc-backlink">‚Ü©Ô∏é</a></p></li>
<li id="fn6"><p> The Unix shell is great for plain-text file, but not for other types of files.<a href="#fnref6" class="footnote-back" role="doc-backlink">‚Ü©Ô∏é</a></p></li>
<li id="fn7"><p> Some of the FastQC example plots were taken from <a href="https://rtsf.natsci.msu.edu/sites/_rtsf/assets/File/FastQC_TutorialAndFAQ_080717.pdf">here</a>.<a href="#fnref7" class="footnote-back" role="doc-backlink">‚Ü©Ô∏é</a></p></li>
</ol>
</section></div></main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "Óßã";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp("https:\/\/jelmerp\.github\.io\/ENTMLGY6703-SP26");
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
<p>2026, Jelmer Poelstra, <a href="https://creativecommons.org/licenses/by-nc/4.0/">CC BY-NC 4.0</a> &nbsp; <img src="http://mirrors.creativecommons.org/presskit/buttons/88x31/png/by-nc.png" class="img-fluid" style="width:10.0%"></p>
</div>   
    <div class="nav-footer-center">
<p><img src="../img/CFAES-Horiz-RGBHEX.svg" class="img-fluid" style="width:40.0%"></p>
</div>
    <div class="nav-footer-right">
      <ul class="footer-items list-unstyled">
    <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/jelmerp/ENTMLGY6703-SP26">
      <i class="bi bi-github" role="img" aria-label="The GitHub repo for this website">
</i> 
    </a>
  </li>  
</ul>
    </div>
  </div>
</footer>
<script>var lightboxQuarto = GLightbox({"closeEffect":"zoom","descPosition":"bottom","loop":false,"openEffect":"zoom","selector":".lightbox"});
(function() {
  let previousOnload = window.onload;
  window.onload = () => {
    if (previousOnload) {
      previousOnload();
    }
    lightboxQuarto.on('slide_before_load', (data) => {
      const { slideIndex, slideNode, slideConfig, player, trigger } = data;
      const href = trigger.getAttribute('href');
      if (href !== null) {
        const imgEl = window.document.querySelector(`a[href="${href}"] img`);
        if (imgEl !== null) {
          const srcAttr = imgEl.getAttribute("src");
          if (srcAttr && srcAttr.startsWith("data:")) {
            slideConfig.href = srcAttr;
          }
        }
      } 
    });
  
    lightboxQuarto.on('slide_after_load', (data) => {
      const { slideIndex, slideNode, slideConfig, player, trigger } = data;
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(slideNode);
      }
    });
  
  };
  
})();
          </script>




</body></html>