<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.6.40">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Jelmer Poelstra">
<meta name="dcterms.date" content="2026-01-30">

<title>Homework for the week 4 lab: R basics – ENTMLGY6703-SP26 - Jelmer’s lectures</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<link href="../img/favicon.png" rel="icon" type="image/png">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting-57476271c5f0467c372e20ff5d25630e.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap-315fdf77af9df3389e506d97b91a2191.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<link href="../site_libs/quarto-contrib/fontawesome6-6.7.2/all.min.css" rel="stylesheet">
<link href="../site_libs/quarto-contrib/fontawesome6-6.7.2/latex-fontsize.css" rel="stylesheet">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


<link rel="stylesheet" href="../styles.css">
</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../index.html">
    <span class="navbar-title">ENTMLGY6703-SP26 - Jelmer’s lectures</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../index.html"> <i class="bi bi-house" role="img" aria-label="This site's homepage">
</i> 
<span class="menu-text"> </span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-week-3-hts" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Week 3: HTS</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-week-3-hts">    
        <li>
    <a class="dropdown-item" href="../pages/sequencing.html">
 <span class="dropdown-text">Lecture: High-throughput Sequencing &amp; Genomes</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../pages/infrastructure.html">
 <span class="dropdown-text">Lab part 1: Computing Environment</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../pages/data.html">
 <span class="dropdown-text">Lab part 2: Working with genome and HTS Data</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-week-4-rna-seq" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Week 4: RNA-seq</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-week-4-rna-seq">    
        <li>
    <a class="dropdown-item" href="../pages/r_homework.html">
 <span class="dropdown-text">R Homework</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../pages/rnaseq.html">
 <span class="dropdown-text">Lecture: RNA-seq</span></a>
  </li>  
        <li class="dropdown-header">Lab: Differential expression analysis</li>
    </ul>
  </li>
</ul>
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="../about.html"> 
<span class="menu-text">About</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="https://ondemand.osc.edu"> 
<span class="menu-text"><i class="fa-solid fa-desktop" aria-label="desktop"></i></span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
    <div class="dropdown">
      <a href="" title="" id="quarto-navigation-tool-dropdown-0" class="quarto-navigation-tool dropdown-toggle px-1" data-bs-toggle="dropdown" aria-expanded="false" role="link" aria-label="The GitHub repo for this website"><i class="bi bi-github"></i></a>
      <ul class="dropdown-menu" aria-labelledby="quarto-navigation-tool-dropdown-0">
          <li>
            <a class="dropdown-item quarto-navbar-tools-item" href="https://github.com/jelmerp/ENTMLGY6703-SP26">
            Source Code
            </a>
          </li>
          <li>
            <a class="dropdown-item quarto-navbar-tools-item" href="https://github.com/jelmerp/ENTMLGY6703-SP26/issues/new">
            Report an Issue
            </a>
          </li>
      </ul>
    </div>
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#introduction" id="toc-introduction" class="nav-link active" data-scroll-target="#introduction"><span class="header-section-number">1</span> Introduction</a>
  <ul class="collapse">
  <li><a href="#learning-objectives" id="toc-learning-objectives" class="nav-link" data-scroll-target="#learning-objectives"><span class="header-section-number">1.1</span> Learning objectives</a></li>
  </ul></li>
  <li><a href="#r-and-rstudio" id="toc-r-and-rstudio" class="nav-link" data-scroll-target="#r-and-rstudio"><span class="header-section-number">2</span> R and RStudio</a>
  <ul class="collapse">
  <li><a href="#why-use-r" id="toc-why-use-r" class="nav-link" data-scroll-target="#why-use-r"><span class="header-section-number">2.1</span> Why use R?</a></li>
  <li><a href="#r-vs.-rstudio" id="toc-r-vs.-rstudio" class="nav-link" data-scroll-target="#r-vs.-rstudio"><span class="header-section-number">2.2</span> R vs.&nbsp;RStudio</a></li>
  </ul></li>
  <li><a href="#start-an-rstudio-session-at-osc" id="toc-start-an-rstudio-session-at-osc" class="nav-link" data-scroll-target="#start-an-rstudio-session-at-osc"><span class="header-section-number">3</span> Start an RStudio session at OSC</a></li>
  <li><a href="#orienting-to-rstudio" id="toc-orienting-to-rstudio" class="nav-link" data-scroll-target="#orienting-to-rstudio"><span class="header-section-number">4</span> Orienting to RStudio</a></li>
  <li><a href="#r-basics" id="toc-r-basics" class="nav-link" data-scroll-target="#r-basics"><span class="header-section-number">5</span> R basics</a>
  <ul class="collapse">
  <li><a href="#r-as-calculator" id="toc-r-as-calculator" class="nav-link" data-scroll-target="#r-as-calculator"><span class="header-section-number">5.1</span> R as calculator</a></li>
  <li><a href="#the-r-prompt" id="toc-the-r-prompt" class="nav-link" data-scroll-target="#the-r-prompt"><span class="header-section-number">5.2</span> The R prompt</a></li>
  <li><a href="#comparing-things" id="toc-comparing-things" class="nav-link" data-scroll-target="#comparing-things"><span class="header-section-number">5.3</span> Comparing things</a></li>
  </ul></li>
  <li><a href="#functions-in-r" id="toc-functions-in-r" class="nav-link" data-scroll-target="#functions-in-r"><span class="header-section-number">6</span> Functions in R</a>
  <ul class="collapse">
  <li><a href="#basics-of-functions" id="toc-basics-of-functions" class="nav-link" data-scroll-target="#basics-of-functions"><span class="header-section-number">6.1</span> Basics of functions</a></li>
  <li><a href="#function-arguments" id="toc-function-arguments" class="nav-link" data-scroll-target="#function-arguments"><span class="header-section-number">6.2</span> Function arguments</a></li>
  </ul></li>
  <li><a href="#r-help-help-and" id="toc-r-help-help-and" class="nav-link" data-scroll-target="#r-help-help-and"><span class="header-section-number">7</span> R Help: <code>help()</code> and <code>?</code></a></li>
  <li><a href="#r-packages" id="toc-r-packages" class="nav-link" data-scroll-target="#r-packages"><span class="header-section-number">8</span> R packages</a></li>
  <li><a href="#r-scripts" id="toc-r-scripts" class="nav-link" data-scroll-target="#r-scripts"><span class="header-section-number">9</span> R scripts</a></li>
  <li><a href="#r-objects" id="toc-r-objects" class="nav-link" data-scroll-target="#r-objects"><span class="header-section-number">10</span> R objects</a>
  <ul class="collapse">
  <li><a href="#assigning-stuff-to-objects" id="toc-assigning-stuff-to-objects" class="nav-link" data-scroll-target="#assigning-stuff-to-objects"><span class="header-section-number">10.1</span> Assigning stuff to objects</a></li>
  <li><a href="#object-names" id="toc-object-names" class="nav-link" data-scroll-target="#object-names"><span class="header-section-number">10.2</span> Object names</a></li>
  </ul></li>
  <li><a href="#vectors" id="toc-vectors" class="nav-link" data-scroll-target="#vectors"><span class="header-section-number">11</span> Vectors</a>
  <ul class="collapse">
  <li><a href="#creating-vectors" id="toc-creating-vectors" class="nav-link" data-scroll-target="#creating-vectors"><span class="header-section-number">11.1</span> Creating vectors</a></li>
  <li><a href="#exploring-vectors" id="toc-exploring-vectors" class="nav-link" data-scroll-target="#exploring-vectors"><span class="header-section-number">11.2</span> Exploring vectors</a></li>
  <li><a href="#extracting-elements-from-vectors" id="toc-extracting-elements-from-vectors" class="nav-link" data-scroll-target="#extracting-elements-from-vectors"><span class="header-section-number">11.3</span> Extracting elements from vectors</a></li>
  </ul></li>
  <li><a href="#data-frames" id="toc-data-frames" class="nav-link" data-scroll-target="#data-frames"><span class="header-section-number">12</span> Data frames</a></li>
  <li><a href="#a-brief-introduction-to-the-tidyverse-and-pipes" id="toc-a-brief-introduction-to-the-tidyverse-and-pipes" class="nav-link" data-scroll-target="#a-brief-introduction-to-the-tidyverse-and-pipes"><span class="header-section-number">13</span> A brief introduction to the <em>tidyverse</em> and pipes</a>
  <ul class="collapse">
  <li><a href="#read_tsv-to-read-tabular-data" id="toc-read_tsv-to-read-tabular-data" class="nav-link" data-scroll-target="#read_tsv-to-read-tabular-data"><span class="header-section-number">13.1</span> <code>read_tsv()</code> to read tabular data</a></li>
  <li><a href="#select-to-pick-columns" id="toc-select-to-pick-columns" class="nav-link" data-scroll-target="#select-to-pick-columns"><span class="header-section-number">13.2</span> <code>select()</code> to pick columns</a></li>
  <li><a href="#filter-to-pick-rows-observations" id="toc-filter-to-pick-rows-observations" class="nav-link" data-scroll-target="#filter-to-pick-rows-observations"><span class="header-section-number">13.3</span> <code>filter()</code> to pick rows (observations)</a></li>
  <li><a href="#the-pipe" id="toc-the-pipe" class="nav-link" data-scroll-target="#the-pipe"><span class="header-section-number">13.4</span> The pipe (<code>|&gt;</code>)</a></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Homework for the week 4 lab: R basics</h1>
</div>


<div class="quarto-title-meta-author">
  <div class="quarto-title-meta-heading">Author</div>
  <div class="quarto-title-meta-heading">Affiliation</div>
  
    <div class="quarto-title-meta-contents">
    <p class="author">Jelmer Poelstra <a href="mailto:poelstra.1@osu.edu" class="quarto-title-author-email"><i class="bi bi-envelope"></i></a> <a href="https://orcid.org/0000-0002-3514-7462" class="quarto-title-author-orcid"> <img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg=="></a></p>
  </div>
  <div class="quarto-title-meta-contents">
        <p class="affiliation">
            <a href="https://mcic.osu.edu/bioinformatics">
            CFAES Bioinformatics Core, Ohio State University
            </a>
          </p>
      </div>
  </div>

<div class="quarto-title-meta">

      
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">January 30, 2026</p>
    </div>
  </div>
  
    
  </div>
  


</header>


<hr>
<p><br></p>
<section id="introduction" class="level2" data-number="1">
<h2 data-number="1" class="anchored" data-anchor-id="introduction"><span class="header-section-number">1</span> Introduction</h2>
<p>This homework is intended to get you introduce you to the basics of R &amp; RStudio, because the week 4 lab will involve using R for RNA-Seq data analysis. As such, this is mostly intended for those with little to no experience with R, <em>but</em>:</p>
<ul>
<li>Even if you’ve worked with R before, at least glance over all material to make sure you’re familiar with all of it</li>
<li>Everyone should install several R packages in RStudio at the Ohio Supercomputer Center (OSC) – <a href="#r-packages">section 8</a></li>
</ul>
<div class="callout callout-style-simple callout-warning">
<div class="callout-body d-flex">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-body-container">
<p>If you run out of steam or time while trying to do this homework, you can stop after section 10 (R objects). But we will see examples of the topics covered in section 11-13 in the week 4 lab.</p>
</div>
</div>
</div>
<p><em>If you run into any issues when going through this material,</em> <em>or have questions, don’t hesitate to <a href="poelstra.1@osu.edu">email Jelmer</a>.</em></p>
<section id="learning-objectives" class="level3" data-number="1.1">
<h3 data-number="1.1" class="anchored" data-anchor-id="learning-objectives"><span class="header-section-number">1.1</span> Learning objectives</h3>
<p>In this session, you will learn:</p>
<ul>
<li>The RStudio layout and the functions of the different panes and tabs</li>
<li>The basics of interacting with R by using R as a calculator</li>
<li>Creating objects in R, and assigning values/data to them</li>
<li>Calling functions</li>
<li>Getting help</li>
<li>Installing and loading packages</li>
<li>Data structures like vectors and data frames</li>
<li>The basics of the tidyverse packages, including piping with <code>|&gt;</code></li>
</ul>
</section>
</section>
<section id="r-and-rstudio" class="level2" data-number="2">
<h2 data-number="2" class="anchored" data-anchor-id="r-and-rstudio"><span class="header-section-number">2</span> R and RStudio</h2>
<section id="why-use-r" class="level3" data-number="2.1">
<h3 data-number="2.1" class="anchored" data-anchor-id="why-use-r"><span class="header-section-number">2.1</span> Why use R?</h3>
<p>After initial processing high-throughput sequencing (HTS) data such as with command-line tools in the Unix shell, you’ll often want to perform <em>statistical analysis and visualization</em> of the results. The R language is probably the best environment for this.</p>
<p>On top of that, R is a great all-round tool to perform the analysis and visualization of any kind of data — together with Python, it is the leading data science language. Because it is free, open-source, cross-platform, and works with code rather than a graphical user interface (GUI), using R also helps with <em>reproducibility</em> of your research.</p>
</section>
<section id="r-vs.-rstudio" class="level3" data-number="2.2">
<h3 data-number="2.2" class="anchored" data-anchor-id="r-vs.-rstudio"><span class="header-section-number">2.2</span> R vs.&nbsp;RStudio</h3>
<p>To run R, all you need is the R program with its “console” where you type commands. However, we will be running R inside a fancy editor called <a href="https://www.rstudio.com/products/rstudio">RStudio</a>, which makes working with R much more pleasant!</p>
</section>
</section>
<section id="start-an-rstudio-session-at-osc" class="level2" data-number="3">
<h2 data-number="3" class="anchored" data-anchor-id="start-an-rstudio-session-at-osc"><span class="header-section-number">3</span> Start an RStudio session at OSC</h2>
<p>While you can install and use R and RStudio on your own computer, for this course, you will use RStudio at OSC: you can run a version of RStudio at OSC within your browser by starting an Interactive App through OnDemand.</p>
<ol type="1">
<li>Log in to OSC’s OnDemand portal at <a href="https://ondemand.osc.edu" class="uri">https://ondemand.osc.edu</a></li>
<li>In the blue top bar, select <strong><code>Interactive Apps</code></strong> and at the bottom, click <strong><code>RStudio Server</code></strong></li>
<li>Fill out the form as follows:
<ul>
<li>Cluster: <strong><code>cardinal</code></strong></li>
<li>R version: <strong><code>4.4.0</code></strong></li>
<li>Project: <strong><code>PAS2880</code></strong></li>
<li>Number of hours: <strong><code>3</code></strong></li>
<li>Node type: <strong><code>any</code></strong></li>
<li>Number of cores: <strong><code>1</code></strong></li>
</ul></li>
<li>Click <strong><code>Launch</code></strong></li>
<li>Click the <strong><code>Connect to RStudio Server</code></strong> button once it appears</li>
</ol>
<div class="callout callout-style-simple callout-tip">
<div class="callout-body d-flex">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-body-container">
<p>If you run out of time for this session, or want to spread this homework across multiple short sessions: you can always start a new session for as many sessions as you want, and you can modify the number of hours in the form as you see fit.</p>
</div>
</div>
</div>
</section>
<section id="orienting-to-rstudio" class="level2" data-number="4">
<h2 data-number="4" class="anchored" data-anchor-id="orienting-to-rstudio"><span class="header-section-number">4</span> Orienting to RStudio</h2>
<p>RStudio is divided into four panes, each of which have multiple tabs – we will highlight the most important ones here:</p>
<ul>
<li><p><strong>Top left: Source (or Editor)</strong><br>
Here, you can open and edit text documents like scripts <em>(May be missing if you don’t have any documents open – if so, we’ll make it appear later)</em></p></li>
<li><p><strong>Bottom left: Console</strong><br>
Here, you can type and execute R code</p></li>
<li><p><strong>Top right: Workspace</strong><br>
The default Environment tab shows you what you have loaded in your R environment</p></li>
<li><p><strong>Bottom right: View</strong><br>
The default tab in this pane is “Files”, a file explorer. Also commonly used are e.g.&nbsp;Plots, where plots appear, and Help, for help info.</p></li>
</ul>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="../img/R_studio_panes.png" class="img-fluid figure-img" style="width:90.0%"></p>
<figcaption>A screenshot showing the four main RStudio panes</figcaption>
</figure>
</div>
</section>
<section id="r-basics" class="level2" data-number="5">
<h2 data-number="5" class="anchored" data-anchor-id="r-basics"><span class="header-section-number">5</span> R basics</h2>
<section id="r-as-calculator" class="level3" data-number="5.1">
<h3 data-number="5.1" class="anchored" data-anchor-id="r-as-calculator"><span class="header-section-number">5.1</span> R as calculator</h3>
<p>To get some familiarity with working in R, you will start by simply using R as calculator. Type <code>5 + 5</code> in the console and press <kbd>Enter</kbd>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="dv">5</span> <span class="sc">+</span> <span class="dv">5</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="regular-r-output"><code>[1] 10</code></pre>
</div>
</div>
<p>R will print the result of the calculation, <code>10</code>. The result is prefaced by <code>[1]</code>, which is simply an index/count of output items – this can be helpful when many items are printed.</p>
<p>A couple more examples:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Multiplication</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="dv">6</span> <span class="sc">*</span> <span class="dv">8</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="regular-r-output"><code>[1] 48</code></pre>
</div>
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Division</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="dv">40</span> <span class="sc">/</span> <span class="dv">5</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="regular-r-output"><code>[1] 8</code></pre>
</div>
</div>
<div class="callout callout-style-simple callout-tip">
<div class="callout-body d-flex">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-body-container">
<p>This code will work with (like shown above) or without spaces between numbers and around operators like <code>+</code>. However, it is good practice to leave space between the number to improve the readability of your code.</p>
</div>
</div>
</div>
</section>
<section id="the-r-prompt" class="level3" data-number="5.2">
<h3 data-number="5.2" class="anchored" data-anchor-id="the-r-prompt"><span class="header-section-number">5.2</span> The R prompt</h3>
<p>The <code>&gt;</code> sign in your console is the R prompt, indicating that that R is ready for you to type something. In other words, this is equivalent to the <code>$</code> Shell prompt you’ve seen in the week 3 lab.</p>
<p>When you are not seeing the prompt, R is either <strong>busy</strong> (because you asked it to do a longer-running computation) or <strong>waiting</strong> for you to complete an incomplete command. If you notice that your prompt turned into a <code>+</code>, you typed an incomplete command – for example:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="dv">10</span> <span class="sc">/</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<pre class="r-out"><code>+</code></pre>
<p>You can either try to finish the command or abort it by pressing <kbd>Ctrl</kbd>+<kbd>C</kbd> or <kbd>Esc</kbd> – go ahead and practice the latter now to get your prompt back after the above command.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Adding comments to your code
</div>
</div>
<div class="callout-body-container callout-body">
<p>Comments in R are also just like in the Unix shell: you can use <code>#</code> signs to comment your code, both inline and on separate lines:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="co"># This line is entirely ignored by R</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a><span class="dv">10</span> <span class="sc">/</span> <span class="dv">5</span>  <span class="co"># This line contains code but everything after the '#' is ignored</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="regular-r-output"><code>[1] 2</code></pre>
</div>
</div>
</div>
</div>
</section>
<section id="comparing-things" class="level3" data-number="5.3">
<h3 data-number="5.3" class="anchored" data-anchor-id="comparing-things"><span class="header-section-number">5.3</span> Comparing things</h3>
<p>In R, you can use comparison operators to compare numbers. When you do this, R will return either <code>FALSE</code> or <code>TRUE</code> (the so-called <code>logical</code> data type). For example:</p>
<ul>
<li><p>Greater than and less than:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="dv">8</span> <span class="sc">&gt;</span> <span class="dv">3</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="regular-r-output"><code>[1] TRUE</code></pre>
</div>
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="dv">2</span> <span class="sc">&lt;</span> <span class="dv">1</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="regular-r-output"><code>[1] FALSE</code></pre>
</div>
</div></li>
<li><p>Equal to or not equal to – note that you need to use <strong>two</strong> equals signs <code>==</code> for the former:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="dv">7</span> <span class="sc">==</span> <span class="dv">7</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="regular-r-output"><code>[1] TRUE</code></pre>
</div>
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="co"># The exclamation mark in general is a "negator"</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a><span class="dv">10</span> <span class="sc">!=</span> <span class="dv">5</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="regular-r-output"><code>[1] TRUE</code></pre>
</div>
</div></li>
</ul>
<p>You will commonly use these types of comparisons to check and subset/filter your data.</p>
</section>
</section>
<section id="functions-in-r" class="level2" data-number="6">
<h2 data-number="6" class="anchored" data-anchor-id="functions-in-r"><span class="header-section-number">6</span> Functions in R</h2>
<p>To do almost anything in R, you will use “<strong>functions</strong>”, which are the equivalent of Unix commands. R has over 3,000 functions that serve many different purposes.</p>
<section id="basics-of-functions" class="level3" data-number="6.1">
<h3 data-number="6.1" class="anchored" data-anchor-id="basics-of-functions"><span class="header-section-number">6.1</span> Basics of functions</h3>
<p>You can recognize function calls by the parentheses. For example:</p>
<ul>
<li><p>The equivalent to the Shell’s <code>pwd</code> is <code>getwd()</code> for “get working dir”:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="co"># You will see a different output depending on your working directory</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a><span class="fu">getwd</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<pre class="r-out"><code>/users/PAS2880/jelmer</code></pre></li>
<li><p>Inside those parentheses, you can provide <strong>arguments</strong> to a function<a href="#fn1" class="footnote-ref" id="fnref1" role="doc-noteref"><sup>1</sup></a>. For example, using <code>setwd()</code>, the equivalent to the Shell’s <code>cd</code>, to set your working dir, the argument is the file path to set as working dir:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="fu">setwd</span>(<span class="st">"/fs/scratch/PAS2880/ENT6703"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="callout callout-style-simple callout-warning">
<div class="callout-body d-flex">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-body-container">
<p>Note that unlike in the Shell, you need to <em>quote (<code>".."</code>) the file path</em>! This is because in R, anything non-numeric that is not quoted is supposed to be a pre-existing “object” – more on those in a bit.</p>
</div>
</div>
</div></li>
</ul>
</section>
<section id="function-arguments" class="level3" data-number="6.2">
<h3 data-number="6.2" class="anchored" data-anchor-id="function-arguments"><span class="header-section-number">6.2</span> Function arguments</h3>
<p>Consider the example below, where you pass the argument <code>6.53</code> to the function <code>round()</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="fu">round</span>(<span class="fl">6.53</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="regular-r-output"><code>[1] 7</code></pre>
</div>
</div>
<p>In most functions, including <code>round()</code>, arguments have <strong>names</strong> that can be used. The argument to <code>round()</code>, which represents the number to be rounded, has the name <code>x</code>, so you can also use:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="co"># This is equivalent to the above:</span></span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a><span class="fu">round</span>(<span class="at">x =</span> <span class="fl">6.53</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="regular-r-output"><code>[1] 7</code></pre>
</div>
</div>
<p>Naming arguments is helpful in particular to avoid confusion when you pass multiple arguments to a function. For example, <code>round()</code> accepts a second argument, which is the number of digits to round to:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="fu">round</span>(<span class="at">x =</span> <span class="fl">6.53</span>, <span class="at">digits =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="regular-r-output"><code>[1] 6.5</code></pre>
</div>
</div>
<p>You could choose not to name these arguments, but that is not as easy to understand for someone reading your code – such as yourself two months later! <a href="#fn2" class="footnote-ref" id="fnref2" role="doc-noteref"><sup>2</sup></a></p>
<div class="callout callout-style-simple callout-note">
<div class="callout-body d-flex">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-body-container">
<p>You can usually see which arguments a function takes by pausing after your type the function and its parenthesis, e.g.&nbsp;<code>round()</code>:</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="../img/round_1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:75.0%"></p>
</figure>
</div>
<p>Below, you’ll also learn about ways to get more extensive help information about a function.</p>
</div>
</div>
</div>
</section>
</section>
<section id="r-help-help-and" class="level2" data-number="7">
<h2 data-number="7" class="anchored" data-anchor-id="r-help-help-and"><span class="header-section-number">7</span> R Help: <code>help()</code> and <code>?</code></h2>
<p>So how can you know what a function does, and which arguments it takes? The <code>help()</code> function and <code>?</code> help operator in R offer access to documentation pages for R functions – as well as built-in data sets and other objects. For example:</p>
<div class="sourceCode" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>?getwd</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>The output should appear in the Help tab of the bottom-left panel in RStudio and look something like this:</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="../img/R_help_setwd.png" class="img-fluid figure-img" style="width:65.0%"></p>
<figcaption>The R help output for the <code>getwd()</code> function</figcaption>
</figure>
</div>
<p>You can also search for Help pages in that same RStudio Help tab.</p>
</section>
<section id="r-packages" class="level2" data-number="8">
<h2 data-number="8" class="anchored" data-anchor-id="r-packages"><span class="header-section-number">8</span> R packages</h2>
<p>R packages are basically add-ons or extensions to the functionality of the R language. These need to installed and loaded:</p>
<ul>
<li><p>The function <code>install.packages()</code> <strong>installs</strong> an R package. This is a one-time operation – you won’t need to reinstall a package when you start a new R session later<a href="#fn3" class="footnote-ref" id="fnref3" role="doc-noteref"><sup>3</sup></a>.</p></li>
<li><p>The function <code>library()</code> will <strong>load</strong> an R package in your current R session. Every time you want to use a package in R, you need to load it – just like everytime you want to use MS Word on your computer, you need to open it.</p></li>
</ul>
<p>To practice with this, you will now install several packages you’ll need in next week’s lab. Note that:</p>
<ul>
<li><p>Unfortunately, on OSC, packages installation can take a while. Especially for large packages like these, so please be patient<a href="#fn4" class="footnote-ref" id="fnref4" role="doc-noteref"><sup>4</sup></a>.</p></li>
<li><p>A lot of messages may be printed to the console when installation is ongoing. Installation of a package is done when you get your prompt back. Unless you see clear messages that include words like “Error” and “Failed”, it should have succeeded (“Warning”s are typically okay).</p></li>
<li><p>Besides <code>install.packages()</code>, you’ll also use the <code>BiocManager::install()</code> function from to install some packages from Bioconductor, a separate repository that hosts many bioinformatics/omics-related R packages.</p></li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Run these lines one-by-one to install the needed packages --</span></span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a><span class="co"># First, the tidyverse, which is discussed at the end of this homework</span></span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">"tidyverse"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<pre class="r-out"><code># [... lots of output ...]
# [The last lines should look something like this:]
** testing if installed package can be loaded from temporary location
** testing if installed package can be loaded from final location
** testing if installed package keeps a record of temporary installation path
* DONE (tidyverse)

The downloaded source packages are in
    ‘/tmp/RtmpS4n222/downloaded_packages’</code></pre>
<div class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="co"># This is needed to in turn install the next two packages:</span></span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a><span class="co"># (Should take &lt;1 minute to install)</span></span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">"BiocManager"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="co"># We will use this for differential expression analysis:</span></span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a><span class="co"># (May take 5-10 minutes to install)</span></span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">"DESeq2"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="co"># We will use this to create a data visualization called a volcano plot:</span></span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a><span class="co"># (Should take &lt;1 minute to install)</span></span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">"EnhancedVolcano"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Test that you can load the packages – see the expected output below:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(EnhancedVolcano)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre class="regular-r-output"><code>Loading required package: ggplot2</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre class="regular-r-output"><code>Loading required package: ggrepel</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(DESeq2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<details>
<summary>
<em>Click here to see the expected output when loading the ‘DESeq2’ package</em>
</summary>
<div class="cell">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(DESeq2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre class="regular-r-output"><code>Loading required package: S4Vectors</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre class="regular-r-output"><code>Loading required package: stats4</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre class="regular-r-output"><code>Loading required package: BiocGenerics</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre class="regular-r-output"><code>Loading required package: generics</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre class="regular-r-output"><code>
Attaching package: 'generics'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre class="regular-r-output"><code>The following objects are masked from 'package:base':

    as.difftime, as.factor, as.ordered, intersect, is.element, setdiff,
    setequal, union</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre class="regular-r-output"><code>
Attaching package: 'BiocGenerics'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre class="regular-r-output"><code>The following objects are masked from 'package:stats':

    IQR, mad, sd, var, xtabs</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre class="regular-r-output"><code>The following objects are masked from 'package:base':

    anyDuplicated, aperm, append, as.data.frame, basename, cbind,
    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,
    get, grep, grepl, is.unsorted, lapply, Map, mapply, match, mget,
    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,
    rbind, Reduce, rownames, sapply, saveRDS, table, tapply, unique,
    unsplit, which.max, which.min</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre class="regular-r-output"><code>
Attaching package: 'S4Vectors'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre class="regular-r-output"><code>The following object is masked from 'package:utils':

    findMatches</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre class="regular-r-output"><code>The following objects are masked from 'package:base':

    expand.grid, I, unname</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre class="regular-r-output"><code>Loading required package: IRanges</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre class="regular-r-output"><code>Loading required package: GenomicRanges</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre class="regular-r-output"><code>Loading required package: GenomeInfoDb</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre class="regular-r-output"><code>Loading required package: SummarizedExperiment</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre class="regular-r-output"><code>Loading required package: MatrixGenerics</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre class="regular-r-output"><code>Loading required package: matrixStats</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre class="regular-r-output"><code>
Attaching package: 'MatrixGenerics'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre class="regular-r-output"><code>The following objects are masked from 'package:matrixStats':

    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,
    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,
    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,
    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,
    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,
    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,
    colWeightedMeans, colWeightedMedians, colWeightedSds,
    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,
    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,
    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,
    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,
    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,
    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,
    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,
    rowWeightedSds, rowWeightedVars</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre class="regular-r-output"><code>Loading required package: Biobase</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre class="regular-r-output"><code>Welcome to Bioconductor

    Vignettes contain introductory material; view with
    'browseVignettes()'. To cite Bioconductor, see
    'citation("Biobase")', and for packages 'citation("pkgname")'.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre class="regular-r-output"><code>
Attaching package: 'Biobase'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre class="regular-r-output"><code>The following object is masked from 'package:MatrixGenerics':

    rowMedians</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre class="regular-r-output"><code>The following objects are masked from 'package:matrixStats':

    anyMissing, rowMedians</code></pre>
</div>
</div>
</details>
<div class="cell">
<div class="sourceCode cell-code" id="cb64"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre class="regular-r-output"><code>── Attaching core tidyverse packages ──────────────────────── tidyverse 2.0.0 ──
✔ dplyr     1.1.4     ✔ readr     2.1.5
✔ forcats   1.0.1     ✔ stringr   1.5.2
✔ lubridate 1.9.4     ✔ tibble    3.3.0
✔ purrr     1.1.0     ✔ tidyr     1.3.1
── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
✖ lubridate::%within%()    masks IRanges::%within%()
✖ dplyr::collapse()        masks IRanges::collapse()
✖ dplyr::combine()         masks Biobase::combine(), BiocGenerics::combine()
✖ dplyr::count()           masks matrixStats::count()
✖ dplyr::desc()            masks IRanges::desc()
✖ tidyr::expand()          masks S4Vectors::expand()
✖ dplyr::filter()          masks stats::filter()
✖ dplyr::first()           masks S4Vectors::first()
✖ dplyr::lag()             masks stats::lag()
✖ BiocGenerics::Position() masks ggplot2::Position(), base::Position()
✖ purrr::reduce()          masks GenomicRanges::reduce(), IRanges::reduce()
✖ dplyr::rename()          masks S4Vectors::rename()
✖ lubridate::second()      masks S4Vectors::second()
✖ lubridate::second&lt;-()    masks S4Vectors::second&lt;-()
✖ dplyr::slice()           masks IRanges::slice()
ℹ Use the conflicted package (&lt;http://conflicted.r-lib.org/&gt;) to force all conflicts to become errors</code></pre>
</div>
</div>
<div class="callout callout-style-simple callout-note">
<div class="callout-body d-flex">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-body-container">
<p><em>For now, don’t worry about the “conflict” warnings shown above.</em></p>
</div>
</div>
</div>
</section>
<section id="r-scripts" class="level2" data-number="9">
<h2 data-number="9" class="anchored" data-anchor-id="r-scripts"><span class="header-section-number">9</span> R scripts</h2>
<p>You should mostly write and save your R code in so-called “R scripts”, which are text files with R code. This helps keep track of what you’ve been doing and makes it easier to modify and rerun things.</p>
<ol type="1">
<li><p>Create and open a new R script by clicking <strong><code>File</code></strong> (top menu bar) &gt; <strong><code>New File</code></strong> &gt; <strong><code>R Script</code></strong>.</p></li>
<li><p>Save this new script right away by clicking <strong><code>File</code></strong> &gt; <strong><code>Save As</code></strong>. In the dialog box, you can save it right in the suggested folder, or you can make a new folder for this course. Regardless, save the script as <strong><code>homework.R</code></strong> (the extension for R scripts is <code>.R</code>).</p></li>
</ol>
<p>But isn’t a hassle to have to copy-paste code from a script to the console? Well, RStudio makes this pretty easy: to execute code from the script in the the console, press <kbd>Ctrl</kbd> + <kbd>Enter</kbd> on Windows, or <kbd>Cmd</kbd> + <kbd>Return</kbd> on a Mac. Practice that after typing the following in the script:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb66"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb66-1"><a href="#cb66-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Type this in the script, then run it with Ctrl + Enter (or Cmd + Return)</span></span>
<span id="cb66-2"><a href="#cb66-2" aria-hidden="true" tabindex="-1"></a><span class="dv">5</span> <span class="sc">+</span> <span class="dv">5</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="regular-r-output"><code>[1] 10</code></pre>
</div>
</div>
</section>
<section id="r-objects" class="level2" data-number="10">
<h2 data-number="10" class="anchored" data-anchor-id="r-objects"><span class="header-section-number">10</span> R objects</h2>
<p>Objects are a key concept in R and other programming languages. They can store data and values that you can manipulate and analyze. It takes some time to get used to working with -and conceptualizing- objects, but you will get the hang of it with some practice.</p>
<section id="assigning-stuff-to-objects" class="level3" data-number="10.1">
<h3 data-number="10.1" class="anchored" data-anchor-id="assigning-stuff-to-objects"><span class="header-section-number">10.1</span> Assigning stuff to objects</h3>
<p>You can assign one or more values to a so-called “object” with the <strong>assignment operator</strong> <code>&lt;-</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb68"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb68-1"><a href="#cb68-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Assign the value 250 the object 'length_cm':</span></span>
<span id="cb68-2"><a href="#cb68-2" aria-hidden="true" tabindex="-1"></a>length_cm <span class="ot">&lt;-</span> <span class="dv">250</span></span>
<span id="cb68-3"><a href="#cb68-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-4"><a href="#cb68-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Now print the value of this object to screen simply by typing its name:</span></span>
<span id="cb68-5"><a href="#cb68-5" aria-hidden="true" tabindex="-1"></a>length_cm</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="regular-r-output"><code>[1] 250</code></pre>
</div>
</div>
<p>Another example:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb70"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb70-1"><a href="#cb70-1" aria-hidden="true" tabindex="-1"></a>conversion <span class="ot">&lt;-</span> <span class="fl">2.54</span></span>
<span id="cb70-2"><a href="#cb70-2" aria-hidden="true" tabindex="-1"></a>conversion</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="regular-r-output"><code>[1] 2.54</code></pre>
</div>
</div>
<p>Importantly, you can use objects as if you had typed their values directly:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb72"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb72-1"><a href="#cb72-1" aria-hidden="true" tabindex="-1"></a>length_cm <span class="sc">/</span> conversion</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="regular-r-output"><code>[1] 98.4252</code></pre>
</div>
</div>
</section>
<section id="object-names" class="level3" data-number="10.2">
<h3 data-number="10.2" class="anchored" data-anchor-id="object-names"><span class="header-section-number">10.2</span> Object names</h3>
<p>Some pointers on object names:</p>
<ul>
<li>Because R is case sensitive, <code>length_inch</code> is different from <code>Length_Inch</code>!</li>
<li>An object name cannot contain spaces — so for readability, you can for example separate words e.g.&nbsp;using underscores, as in <code>length_cm</code>.</li>
<li>Object names can contain but cannot start with a number: <code>x2</code> is valid but <code>2x</code> is not.</li>
<li><em>Make object names descriptive yet not too long — this is not always easy!</em></li>
</ul>
<section id="exercise-assigning-objects-in-r" class="level4 exercise">
<h4 class="anchored" data-anchor-id="exercise-assigning-objects-in-r"><i class="fa-solid fa-user-edit" aria-label="user-edit"></i> Exercise: Assigning objects in R</h4>
<ol type="1">
<li><p>Assign the value 20 to the object <code>num_1</code>, and the value 15 to <code>num_2</code></p>
<details>
<summary>
<p><em>Click to see the solution</em></p>
</summary>
<div class="cell">
<div class="sourceCode cell-code" id="cb74"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb74-1"><a href="#cb74-1" aria-hidden="true" tabindex="-1"></a>num_1 <span class="ot">&lt;-</span> <span class="dv">20</span></span>
<span id="cb74-2"><a href="#cb74-2" aria-hidden="true" tabindex="-1"></a>num_2 <span class="ot">&lt;-</span> <span class="dv">15</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</details></li>
<li><p>Compute the sum of <code>num_1</code> and <code>num_2</code></p>
<details>
<summary>
<p><em>Click to see the solution</em></p>
</summary>
<div class="cell">
<div class="sourceCode cell-code" id="cb75"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb75-1"><a href="#cb75-1" aria-hidden="true" tabindex="-1"></a>num_1 <span class="sc">+</span> num_2</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="regular-r-output"><code>[1] 35</code></pre>
</div>
</div>
</details></li>
<li><p>Assign the result of <code>num_1</code> minus <code>num_2</code> to a new object called <code>difference</code>. What is the value of <code>difference</code>?</p>
<details>
<summary>
<p><em>Click to see a solution</em></p>
</summary>
<div class="cell">
<div class="sourceCode cell-code" id="cb77"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb77-1"><a href="#cb77-1" aria-hidden="true" tabindex="-1"></a>difference <span class="ot">&lt;-</span> num_1 <span class="sc">-</span> num_2</span>
<span id="cb77-2"><a href="#cb77-2" aria-hidden="true" tabindex="-1"></a>difference</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="regular-r-output"><code>[1] 5</code></pre>
</div>
</div>
</details></li>
<li><p>Which of the following are valid object names in R?</p>
<ol type="a">
<li><code>2nd_value</code><br>
</li>
<li><code>value_2</code><br>
</li>
<li><code>total value</code></li>
<li><code>TotalValue</code></li>
</ol>
<details>
<summary>
<p><em>Click to see a solution</em></p>
</summary>
<p>The valid object names are:</p>
<ol start="2" type="a">
<li><code>value_2</code></li>
<li><code>TotalValue</code></li>
</ol>
<p>Object names cannot start with a number and cannot contain spaces (so a and c options are invalid).</p></details></li>
</ol>

</section>
</section>
</section>
<section id="vectors" class="level2" data-number="11">
<h2 data-number="11" class="anchored" data-anchor-id="vectors"><span class="header-section-number">11</span> Vectors</h2>
<p>R has several so-called “data structures”, distinct object types to store different kinds of data. The most basic of these is the <strong>vector</strong>, which is simply a <strong>collections of one or more items</strong>.</p>
<section id="creating-vectors" class="level3" data-number="11.1">
<h3 data-number="11.1" class="anchored" data-anchor-id="creating-vectors"><span class="header-section-number">11.1</span> Creating vectors</h3>
<p>You’ve already created a few vectors, as the single-item <code>num_1</code> etc. objects above were vectors! So let’s see something new here: a common way to make vectors with <strong>multiple elements</strong> is with the <code>c</code> (combine) function:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb79"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb79-1"><a href="#cb79-1" aria-hidden="true" tabindex="-1"></a><span class="co"># longer_vector will contain 3 items:</span></span>
<span id="cb79-2"><a href="#cb79-2" aria-hidden="true" tabindex="-1"></a>longer_vector <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">10</span>, <span class="dv">20</span>, <span class="dv">30</span>)</span>
<span id="cb79-3"><a href="#cb79-3" aria-hidden="true" tabindex="-1"></a>longer_vector</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="regular-r-output"><code>[1] 10 20 30</code></pre>
</div>
</div>
<p>For ranges of integers (whole numbers), you can also use the colon operator <code>:</code></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb81"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb81-1"><a href="#cb81-1" aria-hidden="true" tabindex="-1"></a><span class="co"># longer_vector will contain 10 items:</span></span>
<span id="cb81-2"><a href="#cb81-2" aria-hidden="true" tabindex="-1"></a>sequence_vector <span class="ot">&lt;-</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">10</span></span>
<span id="cb81-3"><a href="#cb81-3" aria-hidden="true" tabindex="-1"></a>sequence_vector</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="regular-r-output"><code> [1]  1  2  3  4  5  6  7  8  9 10</code></pre>
</div>
</div>
</section>
<section id="exploring-vectors" class="level3" data-number="11.2">
<h3 data-number="11.2" class="anchored" data-anchor-id="exploring-vectors"><span class="header-section-number">11.2</span> Exploring vectors</h3>
<p>R has many functions that provide information about vectors and other types of objects, such as to get the number of items, or the mean or sum of the values of these items<a href="#fn5" class="footnote-ref" id="fnref5" role="doc-noteref"><sup>5</sup></a>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb83"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb83-1"><a href="#cb83-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Get the number of items:</span></span>
<span id="cb83-2"><a href="#cb83-2" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(longer_vector)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="regular-r-output"><code>[1] 3</code></pre>
</div>
<div class="sourceCode cell-code" id="cb85"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb85-1"><a href="#cb85-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Get the the mean (average) across all elements:</span></span>
<span id="cb85-2"><a href="#cb85-2" aria-hidden="true" tabindex="-1"></a><span class="fu">mean</span>(longer_vector)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="regular-r-output"><code>[1] 20</code></pre>
</div>
<div class="sourceCode cell-code" id="cb87"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb87-1"><a href="#cb87-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Get the sum of all elements:</span></span>
<span id="cb87-2"><a href="#cb87-2" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(longer_vector)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="regular-r-output"><code>[1] 60</code></pre>
</div>
</div>
</section>
<section id="extracting-elements-from-vectors" class="level3" data-number="11.3">
<h3 data-number="11.3" class="anchored" data-anchor-id="extracting-elements-from-vectors"><span class="header-section-number">11.3</span> Extracting elements from vectors</h3>
<p>You can extract items from objects like vectors using “bracket notation” – get the second item with <code>[2]</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb89"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb89-1"><a href="#cb89-1" aria-hidden="true" tabindex="-1"></a>longer_vector[<span class="dv">2</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="regular-r-output"><code>[1] 20</code></pre>
</div>
</div>
<p>Or get the first and third items with <code>[c(1, 3)]</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb91"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb91-1"><a href="#cb91-1" aria-hidden="true" tabindex="-1"></a>longer_vector[<span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">3</span>)]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="regular-r-output"><code>[1] 10 30</code></pre>
</div>
</div>
</section>
</section>
<section id="data-frames" class="level2" data-number="12">
<h2 data-number="12" class="anchored" data-anchor-id="data-frames"><span class="header-section-number">12</span> Data frames</h2>
<p><strong>Data frames</strong> are a key R data structure that contain <strong>tabular data</strong> – data organized in rows and columns, similar to spreadsheets. An example is the <code>iris</code> data frame that comes shipped with R, which contains measurements for three different species of iris.</p>
<p>You can take a look at the first few rows of this data frame using the <code>head()</code> function:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb93"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb93-1"><a href="#cb93-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Note that each row represents one individual iris flower:</span></span>
<span id="cb93-2"><a href="#cb93-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(iris)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="regular-r-output"><code>  Sepal.Length Sepal.Width Petal.Length Petal.Width Species
1          5.1         3.5          1.4         0.2  setosa
2          4.9         3.0          1.4         0.2  setosa
3          4.7         3.2          1.3         0.2  setosa
4          4.6         3.1          1.5         0.2  setosa
5          5.0         3.6          1.4         0.2  setosa
6          5.4         3.9          1.7         0.4  setosa</code></pre>
</div>
</div>
<details>
<summary>
How many columns does this data frame have?<em>(Click for the solution)</em>
</summary>
This data frame has 5 columns: <code>Sepal.Length</code>, <code>Sepal.Width</code>, <code>Petal.Length</code>, <code>Petal.Width</code>, and <code>Species</code>. The 1 to 6 on the left are the <strong>row names</strong>, in this case a simple numbering of the rows.
</details>
<p>For data frames, the <code>View()</code> function is also neat, as it opens the data frame in a spreadsheet-like viewer – try it!</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb95"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb95-1"><a href="#cb95-1" aria-hidden="true" tabindex="-1"></a><span class="fu">View</span>(iris)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>You can extract individual columns from a data frame using the <code>$</code> operator:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb96"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb96-1"><a href="#cb96-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Extract the 'Sepal.Width' column from the 'iris' data frame:</span></span>
<span id="cb96-2"><a href="#cb96-2" aria-hidden="true" tabindex="-1"></a>iris<span class="sc">$</span>Sepal.Width</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="regular-r-output"><code>  [1] 3.5 3.0 3.2 3.1 3.6 3.9 3.4 3.4 2.9 3.1 3.7 3.4 3.0 3.0 4.0 4.4 3.9 3.5
 [19] 3.8 3.8 3.4 3.7 3.6 3.3 3.4 3.0 3.4 3.5 3.4 3.2 3.1 3.4 4.1 4.2 3.1 3.2
 [37] 3.5 3.6 3.0 3.4 3.5 2.3 3.2 3.5 3.8 3.0 3.8 3.2 3.7 3.3 3.2 3.2 3.1 2.3
 [55] 2.8 2.8 3.3 2.4 2.9 2.7 2.0 3.0 2.2 2.9 2.9 3.1 3.0 2.7 2.2 2.5 3.2 2.8
 [73] 2.5 2.8 2.9 3.0 2.8 3.0 2.9 2.6 2.4 2.4 2.7 2.7 3.0 3.4 3.1 2.3 3.0 2.5
 [91] 2.6 3.0 2.6 2.3 2.7 3.0 2.9 2.9 2.5 2.8 3.3 2.7 3.0 2.9 3.0 3.0 2.5 2.9
[109] 2.5 3.6 3.2 2.7 3.0 2.5 2.8 3.2 3.0 3.8 2.6 2.2 3.2 2.8 2.8 2.7 3.3 3.2
[127] 2.8 3.0 2.8 3.0 2.8 3.8 2.8 2.8 2.6 3.0 3.4 3.1 3.0 3.1 3.1 3.1 2.7 3.2
[145] 3.3 3.0 2.5 3.0 3.4 3.0</code></pre>
</div>
</div>
<p>This kind of operation will return a vector – which can be subsetted as well:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb98"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb98-1"><a href="#cb98-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Extract the 2nd to 5th items from the 'Sepal.Width' column:</span></span>
<span id="cb98-2"><a href="#cb98-2" aria-hidden="true" tabindex="-1"></a>iris<span class="sc">$</span>Sepal.Width[<span class="dv">2</span><span class="sc">:</span><span class="dv">5</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="regular-r-output"><code>[1] 3.0 3.2 3.1 3.6</code></pre>
</div>
</div>
</section>
<section id="a-brief-introduction-to-the-tidyverse-and-pipes" class="level2" data-number="13">
<h2 data-number="13" class="anchored" data-anchor-id="a-brief-introduction-to-the-tidyverse-and-pipes"><span class="header-section-number">13</span> A brief introduction to the <em>tidyverse</em> and pipes</h2>
<p>Finally, you’ll learn a little about a versatile family of R packages designed for data science called the <a href="https://www.tidyverse.org">“tidyverse”</a>. The <em>tidyverse</em> is quite large and includes the famous <em>ggplot2</em> package for data visualization – we’ll barely scratch the surface here, but this should already give you some useful tools.</p>
<section id="read_tsv-to-read-tabular-data" class="level3" data-number="13.1">
<h3 data-number="13.1" class="anchored" data-anchor-id="read_tsv-to-read-tabular-data"><span class="header-section-number">13.1</span> <code>read_tsv()</code> to read tabular data</h3>
<p>The <code>read_tsv()</code> function reads a TSV (Tab-Separated Values) file into a dataframe, taking the path to the file as its main argument – here, read the sample metadata file used in next week’s lab:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb100"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb100-1"><a href="#cb100-1" aria-hidden="true" tabindex="-1"></a>meta <span class="ot">&lt;-</span> <span class="fu">read_tsv</span>(<span class="st">"/fs/scratch/PAS2880/ENT6703/data/meta/metadata.tsv"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="cell-output cell-output-stderr">
<pre class="regular-r-output"><code>Rows: 22 Columns: 3
── Column specification ────────────────────────────────────────────────────────
Delimiter: "\t"
chr (2): sample_id, treatment
dbl (1): dpi

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.</code></pre>
</div>
</div>
<p>Some info about how it interpreted the data is printed, especially about the data types of different columns (see the box below). Take a look at the resulting data frame <code>meta</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb102"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb102-1"><a href="#cb102-1" aria-hidden="true" tabindex="-1"></a>meta</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="regular-r-output"><code># A tibble: 22 × 3
   sample_id     dpi treatment  
   &lt;chr&gt;       &lt;dbl&gt; &lt;chr&gt;      
 1 ERR10802864     1 cathemerium
 2 ERR10802867     1 cathemerium
 3 ERR10802870     1 cathemerium
 4 ERR10802866     1 control    
 5 ERR10802869     1 control    
 6 ERR10802863     1 control    
 7 ERR10802871     1 relictum   
 8 ERR10802874     1 relictum   
 9 ERR10802865     1 relictum   
10 ERR10802868     1 relictum   
# ℹ 12 more rows</code></pre>
</div>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Tibble??
</div>
</div>
<div class="callout-body-container callout-body">
<p>The data frame is a so-called “tibble”, as the top of the output above says. This is a modern take on data frames provided by the tidyverse. Tibbles behave mostly like data frames, but have some nice features mostly in the way they are printed to screen.</p>
<p>For example, the “data types” of columns are shown at the top of the output above, where <code>chr</code> means character (text data) and <code>dbl</code> means double (numeric data; other data types exist as well).</p>
</div>
</div>
</section>
<section id="select-to-pick-columns" class="level3" data-number="13.2">
<h3 data-number="13.2" class="anchored" data-anchor-id="select-to-pick-columns"><span class="header-section-number">13.2</span> <code>select()</code> to pick columns</h3>
<p>The <code>select()</code> function subsets a data frame by including/excluding certain columns. By default, it only includes the columns you specify:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb104"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb104-1"><a href="#cb104-1" aria-hidden="true" tabindex="-1"></a><span class="fu">select</span>(<span class="at">.data =</span> meta, sample_id, treatment)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="regular-r-output"><code># A tibble: 22 × 2
   sample_id   treatment  
   &lt;chr&gt;       &lt;chr&gt;      
 1 ERR10802864 cathemerium
 2 ERR10802867 cathemerium
 3 ERR10802870 cathemerium
 4 ERR10802866 control    
 5 ERR10802869 control    
 6 ERR10802863 control    
 7 ERR10802871 relictum   
 8 ERR10802874 relictum   
 9 ERR10802865 relictum   
10 ERR10802868 relictum   
# ℹ 12 more rows</code></pre>
</div>
</div>
<p>Above, the first argument was the data frame, whereas the other arguments were the (unquoted!) names of columns to be kept. The output is a new, modified dataframe. In the example above, this data frame was simply printed to the screen – but as you’ll see later, it can be stored too.</p>
</section>
<section id="filter-to-pick-rows-observations" class="level3" data-number="13.3">
<h3 data-number="13.3" class="anchored" data-anchor-id="filter-to-pick-rows-observations"><span class="header-section-number">13.3</span> <code>filter()</code> to pick rows (observations)</h3>
<p>The <code>filter()</code> function outputs only those rows that satisfy one or more conditions, similar to Filter functionality in Excel. For example, to only keep rows for the <code>1</code> timepoint:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb106"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb106-1"><a href="#cb106-1" aria-hidden="true" tabindex="-1"></a><span class="fu">filter</span>(<span class="at">.data =</span> meta, dpi <span class="sc">==</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="regular-r-output"><code># A tibble: 10 × 3
   sample_id     dpi treatment  
   &lt;chr&gt;       &lt;dbl&gt; &lt;chr&gt;      
 1 ERR10802864     1 cathemerium
 2 ERR10802867     1 cathemerium
 3 ERR10802870     1 cathemerium
 4 ERR10802866     1 control    
 5 ERR10802869     1 control    
 6 ERR10802863     1 control    
 7 ERR10802871     1 relictum   
 8 ERR10802874     1 relictum   
 9 ERR10802865     1 relictum   
10 ERR10802868     1 relictum   </code></pre>
</div>
</div>
</section>
<section id="the-pipe" class="level3" data-number="13.4">
<h3 data-number="13.4" class="anchored" data-anchor-id="the-pipe"><span class="header-section-number">13.4</span> The pipe (<code>|&gt;</code>)</h3>
<p>It’s common to want to use several consecutive functions to wrangle a dataframe into the format you want. For example, you may want to first <code>select()</code> one or more columns, then <code>filter()</code> the resulting dataframe. You could do that as follows:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb108"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb108-1"><a href="#cb108-1" aria-hidden="true" tabindex="-1"></a>meta_sel <span class="ot">&lt;-</span> <span class="fu">select</span>(<span class="at">.data =</span> meta, sample_id, dpi)</span>
<span id="cb108-2"><a href="#cb108-2" aria-hidden="true" tabindex="-1"></a><span class="fu">filter</span>(<span class="at">.data =</span> meta_sel, dpi <span class="sc">==</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="regular-r-output"><code># A tibble: 10 × 2
   sample_id     dpi
   &lt;chr&gt;       &lt;dbl&gt;
 1 ERR10802864     1
 2 ERR10802867     1
 3 ERR10802870     1
 4 ERR10802866     1
 5 ERR10802869     1
 6 ERR10802863     1
 7 ERR10802871     1
 8 ERR10802874     1
 9 ERR10802865     1
10 ERR10802868     1</code></pre>
</div>
</div>
<p>You could go on along these lines, successively creating new objects that you then use for the next step. But there is a more elegant way of dong this, directly sending (“piping”) output from one function into the next function with the <strong>pipe operator</strong> <code>|&gt;</code> (a vertical bar <kbd>|</kbd> followed by a greater-than sign <kbd>&gt;</kbd>).</p>
<p>Let’s start by seeing a reformulation of the code above with pipes:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb110"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb110-1"><a href="#cb110-1" aria-hidden="true" tabindex="-1"></a>meta <span class="sc">|&gt;</span></span>
<span id="cb110-2"><a href="#cb110-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(sample_id, dpi) <span class="sc">|&gt;</span></span>
<span id="cb110-3"><a href="#cb110-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(dpi <span class="sc">==</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="regular-r-output"><code># A tibble: 10 × 2
   sample_id     dpi
   &lt;chr&gt;       &lt;dbl&gt;
 1 ERR10802864     1
 2 ERR10802867     1
 3 ERR10802870     1
 4 ERR10802866     1
 5 ERR10802869     1
 6 ERR10802863     1
 7 ERR10802871     1
 8 ERR10802874     1
 9 ERR10802865     1
10 ERR10802868     1</code></pre>
</div>
</div>
<p>What happened here? You took the <code>meta</code> data frame, sent (“piped”) it into the <code>select()</code> function, whose output in turn was piped into the <code>filter()</code> function. You can think of the pipe as <strong>“then”</strong>: take <code>meta</code>, <em>then</em> select, <em>then</em> rename.</p>
<p>When using the pipe, you no longer specify the input with the <code>.data</code> argument, because the function now gets its input data via the pipe. Using pipes involves less typing and is more readable than successive assignments.</p>
<div class="callout callout-style-simple callout-tip">
<div class="callout-body d-flex">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-body-container">
<p>For code readability, start a new line after every pipe <code>|&gt;</code> and indent subsequent lines, as seen above.</p>
</div>
</div>
</div>
<section id="exercise-tidyverse-functions-and-pipes" class="level4 exercise">
<h4 class="anchored" data-anchor-id="exercise-tidyverse-functions-and-pipes"><i class="fa-solid fa-user-edit" aria-label="user-edit"></i> Exercise: Tidyverse functions and pipes</h4>
<p>Use a “pipeline” like above to output a data frame with two columns (<code>sample_id</code> and <code>treatment</code>) only for the <em>cathemerium</em> treatment. How many rows does your output data frame have?</p>
<details>
<summary>
Click for the solution
</summary>
<div class="cell">
<div class="sourceCode cell-code" id="cb112"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb112-1"><a href="#cb112-1" aria-hidden="true" tabindex="-1"></a>meta <span class="sc">|&gt;</span></span>
<span id="cb112-2"><a href="#cb112-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(sample_id, treatment) <span class="sc">|&gt;</span></span>
<span id="cb112-3"><a href="#cb112-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(treatment <span class="sc">==</span> <span class="st">"cathemerium"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="regular-r-output"><code># A tibble: 7 × 2
  sample_id   treatment  
  &lt;chr&gt;       &lt;chr&gt;      
1 ERR10802864 cathemerium
2 ERR10802867 cathemerium
3 ERR10802870 cathemerium
4 ERR10802882 cathemerium
5 ERR10802875 cathemerium
6 ERR10802879 cathemerium
7 ERR10802883 cathemerium</code></pre>
</div>
</div>
<p>The output data frame has 7 rows and 2 columns: note that “7 x 2” is helpfully shown at the top of the output in the console.</p>
</details>
</section>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-3-contents" aria-controls="callout-3" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Optional: Using the pipe keyboard shortcut <em>(Click to expand)</em>
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-3" class="callout-3-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>The following RStudio keyboard shortcut will insert a pipe symbol: <kbd>Ctrl/Cmd</kbd> + <kbd>Shift</kbd> + <kbd>M</kbd>.</p>
<p>However, by default this (still) inserts an older pipe symbol, <code>%&gt;%</code>. To make RStudio insert the <code>|&gt;</code> pipe symbol with this keyboard shortcut:</p>
<ol type="1">
<li>Click <code>Tools</code> &gt; <code>Global Options</code> &gt; <code>Code</code> tab on the right</li>
<li>Check the box “Use native pipe operator” as shown below</li>
<li>Click <code>OK</code> at the bottom of the options dialog box to apply the change and close the box.</li>
</ol>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="../img/pipe-shortcut.png" class="img-fluid quarto-figure quarto-figure-center figure-img"></p>
</figure>
</div>
</div>
</div>
</div>


</section>
</section>


<a onclick="window.scrollTo(0, 0); return false;" role="button" id="quarto-back-to-top"><i class="bi bi-arrow-up"></i> Back to top</a><div id="quarto-appendix" class="default"><section id="footnotes" class="footnotes footnotes-end-of-document" role="doc-endnotes"><h2 class="anchored quarto-appendix-heading">Footnotes</h2>

<ol>
<li id="fn1"><p> In R, there is no distinction between options and arguments like with Shell commands.<a href="#fnref1" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn2"><p> And unlike unnamed arguments, named arguments can be presented in any order you like.<a href="#fnref2" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn3"><p>This also applies at OSC<a href="#fnref3" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn4"><p> This should be <em>much</em> faster on your own computer.<a href="#fnref4" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn5"><p>The latter two only with numeric vectors - as vectors can also contain text.<a href="#fnref5" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
</ol>
</section></div></main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp("https:\/\/jelmerp\.github\.io\/ENTMLGY6703-SP26");
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
<p>2026, Jelmer Poelstra, <a href="https://creativecommons.org/licenses/by-nc/4.0/">CC BY-NC 4.0</a> &nbsp; <img src="http://mirrors.creativecommons.org/presskit/buttons/88x31/png/by-nc.png" class="img-fluid" style="width:10.0%"></p>
</div>   
    <div class="nav-footer-center">
<p><img src="../img/CFAES-Horiz-RGBHEX.svg" class="img-fluid" style="width:40.0%"></p>
</div>
    <div class="nav-footer-right">
      <ul class="footer-items list-unstyled">
    <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/jelmerp/ENTMLGY6703-SP26">
      <i class="bi bi-github" role="img" aria-label="The GitHub repo for this website">
</i> 
    </a>
  </li>  
</ul>
    </div>
  </div>
</footer>




</body></html>