<!DOCTYPE html>
<html lang="en"><head>
<link href="../img/favicon.png" rel="icon" type="image/png">
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-html/tabby.min.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/light-border.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting-1092c56c6cadf2eb47b1bc8063ab382a.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<link href="../site_libs/quarto-contrib/fontawesome6-6.7.2/all.min.css" rel="stylesheet">
<link href="../site_libs/quarto-contrib/fontawesome6-6.7.2/latex-fontsize.css" rel="stylesheet">
<script src="../site_libs/quarto-contrib/glightbox/glightbox.min.js"></script>
<link href="../site_libs/quarto-contrib/glightbox/glightbox.min.css" rel="stylesheet">
<link href="../site_libs/quarto-contrib/glightbox/lightbox.css" rel="stylesheet"><meta charset="utf-8">
  <meta name="generator" content="quarto-1.6.40">

  <meta name="author" content="Jelmer Poelstra">
  <meta name="dcterms.date" content="2026-02-05">
  <title>ENTMLGY6703-SP26 - Jelmer's lectures ‚Äì Transcriptomics with RNA-Seq</title>
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no, minimal-ui">
  <link rel="stylesheet" href="../site_libs/revealjs/dist/reset.css">
  <link rel="stylesheet" href="../site_libs/revealjs/dist/reveal.css">
  <style>
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    div.columns{display: flex; gap: min(4vw, 1.5em);}
    div.column{flex: auto; overflow-x: auto;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
    ul.task-list li input[type="checkbox"] {
      width: 0.8em;
      margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
      vertical-align: middle;
    }
    /* CSS for citations */
    div.csl-bib-body { }
    div.csl-entry {
      clear: both;
      margin-bottom: 0em;
    }
    .hanging-indent div.csl-entry {
      margin-left:2em;
      text-indent:-2em;
    }
    div.csl-left-margin {
      min-width:2em;
      float:left;
    }
    div.csl-right-inline {
      margin-left:2em;
      padding-left:1em;
    }
    div.csl-indent {
      margin-left: 2em;
    }  </style>
  <link rel="stylesheet" href="../site_libs/revealjs/dist/theme/quarto-fb88a47fb37db585264132e6c4198fc7.css">
  <link rel="stylesheet" href="../slides.css">
  <link href="../site_libs/revealjs/plugin/quarto-line-highlight/line-highlight.css" rel="stylesheet">
  <link href="../site_libs/revealjs/plugin/reveal-menu/menu.css" rel="stylesheet">
  <link href="../site_libs/revealjs/plugin/reveal-menu/quarto-menu.css" rel="stylesheet">
  <link href="../site_libs/revealjs/plugin/reveal-chalkboard/font-awesome/css/all.css" rel="stylesheet">
  <link href="../site_libs/revealjs/plugin/reveal-chalkboard/style.css" rel="stylesheet">
  <link href="../site_libs/revealjs/plugin/quarto-support/footer.css" rel="stylesheet">
  <style type="text/css">
    .reveal div.sourceCode {
      margin: 0;
      overflow: auto;
    }
    .reveal div.hanging-indent {
      margin-left: 1em;
      text-indent: -1em;
    }
    .reveal .slide:not(.center) {
      height: 100%;
    }
    .reveal .slide.scrollable {
      overflow-y: auto;
    }
    .reveal .footnotes {
      height: 100%;
      overflow-y: auto;
    }
    .reveal .slide .absolute {
      position: absolute;
      display: block;
    }
    .reveal .footnotes ol {
      counter-reset: ol;
      list-style-type: none; 
      margin-left: 0;
    }
    .reveal .footnotes ol li:before {
      counter-increment: ol;
      content: counter(ol) ". "; 
    }
    .reveal .footnotes ol li > p:first-child {
      display: inline-block;
    }
    .reveal .slide ul,
    .reveal .slide ol {
      margin-bottom: 0.5em;
    }
    .reveal .slide ul li,
    .reveal .slide ol li {
      margin-top: 0.4em;
      margin-bottom: 0.2em;
    }
    .reveal .slide ul[role="tablist"] li {
      margin-bottom: 0;
    }
    .reveal .slide ul li > *:first-child,
    .reveal .slide ol li > *:first-child {
      margin-block-start: 0;
    }
    .reveal .slide ul li > *:last-child,
    .reveal .slide ol li > *:last-child {
      margin-block-end: 0;
    }
    .reveal .slide .columns:nth-child(3) {
      margin-block-start: 0.8em;
    }
    .reveal blockquote {
      box-shadow: none;
    }
    .reveal .tippy-content>* {
      margin-top: 0.2em;
      margin-bottom: 0.7em;
    }
    .reveal .tippy-content>*:last-child {
      margin-bottom: 0.2em;
    }
    .reveal .slide > img.stretch.quarto-figure-center,
    .reveal .slide > img.r-stretch.quarto-figure-center {
      display: block;
      margin-left: auto;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-left,
    .reveal .slide > img.r-stretch.quarto-figure-left  {
      display: block;
      margin-left: 0;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-right,
    .reveal .slide > img.r-stretch.quarto-figure-right  {
      display: block;
      margin-left: auto;
      margin-right: 0; 
    }
  </style>
</head>
<body class="quarto-light">
  <div class="reveal">
    <div class="slides">

<section id="title-slide" class="quarto-title-block center">
  <h1 class="title">Transcriptomics with RNA-Seq</h1>

<div class="quarto-title-authors">
<div class="quarto-title-author">
<div class="quarto-title-author-name">
Jelmer Poelstra 
</div>
        <p class="quarto-title-affiliation">
            CFAES Bioinformatics Core, OSU
          </p>
    </div>
</div>

  <p class="date">2026-02-05</p>
</section><section id="TOC">
<nav role="doc-toc"> 
<h2 id="toc-title">Overview</h2>
<ul>
<li><a href="#/why-study-gene-expression" id="/toc-why-study-gene-expression">Why study gene expression?</a></li>
<li><a href="#/introduction-to-rna-seq" id="/toc-introduction-to-rna-seq">Introduction to RNA-Seq</a></li>
<li><a href="#/stage-i-experimental-design" id="/toc-stage-i-experimental-design">Stage I: Experimental design</a></li>
<li><a href="#/stage-ii-from-samples-to-reads" id="/toc-stage-ii-from-samples-to-reads">Stage II: From samples to reads</a></li>
<li><a href="#/stage-iii-from-reads-to-counts" id="/toc-stage-iii-from-reads-to-counts">Stage III: From reads to counts</a></li>
<li><a href="#/stage-iv-from-counts-to-conclusions" id="/toc-stage-iv-from-counts-to-conclusions">Stage IV: From counts to conclusions</a></li>
<li><a href="#/questions" id="/toc-questions">Questions?</a></li>
</ul>
</nav>
</section>
<section>
<section id="why-study-gene-expression" class="title-slide slide level1 center" data-background-color="silver">
<h1>Why study gene expression?</h1>

</section>
<section id="gene-expression-is-highly-dynamic" class="slide level2">
<h2>Gene expression is highly dynamic</h2>
<p>Unlike an organism‚Äôs genome content, the expression of genes is highly dynamic ‚Äì it varies:</p>
<ul>
<li><p>Across time and space</p></li>
<li><p>Qualitatively (<em>which</em> genes are expressed) and especially quantitatively (<em>how much</em> of each gene is expressed)</p></li>
</ul>
</section>
<section id="gene-expression-is-functionally-important" class="slide level2">
<h2>Gene expression is functionally important</h2>
<p>Considering:</p>
<div>
<ol type="1">
<li class="fragment"><p>That <strong>protein production</strong> tells us about the activity of biological functions,<br>
and the molecular mechanisms underlying those functions</p></li>
<li class="fragment"><p>That it is easier to measure <strong>transcript (mRNA)</strong> than protein abundance</p></li>
<li class="fragment"><p>The central dogma</p></li>
</ol>
</div>
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="../img/centraldogma3.png" class="quarto-figure quarto-figure-center" style="width:100.0%"></p>
</figure>
</div>
<div class="fragment">
<p>‚Ä¶ <strong><em>gene expression can be used as a proxy for protein expression</em></strong> to make functional inferences</p>
</div>
<div class="fragment">
<p><br></p>
<div class="callout callout-warning callout-style-simple">
<div class="callout-body">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-content">
<p><strong>Caveat:</strong> the correlation between mRNA and protein levels is rather imperfect ‚Äì<br>
see e.g. <span class="citation" data-cites="ponomarenko2023">Ponomarenko et al. (<a href="#/references" role="doc-biblioref" onclick="">2023</a>)</span>.</p>
</div>
</div>
</div>
</div>
</section>
<section id="studying-gene-expression" class="slide level2">
<h2>Studying gene expression</h2>
<p>For example, gene expression can be quantified to associate genes and genetic pathways with:</p>
<ul>
<li>Phenotypic responses e.g.&nbsp;to environmental conditions</li>
<li>Differences between groups (treatments, genotypes, sexes, tissues, etc.)</li>
</ul>
<div class="fragment">
<hr style="height:1pt; visibility:hidden;">
<p>So, studying gene expression can <strong>identify candidate genes underlying phenotypes ‚Äì</strong><br>
these can next by functionally validated using experiments like knock-outs.</p>
<p><br></p>
</div>
<div class="fragment">
<div class="callout callout-tip callout-style-simple">
<div class="callout-body">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-content">
<p>And further down the line, understanding molecular basis of phenotypes has all sorts of uses!</p>
<p>E.g, in an agricultural context, to <strong>manipulate</strong> phenotypes such as resistance, pathogenicity, and yield.</p>
</div>
</div>
</div>
</div>
</section>
<section id="how-to-quantify-gene-expression" class="slide level2">
<h2>How to quantify gene expression?</h2>
<ul>
<li>Techniques like <strong>qPCR/ddPCR</strong> can quantify expression of a few genes at a time
<ul>
<li>Useful for hypothesis-driven research on specific genes</li>
<li>Not suitable for discovery-based research on many genes</li>
</ul></li>
</ul>
<hr style="height:1pt; visibility:hidden;">
<div class="fragment">
<ul>
<li>With <strong>transcriptomics</strong>, expression of thousands of genes is quantified simultaneously
<ul>
<li>It was already common practice before HTS: with microarrays</li>
<li>But is now dominated by the HTS-based method <strong>RNA-Seq</strong></li>
</ul></li>
</ul>
</div>
</section></section>
<section>
<section id="introduction-to-rna-seq" class="title-slide slide level1 center" data-background-color="silver">
<h1>Introduction to RNA-Seq</h1>

</section>
<section id="what-is-rna-seq" class="slide level2">
<h2>What is RNA-Seq?</h2>
<ul>
<li><p>To estimate gene expression levels genome-wide, RNA-Seq takes a brute-force approach by <strong>randomly sequencing of millions of RNA fragments per sample</strong></p></li>
<li><p>The resulting reads can be assigned a gene of origin,<br>
and the core idea is that a gene‚Äôs read count reflects that gene‚Äôs expression level</p></li>
</ul>
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="../img/reads.png" class="quarto-figure quarto-figure-center" style="width:35.0%"></p>
</figure>
</div>
<div class="fragment">
<details>
<summary>
<strong>How can we tell which gene each read originates from?</strong> &nbsp; <span style="color:darkolivegreen;"><i class="fa-solid fa-comments" aria-label="comments"></i></span>
</summary>
Most commonly by aligning the reads to a reference genome.
</details>
<div class="footer">

</div>
</div>
</section>
<section id="what-is-rna-seq-1" class="slide level2">
<h2>What is RNA-Seq?</h2>
<ul>
<li><p>To estimate gene expression levels genome-wide, RNA-Seq takes a brute-force approach by <strong>randomly sequencing of millions of RNA fragments per sample</strong></p></li>
<li><p>The resulting reads can be assigned a gene of origin,<br>
and the core idea is that a gene‚Äôs read count reflects that gene‚Äôs expression level</p></li>
</ul>
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="../img/reads.png" class="quarto-figure quarto-figure-center" style="width:35.0%"></p>
</figure>
</div>
<div class="callout callout-note callout-style-simple">
<div class="callout-body">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-content">
<p>RNA-Seq is a very widely used technique ‚Äî<br>
it constitutes <em>the</em> most common usage of high-throughput sequencing</p>
</div>
</div>
</div>
<div class="footer">

</div>
</section>
<section id="the-most-common-type-of-rna-seq" class="slide level2">
<h2>The most common type of RNA-Seq</h2>
<p>We‚Äôll focus on the most common type of RNA-Seq, which:</p>
<div>
<ol type="1">
<li class="fragment">Does not sequence RNA directly, but first <strong>reverse transcribes RNA to cDNA</strong></li>
<li class="fragment">Attempts to sequence only mRNA, avoiding non-coding RNAs (‚Äú<strong>mRNA-Seq</strong>‚Äù)</li>
<li class="fragment">Does not distinguish between RNA from different cell types (‚Äú<strong>bulk RNA-Seq</strong>‚Äù)</li>
<li class="fragment">Uses <strong>short reads</strong> (‚â§150 bp) that do not cover full transcripts but do uniquely ID genes</li>
<li class="fragment">Uses a reference genome for read alignment (‚Äú<strong>reference-based RNA-Seq</strong>‚Äù)</li>
</ol>
</div>
<div class="fragment">
<hr style="height:1pt; visibility:hidden;">
<div class="callout callout-note callout-style-simple">
<div class="callout-body">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-content">
<p><em>When I refer to RNA-Seq from now on, this specific type of RNA-Seq is implied.</em></p>
</div>
</div>
</div>
</div>
</section>
<section id="side-note-types-of-rna" class="slide level2">
<h2>Side note: Types of RNA</h2>
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="../img/rna-types2.png" class="quarto-figure quarto-figure-center" style="width:83.0%"></p>
</figure>
</div>
<div class="legend2">
<p>From <a href="https://biorender.com/">BioRender</a></p>
</div>
<div class="footer">

</div>
</section>
<section id="rna-seq-project-examples" class="slide level2">
<h2>RNA-Seq project examples</h2>
<p>RNA-Seq is the most common data type I help analyze in my role.<br>
For a taste of what it‚Äôs used for in CFAES and beyond ‚Äî<br>
the following projects aimed to identify genes &amp; pathways differing between:</p>
<div class="fragment">
<ul>
<li><p>Soybean cultivars in response to <em>Phytophtora sojae</em> inoculation <em>(Dorrance lab, PlantPath)</em></p></li>
<li><p>Response of maize to infection with <em>Pantoea</em> with and without an effector gene <em>(Mackey lab, HCS)</em></p></li>
</ul>
</div>
<div class="fragment">
<ul>
<li><p>Mated and unmated mosquitos <em>(Sirot lab, College of Wooster)</em></p></li>
<li><p>Tissues of the ambrosia beetle and its symbiotic fungus <em>(Ranger lab, USDA)</em></p></li>
<li><p>Diapause-inducing conditions for two pest stink bug species <em>(Michel lab, Entomology)</em></p></li>
</ul>
</div>
<div class="fragment">
<ul>
<li><p>Pig coronaviruses with vs.&nbsp;without an experimental insertion <em>(Wang lab, CFAH)</em></p></li>
<li><p>Human carcinoma cell lines with vs.&nbsp;without a manipulated gene <em>(Cruz lab, CCC)</em></p></li>
</ul>
</div>
<div class="fragment">
<p>And to improve the annotation of a nematode genome assembly <em>(Taylor lab, PlantPath)</em></p>
</div>
</section></section>
<section>
<section id="stage-i-experimental-design" class="title-slide slide level1 center" data-background-color="silver">
<h1>Stage I: Experimental design</h1>

</section>
<section id="experimental-design-treatmentsgroups" class="slide level2">
<h2>Experimental design: treatments/groups</h2>
<p>RNA-Seq typically compares groups of samples defined by differences in:</p>
<ul>
<li><p><strong>Treatments</strong> ‚Äî e.g.&nbsp;different host plants, temperature, diet, mated/unmated</p></li>
<li><p><strong>Organismal variants</strong> ‚Äî e.g.&nbsp;ages/developmental stages, sexes, subspecies</p></li>
<li><p><strong>Tissues</strong></p></li>
</ul>
</section>
<section id="the-garrig√≥s-et-al.-experimental-design" class="slide level2">
<h2>The Garrig√≥s et al.&nbsp;experimental design</h2>
<div class="quarto-figure quarto-figure-center">
<figure>
<p><a href="../img/garrigos_paper.png" class="lightbox" data-gallery="quarto-lightbox-gallery-1"><img data-src="../img/garrigos_paper.png" class="quarto-figure quarto-figure-center" style="width:75.0%" alt="A screenshot of the paper's front matter."></a></p>
</figure>
</div>
<div class="legend2">
<p><span class="citation" data-cites="garrig√≥s2025">Garrig√≥s et al. (<a href="#/references" role="doc-biblioref" onclick="">2025</a>)</span></p>
</div>
<hr style="height:1pt; visibility:hidden;">
<p><em>Culex pipiens</em> mosquitos infected with malaria-causing <em>Plasmodium</em> protozoans:</p>
<ul>
<li><em>Plasmodium relictum</em> ‚Äì higher virulence, lower tranmission rate</li>
<li><em>Plasmodium cathemerium</em> ‚Äì lower virulence, higher tranmission rate</li>
</ul>
</section>
<section id="the-garrig√≥s-et-al.-experimental-design-1" class="slide level2">
<h2>The Garrig√≥s et al.&nbsp;experimental design</h2>
<div class="quarto-figure quarto-figure-center">
<figure>
<p><a href="../img/exp-design_a.png" class="lightbox" data-gallery="quarto-lightbox-gallery-2"><img data-src="../img/exp-design_a.png" class="quarto-figure quarto-figure-center" style="width:75.0%" alt="A diagram showing the experimental design of the Garrig√≥s et al. 2025 study."></a></p>
</figure>
</div>
<div class="footer">

</div>
</section>
<section id="the-garrig√≥s-et-al.-experimental-design-2" class="slide level2">
<h2>The Garrig√≥s et al.&nbsp;experimental design</h2>
<div class="quarto-figure quarto-figure-center">
<figure>
<p><a href="../img/exp-design_b.png" class="lightbox" data-gallery="quarto-lightbox-gallery-3"><img data-src="../img/exp-design_b.png" class="quarto-figure quarto-figure-center" style="width:75.0%" alt="A diagram showing the experimental design of the Garrig√≥s et al. 2025 study."></a></p>
</figure>
</div>
<div class="footer">

</div>
</section>
<section id="the-garrig√≥s-et-al.-experimental-design-3" class="slide level2">
<h2>The Garrig√≥s et al.&nbsp;experimental design</h2>
<div class="quarto-figure quarto-figure-center">
<figure>
<p><a href="../img/exp-design_c.png" class="lightbox" data-gallery="quarto-lightbox-gallery-4"><img data-src="../img/exp-design_c.png" class="quarto-figure quarto-figure-center" style="width:75.0%" alt="A diagram showing the experimental design of the Garrig√≥s et al. 2025 study."></a></p>
</figure>
</div>
<div class="footer">

</div>
</section>
<section id="experimental-design" class="slide level2">
<h2>Experimental design</h2>
<details>
<summary>
<em>With this experimental design,</em> <em>would it be appropriate to sequence 9 samples total?</em> &nbsp; <span style="color:darkolivegreen;"><i class="fa-solid fa-comments" aria-label="comments"></i></span>
</summary>
No ‚Äì see the next slide
</details>
<div class="quarto-figure quarto-figure-center">
<figure>
<p><a href="../img/exp-design_c.png" class="lightbox" data-gallery="quarto-lightbox-gallery-5"><img data-src="../img/exp-design_c.png" class="quarto-figure quarto-figure-center" style="width:65.0%" alt="A diagram showing the experimental design of the Garrig√≥s et al. 2025 study."></a></p>
</figure>
</div>
<div class="footer">

</div>
</section>
<section id="experimental-design-biological-replicates" class="slide level2">
<h2>Experimental design: Biological replicates</h2>
<p>To make statistically supported conclusions about expression differences,<br>
we need <strong>biological replication</strong> (at least 3-5 samples per group):</p>
<div class="quarto-figure quarto-figure-center">
<figure>
<p><a href="../img/exp-design_d.png" class="lightbox" data-gallery="quarto-lightbox-gallery-6"><img data-src="../img/exp-design_d.png" class="quarto-figure quarto-figure-center" style="width:65.0%" alt="A diagram showing the experimental design of the Garrig√≥s et al. 2025 study."></a></p>
</figure>
</div>
<div class="footer">

</div>
</section></section>
<section>
<section id="stage-ii-from-samples-to-reads" class="title-slide slide level1 center" data-background-color="silver">
<h1>Stage II: From samples to reads</h1>

</section>
<section id="from-samples-to-reads" class="slide level2">
<h2>From samples to reads</h2>
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="../img/samples2reads_a2.png" class="quarto-figure quarto-figure-center" style="width:80.0%"></p>
</figure>
</div>
<details>
<summary>
<em>Which type of RNA do the red bars followed by <span style="color:red;">AAAAAAA</span> represent?</em> &nbsp; <span style="color:darkolivegreen;"><i class="fa-solid fa-comments" aria-label="comments"></i></span>
</summary>
These are the mRNAs, which have poly-A tails (<code>AAAAAA...</code>)
</details>
</section>
<section id="from-samples-to-reads-1" class="slide level2">
<h2>From samples to reads</h2>
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="../img/samples2reads_a2.png" class="quarto-figure quarto-figure-center" style="width:80.0%"></p>
</figure>
</div>
<p>We want mRNAs but these often make up only a few percent of RNAs!<br>
The two main ways to select for mRNAs are <strong>poly-A selection</strong> and <strong>ribo-depletion</strong>.</p>
<div class="fragment">
<hr style="height:1pt; visibility:hidden;">
<div class="callout callout-note callout-titled callout-style-default">
<div class="callout-body">
<div class="callout-title">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<p><strong>As mentioned in the previous lecture:</strong></p>
</div>
<div class="callout-content">
<ul>
<li><p>Library preparation is typically done by sequencing facilities / companies</p></li>
<li><p>Many samples can be ‚Äúmultiplexed‚Äù into a single (RNA-Seq) library</p></li>
</ul>
</div>
</div>
</div>
</div>
</section>
<section id="from-samples-to-reads-2" class="slide level2">
<h2>From samples to reads</h2>
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="../img/samples2reads_b.png" class="quarto-figure quarto-figure-center" style="width:80.0%"></p>
</figure>
</div>
<div class="legend2">
<p>Modified after <a href="https://sydney-informatics-hub.github.io" class="uri">https://sydney-informatics-hub.github.io</a></p>
</div>
<div class="footer">

</div>
</section>
<section id="how-much-sequencing-is-needed" class="slide level2">
<h2>How much sequencing is needed?</h2>
<div class="fragment">
<ul>
<li>Guidelines highly approximate ‚Äî required amount depends not just on transcriptome size but also on expression level distribution, expression levels of genes of interest, etc.</li>
</ul>
</div>
<div class="fragment">
<ul>
<li>Typical recommendations are 20-50 million reads per sample in eukaryotes</li>
</ul>
</div>
</section></section>
<section>
<section id="stage-iii-from-reads-to-counts" class="title-slide slide level1 center" data-background-color="silver">
<h1>Stage III: From reads to counts</h1>

</section>
<section id="overview-of-rna-seq-data-analysis" class="slide level2">
<h2>Overview of RNA-Seq data analysis</h2>
<p>The analysis of RNA-Seq data can be divided into two main parts:</p>
<ol type="1">
<li><strong>From reads to counts</strong>: from the raw reads, produce a count table</li>
<li><strong>From counts to conclusions</strong>: analyze the count table to draw biological conclusions</li>
</ol>
</section>
<section id="a-quick-primer-on-the-count-table" class="slide level2">
<h2>A quick primer on the count table</h2>
<p>The <strong>count table</strong> has one row for each gene and one column for each sample,<br>
with the entries being the number of reads mapping to each gene in each sample:</p>
<div class="fragment">
<hr style="height:1pt; visibility:hidden;">
<table class="table-striped caption-top">
<thead>
<tr class="header">
<th></th>
<th style="text-align: center;">Sample 1</th>
<th style="text-align: center;">Sample 2</th>
<th style="text-align: center;">Sample 3</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><strong>Gene A</strong></td>
<td style="text-align: center;">1500</td>
<td style="text-align: center;">2300</td>
<td style="text-align: center;">1800</td>
</tr>
<tr class="even">
<td><strong>Gene B</strong></td>
<td style="text-align: center;">0</td>
<td style="text-align: center;">5</td>
<td style="text-align: center;">2</td>
</tr>
<tr class="odd">
<td><strong>Gene C</strong></td>
<td style="text-align: center;">300</td>
<td style="text-align: center;">250</td>
<td style="text-align: center;">400</td>
</tr>
</tbody>
</table>
<p><br></p>
<div class="callout callout-tip callout-style-simple">
<div class="callout-body">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-content">
<p><em>Actual count tables have thousands of genes (rows) and usually dozens of</em> <em>samples (columns)</em></p>
</div>
</div>
</div>
</div>
<div class="fragment">
<p><br></p>
<p><strong>How do you get to such a count table?</strong></p>
</div>
</section>
<section id="section" class="slide level2">
<h2></h2>
<div class="quarto-figure quarto-figure-center">
<figure>
<p><a href="../img/alignment1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-7"><img data-src="../img/alignment1.png" class="quarto-figure quarto-figure-center" style="width:82.0%" alt="A diagram of RNA-Seq analysis steps."></a></p>
</figure>
</div>
<div class="footer">

</div>
</section>
<section id="section-1" class="slide level2">
<h2></h2>
<div class="quarto-figure quarto-figure-center">
<figure>
<p><a href="../img/alignment2.png" class="lightbox" data-gallery="quarto-lightbox-gallery-8"><img data-src="../img/alignment2.png" class="quarto-figure quarto-figure-center" style="width:82.0%" alt="A diagram of RNA-Seq analysis steps."></a></p>
</figure>
</div>
<div class="footer">

</div>
</section>
<section id="section-2" class="slide level2">
<h2></h2>
<div class="quarto-figure quarto-figure-center">
<figure>
<p><a href="../img/alignment3.png" class="lightbox" data-gallery="quarto-lightbox-gallery-9"><img data-src="../img/alignment3.png" class="quarto-figure quarto-figure-center" style="width:82.0%" alt="A diagram of RNA-Seq analysis steps."></a></p>
</figure>
</div>
<div class="footer">

</div>
</section>
<section id="section-3" class="slide level2">
<h2></h2>
<div class="quarto-figure quarto-figure-center">
<figure>
<p><a href="../img/alignment4.png" class="lightbox" data-gallery="quarto-lightbox-gallery-10"><img data-src="../img/alignment4.png" class="quarto-figure quarto-figure-center" style="width:82.0%" alt="A diagram of RNA-Seq analysis steps."></a></p>
</figure>
</div>
<div class="footer">

</div>
</section>
<section id="section-4" class="slide level2">
<h2></h2>
<div class="quarto-figure quarto-figure-center">
<figure>
<p><a href="../img/alignment5.png" class="lightbox" data-gallery="quarto-lightbox-gallery-11"><img data-src="../img/alignment5.png" class="quarto-figure quarto-figure-center" style="width:82.0%" alt="A diagram of RNA-Seq analysis steps."></a></p>
</figure>
</div>
<div class="footer">

</div>
</section>
<section id="a-genai-diagram-that-explains-rna-seq-analysis" class="slide level2">
<h2>A genAI diagram that explains RNA-Seq analysis</h2>
<p>I though I‚Äôd get some generative AI help with the diagram on the previous slide ‚Äì<br>
this is what Adobe Firefly came up with: üòµ‚Äçüí´</p>
<div class="quarto-figure quarto-figure-center">
<figure>
<p><a href="../img/firefly_diagram.png" class="lightbox" data-gallery="quarto-lightbox-gallery-12"><img data-src="../img/firefly_diagram.png" class="quarto-figure quarto-figure-center" style="width:45.0%" alt="Adobe Firefly's image when asking it create a diagram of RNA-Seq analysis steps."></a></p>
</figure>
</div>
<div class="footer">

</div>
</section>
<section id="from-reads-to-counts-overview" class="slide level2">
<h2>From reads to counts: overview</h2>
<p>It specifically involves, at a minimum:</p>
<ol type="1">
<li><p><strong>Read preprocessing</strong></p></li>
<li><p><strong>Aligning reads to a reference genome</strong></p></li>
<li><p><strong>Quantifying expression levels</strong></p></li>
</ol>
<p><br></p>
<div class="fragment">
<div class="callout callout-note callout-style-simple">
<div class="callout-body">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-content">
<p>This part is ‚Äúbioinformatics‚Äù-heavy, with large files, high computing needs, and using command-line tools ran in the Unix shell.</p>
<p>If this is a problem: the process is fairly standardized and suitable to be outsourced.</p>
</div>
</div>
</div>
</div>
</section>
<section id="read-pre-processing" class="slide level2">
<h2>Read pre-processing</h2>
<p>Read pre-processing includes:</p>
<hr style="height:1pt; visibility:hidden;">
<ul>
<li><strong>Checking the quantity and quality of your reads</strong> (e.g.&nbsp;with <code>FastQC</code>)</li>
</ul>
<div class="fragment">
<ul>
<li><strong>Removing unwanted sequences</strong>, such as:
<ul>
<li>Adapters, low-quality bases, and very short reads</li>
<li>rRNA-derived reads (<em>optional</em>)</li>
<li>Contaminant sequences (<em>optional</em>)</li>
</ul></li>
</ul>
</div>
</section>
<section id="read-alignment-to-a-reference-genome" class="slide level2">
<h2>Read alignment to a reference genome</h2>
<details>
<summary>
Consider what you know about how mRNAs are produced. Does an RNA transcript always correspond to a contiguous stretch of the genome? If not, what implications does this have for read alignment? &nbsp; <span style="color:darkolivegreen;"><i class="fa-solid fa-comments" aria-label="comments"></i></span>
</summary>
No, it doesn‚Äôt, because of splicing. This means that a read can align partially to two different exons, skipping the intron in between. ‚ÄúRegular‚Äù genome alignment would not be able to handle this properly ‚Äì see the next slides.
</details>
</section>
<section id="side-note-splicing" class="slide level2">
<h2>Side note: Splicing</h2>
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="../img/splicing_biorender2a.png" class="quarto-figure quarto-figure-center" style="width:85.0%"></p>
</figure>
</div>
<div class="legend2">
<p>From <a href="https://biorender.com/">BioRender</a></p>
</div>
</section>
<section id="side-note-splicing-1" class="slide level2">
<h2>Side note: Splicing</h2>
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="../img/splicing_biorender2.png" class="quarto-figure quarto-figure-center" style="width:85.0%"></p>
</figure>
</div>
<div class="legend2">
<p>From <a href="https://biorender.com/">BioRender</a></p>
</div>
</section>
<section id="read-alignment-to-a-reference-genome-1" class="slide level2">
<h2>Read alignment to a reference genome</h2>
<p>So, the alignment of reads to a reference genome needs to be ‚Äú<em>splice-aware</em>‚Äù:</p>
<hr style="height:1pt; visibility:hidden;">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="../img/alignment_ed2.png" class="quarto-figure quarto-figure-center" style="width:70.0%"></p>
</figure>
</div>
<div class="legend2">
<p><span class="citation" data-cites="vandenberge2019">Van den Berge et al. (<a href="#/references" role="doc-biblioref" onclick="">2019</a>)</span></p>
</div>
<div class="footer">

</div>
</section>
<section id="read-alignment-to-a-reference-genome-2" class="slide level2">
<h2>Read alignment to a reference genome</h2>
<p>Alternatively, you can align to the <em>transcriptome</em> (i.e., all mature transcripts):</p>
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="../img/alignment_ed1.jpg" class="quarto-figure quarto-figure-center" style="width:70.0%"></p>
</figure>
</div>
<div class="legend2">
<p><span class="citation" data-cites="vandenberge2019">Van den Berge et al. (<a href="#/references" role="doc-biblioref" onclick="">2019</a>)</span></p>
</div>
<div class="footer">

</div>
</section>
<section id="read-alignment-to-a-reference-genome-3" class="slide level2">
<h2>Read alignment to a reference genome</h2>
<details>
<summary>
Why are there multiple bars in panel <strong>b</strong>? What do these represent? &nbsp; <span style="color:darkolivegreen;"><i class="fa-solid fa-comments" aria-label="comments"></i></span>
</summary>
<p>These represent different transcripts originating from the same gene due to <strong><em>alternative splicing</em></strong>. These will produce different proteins, which are called <strong><em>isoforms</em></strong>.</p>
<p>Most short-read RNA-Seq studies do not attempt to distinguish between isoforms,<br>
but rather quantify expression at the gene level.</p>
If distinguishing between isoforms is important, performing RNA-Seq with a <strong>long-read HTS</strong> platform is a better option.
</details>
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="../img/alignment_ed1.jpg" class="quarto-figure quarto-figure-center" style="width:70.0%"></p>
</figure>
</div>
<div class="legend2">
<p><span class="citation" data-cites="vandenberge2019">Van den Berge et al. (<a href="#/references" role="doc-biblioref" onclick="">2019</a>)</span></p>
</div>
<div class="footer">

</div>
</section>
<section id="gene-wise-quantification" class="slide level2">
<h2>Gene-wise quantification</h2>
<p>In essence, a simple counting exercise once you have the alignments in hand:<br>
for each sample, how many reads map to each gene?</p>
<div class="quarto-figure quarto-figure-center">
<figure>
<p><a href="../img/alignment_counting.png" class="lightbox" data-gallery="quarto-lightbox-gallery-13"><img data-src="../img/alignment_counting.png" class="quarto-figure quarto-figure-center" style="width:72.0%" alt="A diagram of RNA-Seq analysis steps."></a></p>
</figure>
</div>
<div class="fragment">
<div class="callout callout-note callout-style-simple">
<div class="callout-body">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-content">
<p>Though in practice, a bit more complicated than this, due to e.g.:</p>
<ul>
<li>Reads that map to multiple genes (‚Äúmulti-mapping reads‚Äù)</li>
<li>Sequencing biases and RNA fragmentation</li>
</ul>
</div>
</div>
</div>
</div>
</section>
<section id="a-best-practice-pipeline-to-produce-counts" class="slide level2">
<h2>A best-practice pipeline to produce counts</h2>
<p>The ‚Äúnf-core‚Äù initiative (<a href="https://nf-co.re" class="uri">https://nf-co.re</a>, <span class="citation" data-cites="ewels2020">Ewels et al. (<a href="#/references" role="doc-biblioref" onclick="">2020</a>)</span>) aims to produce best-practice and automated bioinformatics pipelines, like for RNA-Seq (<a href="https://nf-co.re/rnaseq" class="uri">https://nf-co.re/rnaseq</a>):</p>
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="../img/nfcore_rnaseq.png" class="quarto-figure quarto-figure-center" style="width:100.0%"></p>
</figure>
</div>
</section></section>
<section>
<section id="stage-iv-from-counts-to-conclusions" class="title-slide slide level1 center" data-background-color="silver">
<h1>Stage IV: From counts to conclusions</h1>

</section>
<section id="count-table-analysis-overview" class="slide level2">
<h2>Count table analysis: overview</h2>
<p>The second part of RNA-Seq data analysis involves analyzing the count table.</p>
<p>In contrast to the first part, this can be done on a laptop and instead is heavier on<br>
<strong>statistics</strong>, <strong>data visualization</strong> and <strong>biological interpretation</strong>.</p>
<p><br></p>
<div class="fragment">
<p>It is typically done with the R language, and common aspects include:</p>
<ul>
<li><p><strong>Principal Component Analysis (PCA)</strong><br>
Assessing overall sample ‚Äúclustering‚Äù (similarity) patterns</p></li>
<li><p><strong>Differential Expression (DE) analysis</strong><br>
Finding genes that differ in expression level between sample groups (DEGs)</p></li>
<li><p><strong>Functional enrichment analysis</strong><br>
See whether certain gene function categories are overrepresented among DEGs</p></li>
</ul>
</div>
</section>
<section id="principal-component-analysis-pca" class="slide level2">
<h2>Principal Component Analysis (PCA)</h2>
<p>PCA examines overall patterns of dissimilarity among samples,<br>
such as whether groups of interest form distinct clusters:</p>
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="../img/PCA.jpeg" class="quarto-figure quarto-figure-center" style="width:85.0%"></p>
</figure>
</div>
<div class="legend2">
<p>Fig. 1 from <span class="citation" data-cites="garrig√≥s2025">Garrig√≥s et al. (<a href="#/references" role="doc-biblioref" onclick="">2025</a>)</span></p>
</div>
<div class="fragment">
<hr style="height:1pt; visibility:hidden;">
<div class="callout callout-note callout-style-simple">
<div class="callout-body">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-content">
<p>We‚Äôll talk more about the interpretation of this PCA plot in tomorrow‚Äôs lab</p>
</div>
</div>
</div>
<div class="footer">

</div>
</div>
</section>
<section id="principal-component-analysis-pca-1" class="slide level2">
<h2>Principal Component Analysis (PCA)</h2>
<p>PCA examines overall patterns of dissimilarity among samples,<br>
such as whether groups of interest form distinct clusters:</p>
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="../img/PCA.jpeg" class="quarto-figure quarto-figure-center" style="width:85.0%"></p>
</figure>
</div>
<div class="legend2">
<p>Fig. 1 from <span class="citation" data-cites="garrig√≥s2025">Garrig√≥s et al. (<a href="#/references" role="doc-biblioref" onclick="">2025</a>)</span></p>
</div>
<hr style="height:1pt; visibility:hidden;">
<div class="callout callout-tip callout-style-simple">
<div class="callout-body">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-content">
<p>PCA is a very useful technique, not just for RNA-Seq data.<br>
To learn more about how it works, see <a href="https://www.youtube.com/watch?v=HMOI_lkzW08">this video</a> (short overview) and <a href="https://www.youtube.com/watch?v=FgakZw6K1QQ">this video</a> (more detailed).</p>
</div>
</div>
</div>
<div class="footer">

</div>
</section>
<section id="differential-expression-de-analysis" class="slide level2">
<h2>Differential expression (DE) analysis</h2>
<p>Differential Expression (DE) analysis allows you to test, separately for every expressed gene in your dataset, whether it significantly differs in expression level between groups.</p>
<div class="fragment">
<p>Typically, this is done with pairwise comparisons between groups:</p>
<div class="columns">
<div class="column">
<p><img data-src="../img/geneplot.png"></p>
</div><div class="column">

</div></div>
</div>
</section>
<section id="differential-expression-de-analysis-1" class="slide level2">
<h2>Differential expression (DE) analysis</h2>
<p>Differential Expression (DE) analysis allows you to test, separately for every expressed gene in your dataset, whether it significantly differs in expression level between groups.</p>
<p>Typically, this is done with pairwise comparisons between groups:</p>
<div class="columns">
<div class="column">
<p><img data-src="../img/geneplot.png"></p>
</div><div class="column">
<p><img data-src="../img/volcano.png"></p>
</div></div>
</section>
<section id="statistical-considerations-for-the-de-analysis" class="slide level2">
<h2>Statistical considerations for the DE analysis</h2>
<ol type="1">
<li>Gene count normalization</li>
<li>Probability distribution of the gene count data</li>
<li>Multiple-testing correction</li>
</ol>
<div class="callout callout-note callout-style-simple">
<div class="callout-body">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-content">
<p>We will talk about these during tomorrow‚Äôs lab.</p>
<p>While not necessarily easy conceptually, this is all fairly straightforward in practice: specialized R packages like <code>DESeq2</code> take care of the details.</p>
</div>
</div>
</div>
</section>
<section id="functional-enrichment-introduction" class="slide level2">
<h2>Functional enrichment: introduction</h2>
<p>Lists of DEGs can be long and not always easy to make biological sense of.</p>
<p>Functional <strong>enrichment analysis</strong> helps with this, asking whether certain <strong>functional categories</strong> of genes are statistically overrepresented among DEGs.</p>
<p><br></p>
<div class="fragment">
<p>Several databases group genes into such functional categories ‚Äî<br>
the two main ones used for enrichment analysis are:</p>
<ul>
<li><strong>Gene Ontology (GO)</strong><br>
Have a hierarchical structure with more specific terms grouping into more general terms</li>
<li><strong>Kyoto Encyclopedia of Genes and Genomes (KEGG)</strong><br>
Less general than GO, has pathways whose genes can be drawn and connected in diagrams</li>
</ul>
</div>
</section>
<section id="functional-enrichment-conceptually" class="slide level2">
<h2>Functional enrichment: conceptually</h2>
<p>A GO term called ‚Äú<strong>photosynthesis</strong>‚Äù is associated with a set of genes involved in photosynthesis.</p>
<ul>
<li><p>Say that 300 genes out of 30,000 genes in the genome have been annotated to this term = 1%</p></li>
<li><p>Say that among your 50 out of your 500 DEGs are annotated to this term = 10%</p></li>
</ul>
<div class="fragment">
<p>The larger the difference between these percentages, the stronger the indication that the function in question ‚Äìhere, photosynthesis‚Äì is overrepresented among your DEGs.</p>
</div>
</section>
<section id="functional-enrichment-go" class="slide level2">
<h2>Functional enrichment: GO</h2>
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="../img/garrigo_go.jpeg" class="quarto-figure quarto-figure-center" style="width:60.0%"></p>
</figure>
</div>
<div class="legend2">
<p>Fig. 4 from <span class="citation" data-cites="garrig√≥s2025">Garrig√≥s et al. (<a href="#/references" role="doc-biblioref" onclick="">2025</a>)</span></p>
</div>
<div class="footer">

</div>
</section>
<section id="functional-enrichment-kegg" class="slide level2 nostretch">
<h2>Functional enrichment: KEGG</h2>
<p><br></p>
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="../img/kegg_example.jpg" class="quarto-figure quarto-figure-center" style="width:90.0%"></p>
</figure>
</div>
<div class="legend2">
<p><span class="citation" data-cites="rodriguez2020">Rodriguez et al. (<a href="#/references" role="doc-biblioref" onclick="">2020</a>)</span></p>
<blockquote>
<p><em>KEGG representation of up-regulated genes related to jasmonic acid (JA) signal</em> <em>transduction pathways (ko04075) in banana cv. Calcutta 4 after inoculation with</em> <em>Pseudocercospora fijiensis</em>. <em>Genes or chemicals up-regulated at any time point were highlighted in green.</em></p>
</blockquote>
</div>
</section></section>
<section>
<section id="questions" class="title-slide slide level1 center" data-background-color="silver">
<h1>Questions?</h1>

</section>
<section id="bonus-read-alignment-qc" class="slide level2">
<h2>Bonus: Read alignment QC</h2>
<ul>
<li><strong>Alignment rates</strong><br>
What percentage of reads was successfully aligned? <em>(Should be &gt;80%)</em></li>
</ul>
<div class="fragment">
<ul>
<li><strong>Alignment targets</strong><br>
What percentages of aligned reads mapped to <em>exons</em> vs.&nbsp;<em>introns</em> vs.&nbsp;<em>intergenic regions</em>?</li>
</ul>
</div>
<div class="fragment">
<hr style="height:1pt; visibility:hidden;">
<details>
<summary>
What might cause high intronic mapping rates? &nbsp; <span style="color:darkolivegreen;"><i class="fa-solid fa-comments" aria-label="comments"></i></span>
</summary>
An abundance of pre-mRNA versus mature-mRNA.
</details>
</div>
<div class="fragment">
<details>
<summary>
What might cause high intergenic mapping rates? &nbsp; <span style="color:darkolivegreen;"><i class="fa-solid fa-comments" aria-label="comments"></i></span>
</summary>
DNA contamination or poor genome assembly/annotation quality
</details>
</div>
</section>
<section id="references" class="slide level2 smaller scrollable">
<h2>References</h2>
<div class="footer">

</div>


<div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" role="list">
<div id="ref-ewels2020" class="csl-entry" role="listitem">
Ewels, Philip A., Alexander Peltzer, Sven Fillinger, Harshil Patel, Johannes Alneberg, Andreas Wilm, Maxime Ulysse Garcia, Paolo Di Tommaso, and Sven Nahnsen. 2020. <span>‚ÄúThe Nf-Core Framework for Community-Curated Bioinformatics Pipelines.‚Äù</span> <em>Nature Biotechnology</em> 38 (3): 276‚Äì78. <a href="https://doi.org/10.1038/s41587-020-0439-x">https://doi.org/10.1038/s41587-020-0439-x</a>.
</div>
<div id="ref-garrig√≥s2025" class="csl-entry" role="listitem">
Garrig√≥s, Marta, Guillem Ylla, Josu√© Mart√≠nez-de la Puente, Jordi Figuerola, and Mar√≠a Jos√© Ruiz-L√≥pez. 2025. <span>‚ÄúTwo Avian <em>Plasmodium</em> Species Trigger Different Transcriptional Responses on Their Vector <em>Culex <span>p</span>ipiens</em>.‚Äù</span> <em>Molecular Ecology</em> 34 (15): e17240. <a href="https://doi.org/10.1111/mec.17240">https://doi.org/10.1111/mec.17240</a>.
</div>
<div id="ref-ponomarenko2023" class="csl-entry" role="listitem">
Ponomarenko, Elena A., George S. Krasnov, Olga I. Kiseleva, Polina A. Kryukova, Viktoriia A. Arzumanian, Georgii V. Dolgalev, Ekaterina V. Ilgisonis, Andrey V. Lisitsa, and Ekaterina V. Poverennaya. 2023. <span>‚ÄúWorkability of <span class="nocase">mRNA</span> <span>Sequencing</span> for <span>Predicting</span> <span>Protein</span> <span>Abundance</span>.‚Äù</span> <em>Genes</em> 14 (11): 2065. <a href="https://doi.org/10.3390/genes14112065">https://doi.org/10.3390/genes14112065</a>.
</div>
<div id="ref-rodriguez2020" class="csl-entry" role="listitem">
Rodriguez, H√©ctor Alejandro, William F. Hidalgo, J. Danilo Sanchez, Riya C. Menezes, Bernd Schneider, Rafael Eduardo Arango, and Juan Gonzalo Morales. 2020. <span>‚ÄúDifferential Regulation of Jasmonic Acid Pathways in Resistant (Calcutta 4) and Susceptible (Williams) Banana Genotypes During the Interaction with Pseudocercospora Fijiensis.‚Äù</span> <em>Plant Pathology</em> 69 (5): 872‚Äì82. https://doi.org/<a href="https://doi.org/10.1111/ppa.13165">https://doi.org/10.1111/ppa.13165</a>.
</div>
<div id="ref-vandenberge2019" class="csl-entry" role="listitem">
Van den Berge, Koen, Charlotte Soneson, Mark D. Robinson, and Christian Heinis. 2019. <span>‚ÄúRNA Sequencing Data: Hitchhiker‚Äôs Guide to Expression Analysis.‚Äù</span> <em>Annual Review of Biomedical Data Science</em> 2 (1): 139‚Äì73. <a href="https://doi.org/10.1146/annurev-biodatasci-072018-021255">https://doi.org/10.1146/annurev-biodatasci-072018-021255</a>.
</div>
</div>
</section></section>
    </div>
  <div class="quarto-auto-generated-content" style="display: none;">
<p><img src="../img/CFAES-Horiz-RGBHEX.svg" class="slide-logo"></p>
<div class="footer footer-default">
<p><a href="https://mcic-osu.github.io/ENTMLGY6703-SP26">~ ENTMLGY6703-SP26 ~</a></p>
</div>
</div></div>

  <script>window.backupDefine = window.define; window.define = undefined;</script>
  <script src="../site_libs/revealjs/dist/reveal.js"></script>
  <!-- reveal.js plugins -->
  <script src="../site_libs/revealjs/plugin/quarto-line-highlight/line-highlight.js"></script>
  <script src="../site_libs/revealjs/plugin/pdf-export/pdfexport.js"></script>
  <script src="../site_libs/revealjs/plugin/reveal-menu/menu.js"></script>
  <script src="../site_libs/revealjs/plugin/reveal-menu/quarto-menu.js"></script>
  <script src="../site_libs/revealjs/plugin/reveal-chalkboard/plugin.js"></script>
  <script src="../site_libs/revealjs/plugin/quarto-support/support.js"></script>
  

  <script src="../site_libs/revealjs/plugin/notes/notes.js"></script>
  <script src="../site_libs/revealjs/plugin/search/search.js"></script>
  <script src="../site_libs/revealjs/plugin/zoom/zoom.js"></script>
  <script src="../site_libs/revealjs/plugin/math/math.js"></script>
  <script>window.define = window.backupDefine; window.backupDefine = undefined;</script>

  <script>

      // Full list of configuration options available at:
      // https://revealjs.com/config/
      Reveal.initialize({
'controlsAuto': true,
'previewLinksAuto': false,
'pdfSeparateFragments': false,
'autoAnimateEasing': "ease",
'autoAnimateDuration': 1,
'autoAnimateUnmatched': true,
'jumpToSlide': true,
'menu': {"side":"left","useTextContentForMissingTitles":true,"markers":false,"loadIcons":false,"custom":[{"title":"Tools","icon":"<i class=\"fas fa-gear\"></i>","content":"<ul class=\"slide-menu-items\">\n<li class=\"slide-tool-item active\" data-item=\"0\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.fullscreen(event)\"><kbd>f</kbd> Fullscreen</a></li>\n<li class=\"slide-tool-item\" data-item=\"1\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.speakerMode(event)\"><kbd>s</kbd> Speaker View</a></li>\n<li class=\"slide-tool-item\" data-item=\"2\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.overview(event)\"><kbd>o</kbd> Slide Overview</a></li>\n<li class=\"slide-tool-item\" data-item=\"3\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.togglePdfExport(event)\"><kbd>e</kbd> PDF Export Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"4\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleScrollView(event)\"><kbd>r</kbd> Scroll View Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"5\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleChalkboard(event)\"><kbd>b</kbd> Toggle Chalkboard</a></li>\n<li class=\"slide-tool-item\" data-item=\"6\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleNotesCanvas(event)\"><kbd>c</kbd> Toggle Notes Canvas</a></li>\n<li class=\"slide-tool-item\" data-item=\"7\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.downloadDrawings(event)\"><kbd>d</kbd> Download Drawings</a></li>\n<li class=\"slide-tool-item\" data-item=\"8\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.keyboardHelp(event)\"><kbd>?</kbd> Keyboard Help</a></li>\n</ul>"}],"openButton":true,"width":"wide","numbers":true},
'chalkboard': {"buttons":true,"theme":"whiteboard"},
'smaller': true,
 
        // Display controls in the bottom right corner
        controls: false,

        // Help the user learn the controls by providing hints, for example by
        // bouncing the down arrow when they first encounter a vertical slide
        controlsTutorial: false,

        // Determines where controls appear, "edges" or "bottom-right"
        controlsLayout: 'edges',

        // Visibility rule for backwards navigation arrows; "faded", "hidden"
        // or "visible"
        controlsBackArrows: 'faded',

        // Display a presentation progress bar
        progress: true,

        // Display the page number of the current slide
        slideNumber: 'c/t',

        // 'all', 'print', or 'speaker'
        showSlideNumber: 'all',

        // Add the current slide number to the URL hash so that reloading the
        // page/copying the URL will return you to the same slide
        hash: true,

        // Start with 1 for the hash rather than 0
        hashOneBasedIndex: false,

        // Flags if we should monitor the hash and change slides accordingly
        respondToHashChanges: true,

        // Push each slide change to the browser history
        history: true,

        // Enable keyboard shortcuts for navigation
        keyboard: true,

        // Enable the slide overview mode
        overview: true,

        // Disables the default reveal.js slide layout (scaling and centering)
        // so that you can use custom CSS layout
        disableLayout: false,

        // Vertical centering of slides
        center: false,

        // Enables touch navigation on devices with touch input
        touch: true,

        // Loop the presentation
        loop: false,

        // Change the presentation direction to be RTL
        rtl: false,

        // see https://revealjs.com/vertical-slides/#navigation-mode
        navigationMode: 'linear',

        // Randomizes the order of slides each time the presentation loads
        shuffle: false,

        // Turns fragments on and off globally
        fragments: true,

        // Flags whether to include the current fragment in the URL,
        // so that reloading brings you to the same fragment position
        fragmentInURL: false,

        // Flags if the presentation is running in an embedded mode,
        // i.e. contained within a limited portion of the screen
        embedded: false,

        // Flags if we should show a help overlay when the questionmark
        // key is pressed
        help: true,

        // Flags if it should be possible to pause the presentation (blackout)
        pause: true,

        // Flags if speaker notes should be visible to all viewers
        showNotes: false,

        // Global override for autoplaying embedded media (null/true/false)
        autoPlayMedia: null,

        // Global override for preloading lazy-loaded iframes (null/true/false)
        preloadIframes: null,

        // Number of milliseconds between automatically proceeding to the
        // next slide, disabled when set to 0, this value can be overwritten
        // by using a data-autoslide attribute on your slides
        autoSlide: 0,

        // Stop auto-sliding after user input
        autoSlideStoppable: true,

        // Use this method for navigation when auto-sliding
        autoSlideMethod: null,

        // Specify the average time in seconds that you think you will spend
        // presenting each slide. This is used to show a pacing timer in the
        // speaker view
        defaultTiming: null,

        // Enable slide navigation via mouse wheel
        mouseWheel: false,

        // The display mode that will be used to show slides
        display: 'block',

        // Hide cursor if inactive
        hideInactiveCursor: true,

        // Time before the cursor is hidden (in ms)
        hideCursorTime: 5000,

        // Opens links in an iframe preview overlay
        previewLinks: false,

        // Transition style (none/fade/slide/convex/concave/zoom)
        transition: 'none',

        // Transition speed (default/fast/slow)
        transitionSpeed: 'default',

        // Transition style for full page slide backgrounds
        // (none/fade/slide/convex/concave/zoom)
        backgroundTransition: 'none',

        // Number of slides away from the current that are visible
        viewDistance: 3,

        // Number of slides away from the current that are visible on mobile
        // devices. It is advisable to set this to a lower number than
        // viewDistance in order to save resources.
        mobileViewDistance: 2,

        // The "normal" size of the presentation, aspect ratio will be preserved
        // when the presentation is scaled to fit different resolutions. Can be
        // specified using percentage units.
        width: 1050,

        height: 700,

        // Factor of the display size that should remain empty around the content
        margin: 0.1,

        math: {
          mathjax: 'https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/MathJax.js',
          config: 'TeX-AMS_HTML-full',
          tex2jax: {
            inlineMath: [['\\(','\\)']],
            displayMath: [['\\[','\\]']],
            balanceBraces: true,
            processEscapes: false,
            processRefs: true,
            processEnvironments: true,
            preview: 'TeX',
            skipTags: ['script','noscript','style','textarea','pre','code'],
            ignoreClass: 'tex2jax_ignore',
            processClass: 'tex2jax_process'
          },
        },

        // reveal.js plugins
        plugins: [QuartoLineHighlight, PdfExport, RevealMenu, RevealChalkboard, QuartoSupport,

          RevealMath,
          RevealNotes,
          RevealSearch,
          RevealZoom
        ]
      });
    </script>
    <script id="quarto-html-after-body" type="application/javascript">
    window.document.addEventListener("DOMContentLoaded", function (event) {
      const toggleBodyColorMode = (bsSheetEl) => {
        const mode = bsSheetEl.getAttribute("data-mode");
        const bodyEl = window.document.querySelector("body");
        if (mode === "dark") {
          bodyEl.classList.add("quarto-dark");
          bodyEl.classList.remove("quarto-light");
        } else {
          bodyEl.classList.add("quarto-light");
          bodyEl.classList.remove("quarto-dark");
        }
      }
      const toggleBodyColorPrimary = () => {
        const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
        if (bsSheetEl) {
          toggleBodyColorMode(bsSheetEl);
        }
      }
      toggleBodyColorPrimary();  
      const tabsets =  window.document.querySelectorAll(".panel-tabset-tabby")
      tabsets.forEach(function(tabset) {
        const tabby = new Tabby('#' + tabset.id);
      });
      const isCodeAnnotation = (el) => {
        for (const clz of el.classList) {
          if (clz.startsWith('code-annotation-')) {                     
            return true;
          }
        }
        return false;
      }
      const onCopySuccess = function(e) {
        // button target
        const button = e.trigger;
        // don't keep focus
        button.blur();
        // flash "checked"
        button.classList.add('code-copy-button-checked');
        var currentTitle = button.getAttribute("title");
        button.setAttribute("title", "Copied!");
        let tooltip;
        if (window.bootstrap) {
          button.setAttribute("data-bs-toggle", "tooltip");
          button.setAttribute("data-bs-placement", "left");
          button.setAttribute("data-bs-title", "Copied!");
          tooltip = new bootstrap.Tooltip(button, 
            { trigger: "manual", 
              customClass: "code-copy-button-tooltip",
              offset: [0, -8]});
          tooltip.show();    
        }
        setTimeout(function() {
          if (tooltip) {
            tooltip.hide();
            button.removeAttribute("data-bs-title");
            button.removeAttribute("data-bs-toggle");
            button.removeAttribute("data-bs-placement");
          }
          button.setAttribute("title", currentTitle);
          button.classList.remove('code-copy-button-checked');
        }, 1000);
        // clear code selection
        e.clearSelection();
      }
      const getTextToCopy = function(trigger) {
          const codeEl = trigger.previousElementSibling.cloneNode(true);
          for (const childEl of codeEl.children) {
            if (isCodeAnnotation(childEl)) {
              childEl.remove();
            }
          }
          return codeEl.innerText;
      }
      const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
        text: getTextToCopy
      });
      clipboard.on('success', onCopySuccess);
      if (window.document.getElementById('quarto-embedded-source-code-modal')) {
        const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
          text: getTextToCopy,
          container: window.document.getElementById('quarto-embedded-source-code-modal')
        });
        clipboardModal.on('success', onCopySuccess);
      }
        var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
        var mailtoRegex = new RegExp(/^mailto:/);
          var filterRegex = new RegExp("https:\/\/jelmerp\.github\.io\/ENTMLGY6703-SP26");
        var isInternal = (href) => {
            return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
        }
        // Inspect non-navigation links and adorn them if external
     	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
        for (var i=0; i<links.length; i++) {
          const link = links[i];
          if (!isInternal(link.href)) {
            // undo the damage that might have been done by quarto-nav.js in the case of
            // links that we want to consider external
            if (link.dataset.originalHref !== undefined) {
              link.href = link.dataset.originalHref;
            }
          }
        }
      function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
        const config = {
          allowHTML: true,
          maxWidth: 500,
          delay: 100,
          arrow: false,
          appendTo: function(el) {
              return el.closest('section.slide') || el.parentElement;
          },
          interactive: true,
          interactiveBorder: 10,
          theme: 'light-border',
          placement: 'bottom-start',
        };
        if (contentFn) {
          config.content = contentFn;
        }
        if (onTriggerFn) {
          config.onTrigger = onTriggerFn;
        }
        if (onUntriggerFn) {
          config.onUntrigger = onUntriggerFn;
        }
          config['offset'] = [0,0];
          config['maxWidth'] = 700;
        window.tippy(el, config); 
      }
      const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
      for (var i=0; i<noterefs.length; i++) {
        const ref = noterefs[i];
        tippyHover(ref, function() {
          // use id or data attribute instead here
          let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
          try { href = new URL(href).hash; } catch {}
          const id = href.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note) {
            return note.innerHTML;
          } else {
            return "";
          }
        });
      }
      const findCites = (el) => {
        const parentEl = el.parentElement;
        if (parentEl) {
          const cites = parentEl.dataset.cites;
          if (cites) {
            return {
              el,
              cites: cites.split(' ')
            };
          } else {
            return findCites(el.parentElement)
          }
        } else {
          return undefined;
        }
      };
      var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
      for (var i=0; i<bibliorefs.length; i++) {
        const ref = bibliorefs[i];
        const citeInfo = findCites(ref);
        if (citeInfo) {
          tippyHover(citeInfo.el, function() {
            var popup = window.document.createElement('div');
            citeInfo.cites.forEach(function(cite) {
              var citeDiv = window.document.createElement('div');
              citeDiv.classList.add('hanging-indent');
              citeDiv.classList.add('csl-entry');
              var biblioDiv = window.document.getElementById('ref-' + cite);
              if (biblioDiv) {
                citeDiv.innerHTML = biblioDiv.innerHTML;
              }
              popup.appendChild(citeDiv);
            });
            return popup.innerHTML;
          });
        }
      }
    });
    </script>
    <script>var lightboxQuarto = GLightbox({"closeEffect":"zoom","descPosition":"bottom","loop":false,"openEffect":"zoom","selector":".lightbox"});
    (function() {
      let previousOnload = window.onload;
      window.onload = () => {
        if (previousOnload) {
          previousOnload();
        }
        lightboxQuarto.on('slide_before_load', (data) => {
          const { slideIndex, slideNode, slideConfig, player, trigger } = data;
          const href = trigger.getAttribute('href');
          if (href !== null) {
            const imgEl = window.document.querySelector(`a[href="${href}"] img`);
            if (imgEl !== null) {
              const srcAttr = imgEl.getAttribute("src");
              if (srcAttr && srcAttr.startsWith("data:")) {
                slideConfig.href = srcAttr;
              }
            }
          } 
        });
      
        lightboxQuarto.on('slide_after_load', (data) => {
          const { slideIndex, slideNode, slideConfig, player, trigger } = data;
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(slideNode);
          }
        });
      
      };
      
    })();
              </script>
    

</body></html>