---
title: "Transcriptomics with RNA-Seq"
author: Jelmer Poelstra
date: 2026-02-05
institute: "CFAES Bioinformatics Core, OSU"
format:
  revealjs:
    theme: [default, ../theme_mod.scss]
    css: ../slides.css
    auto-stretch: false
    smaller: true
    scrollable: false
    progress: true
    incremental: false
    slide-number: true
    slide-level: 2
    highlight-style: pygments
    #embed-resources: true
    chalkboard:
      theme: whiteboard  # Need to turn off embed-resources for this
    toc: true
    toc-depth: 1
    toc-title: Overview
    footer: "[~ ENTMLGY6703-SP26 ~](https://mcic-osu.github.io/ENTMLGY6703-SP26)"
    logo: ../img/CFAES-Horiz-RGBHEX.svg
    menu:
      width: wide
      numbers: true
editor_options: 
  chunk_output_type: console
bibliography: ../references.bib
---

# Why study gene expression? {background-color="silver"}

## Gene expression is highly dynamic

Unlike an organism's genome content, the expression of genes is highly dynamic,
varying:

- Across time and space

- Qualitatively (_which_ genes are expressed) and especially
  quantitatively (_how much_ of each gene is expressed)

## Gene expression is functionally important 

Considering:

::: incremental

1. That **protein production** tells us about the activity of biological functions,\
   and the molecular mechanisms underlying those functions

2. That it is easier to measure **transcript (mRNA)** than protein abundance

3. The central dogma

:::

![](../img/centraldogma3.png){fig-align="center" width="100%"}

. . .

... ***gene expression can be used as a proxy for protein expression***
to make functional inferences

## Studying gene expression

For example,
gene expression can be quantified to associate genes and genetic pathways with:

- Phenotypic responses e.g. to environmental conditions
- Differences between groups (treatments, genotypes, sexes, tissues, etc.)

. . .

<hr style="height:1pt; visibility:hidden;" />

So, studying gene expression can
**identify candidate genes underlying phenotypes --**\
these can next by functionally validated using experiments like knock-outs.

<br>

. . .

::: {.callout-tip appearance="simple"}
And further down the line,
understanding molecular basis of phenotypes has all sorts of uses!

E.g, in an agricultural context, to **manipulate** phenotypes such as resistance,
pathogenicity, and yield.
:::

## How to quantify gene expression?

- Techniques like **qPCR/ddPCR** can quantify expression of a few genes at a time
  - Useful for hypothesis-driven research on specific genes
  - Not suitable for discovery-based research on many genes

<hr style="height:1pt; visibility:hidden;" />

. . .

- With **transcriptomics**, expression of thousands of genes is quantified simultaneously
  - It was already common practice before HTS: with microarrays
  - But is now dominated by the HTS-based method **RNA-Seq**

# Introduction to RNA-Seq {background-color="silver"}

## What is RNA-Seq?

- To estimate gene expression levels genome-wide,
  RNA-Seq takes a brute-force approach by
  **randomly sequencing of millions of RNA fragments per sample**

- The resulting reads can be assigned a gene of origin,\
  and the core idea is that a gene's read count reflects that gene's expression level

![](../img/reads.png){fig-align="center" width="35%"}

. . .

<details><summary>
**How can we tell which gene each read originates from?**
&nbsp; <span style="color:darkolivegreen;">{{< fa comments >}}</span>
</summary>
Most commonly by aligning them to a reference genome.
</details>

:::footer
:::

## What is RNA-Seq?

- To estimate gene expression levels genome-wide,
  RNA-Seq takes a brute-force approach by
  **randomly sequencing of millions of RNA fragments per sample**

- The resulting reads can be assigned a gene of origin,\
  and the core idea is that a gene's read count reflects that gene's expression level

![](../img/reads.png){fig-align="center" width="35%"}

::: {.callout-note appearance="simple"}
RNA-Seq is a very widely used technique ‚Äî\
it constitutes *the* most common usage of high-throughput sequencing
:::

:::footer
:::

## The most common type of RNA-Seq

We'll focus on the most common type of RNA-Seq, which:

::: incremental
1. Does not sequence RNA directly, but first **reverse transcribes RNA to cDNA**
2. Attempts to sequence only mRNA, avoiding non-coding RNAs ("**mRNA-Seq**")
3. Does not distinguish between RNA from different cell types ("**bulk RNA-Seq**")
4. Uses **short reads** (‚â§150 bp) that do not cover full transcripts but do uniquely ID genes
5. Uses a reference genome for read alignment ("**reference-based RNA-Seq**")
:::

. . .

<hr style="height:1pt; visibility:hidden;" />

::: {.callout-note appearance="simple"}
_When I refer to RNA-Seq from now on, this kind of RNA-Seq is impled._
:::

## Side note: Types of RNA

![](../img/rna-types2.png){fig-align="center" width="83%"}

:::legend2
From [BioRender](https://biorender.com/)
:::

:::footer
:::

## RNA-Seq project examples

RNA-Seq is the most common data type I help analyze in my role.\
For a taste of what it's used for in CFAES and beyond ---\
the following projects aimed to identify genes & pathways differing between:

. . .

- Soybean cultivars in response to *Phytophtora sojae* inoculation
  _(Dorrance lab, PlantPath)_

- Response of maize to _Pantoea_ infection with and without an effector gene
  _(Mackey lab, HCS)_

. . .

- Mated and unmated mosquitos _(Sirot lab, College of Wooster)_

- Tissues of the ambrosia beetle and its symbiotic fungus _(Ranger lab, USDA)_

- Diapause-inducing conditions for two pest stink bug species _(Michel lab, Entomology)_

. . .

- Pig coronaviruses with vs. without an experimental insertion _(Wang lab, CFAH)_

- Human carcinoma cell lines with vs. without a manipulated gene _(Cruz lab, CCC)_

. . .

And to improve the annotation of a nematode genome assembly _(Taylor lab, PlantPath)_

# Stage I: Experimental design {background-color="silver"}

## Experimental design: treatments/groups

RNA-Seq typically compares groups of samples defined by differences in:

- **Treatments** --- e.g. different host plants, temperature, diet, mated/unmated

- **Organismal variants** --- e.g. ages/developmental stages, sexes, subspecies

- **Tissues**

## The Garrig√≥s et al. 2025 dataset

![](../img/garrigos_paper.png){.lightbox fig-align="center" width="75%" fig-alt="A screenshot of the paper's front matter."}

:::legend2
@garrig√≥s2025
:::

<br>

This paper used RNA-Seq data to study gene expression in *Culex pipiens*
mosquitos infected with malaria-causing *Plasmodium* protozoans

## Experimental design: treatments/groups

![](../img/exp-design_a.png){.lightbox fig-align="center" width="75%" fig-alt="A diagram showing the experimental design of the Garrig√≥s et al. 2025 study."}

:::footer
:::

## Experimental design: treatments/groups

![](../img/exp-design_b.png){.lightbox fig-align="center" width="75%" fig-alt="A diagram showing the experimental design of the Garrig√≥s et al. 2025 study."}

:::footer
:::

## Experimental design: treatments/groups

![](../img/exp-design_c.png){.lightbox fig-align="center" width="75%" fig-alt="A diagram showing the experimental design of the Garrig√≥s et al. 2025 study."}

:::footer
:::

## Experimental design: treatments/groups

<details><summary>
_With this experimental design,_
_would it be appropriate to sequence 9 samples total?_
&nbsp; <span style="color:darkolivegreen;">{{< fa comments >}}</span>
</summary>
No -- see the next slide
</details>

![](../img/exp-design_c.png){.lightbox fig-align="center" width="65%" fig-alt="A diagram showing the experimental design of the Garrig√≥s et al. 2025 study."}

:::footer
:::

## Biological replicates

To make statistically supported conclusions about expression differences,\
we need **biological replication** (at least 3-5 samples per group):

![](../img/exp-design_d.png){.lightbox fig-align="center" width="65%" fig-alt="A diagram showing the experimental design of the Garrig√≥s et al. 2025 study."}

:::footer
:::

# Stage II: From samples to reads {background-color="silver"}

## From samples to reads

![](../img/samples2reads_a2.png){fig-align="center" width="80%"}

<details><summary>
**Which type of RNA do the red bars followed by [AAAAAAAA]{style="color:red;"} represent?**
&nbsp; <span style="color:darkolivegreen;">{{< fa comments >}}</span>
</summary>
These are the mRNAs, which have poly-A tails (AAAA...)
</details>

## From samples to reads

![](../img/samples2reads_a2.png){fig-align="center" width="80%"}

We want mRNAs but these often make up only a few percent of RNAs!\
The two main ways to select for mRNAs are **poly-A selection** and **ribo-depletion**.

. . .

<hr style="height:1pt; visibility:hidden;" />

::: callout-note
#### As mentioned in the previous lecture:

- Library preparation is typically done by sequencing facilities / companies

- Many samples can be "multiplexed" into a single (RNA-Seq) library
:::

## From samples to reads

![](../img/samples2reads_b.png){fig-align="center" width="80%"}

:::legend2
Modified after <https://sydney-informatics-hub.github.io>
:::

:::footer
:::

## How much sequencing is needed?

. . .

- Guidelines highly approximate ‚Äî
  required amount depends not just on transcriptome size but also on expression
  level distribution, expression levels of genes of interest, etc.

. . .

- Typical recommendations are 20-50 million reads per sample in eukaryotes
  
# Stage III: From reads to counts {background-color="silver"}

## Overview of RNA-Seq data analysis

The analysis of RNA-Seq data can be divided into two main parts:

1. **From reads to counts**:
   from the raw reads, produce a count table
2. **From counts to conclusions**:
   analyze the count table to draw biological conclusions

## A quick primer on the count table

The **count table** has one row for each gene and one column for each sample,\
with the entries being the number of reads mapping to each gene in each sample:

. . .

<hr style="height:1pt; visibility:hidden;" />

|  | Sample 1 | Sample 2 | Sample 3 |
|---------|:----------:|:----------:|:----------:|
| **Gene A**  | 1500     | 2300     | 1800     |
| **Gene B**  | 0        | 5        | 2        |
| **Gene C**  | 300      | 250      | 400      |

: {.striped}

<br>

::: {.callout-tip appearance="simple"}
_Actual count tables have thousands of genes (rows) and usually dozens of_
_samples (columns)_
:::

. . .

<br>

**How do you get to such a count table?**

##

![](../img/alignment1.png){fig-align="center" width="82%" fig-alt="A diagram of RNA-Seq analysis steps." .lightbox}

::: footer
:::

##

![](../img/alignment2.png){fig-align="center" width="82%" fig-alt="A diagram of RNA-Seq analysis steps." .lightbox}

::: footer
:::

##

![](../img/alignment3.png){fig-align="center" width="82%" fig-alt="A diagram of RNA-Seq analysis steps." .lightbox}

::: footer
:::

##

![](../img/alignment4.png){fig-align="center" width="82%" fig-alt="A diagram of RNA-Seq analysis steps." .lightbox}

::: footer
:::

##

![](../img/alignment5.png){fig-align="center" width="82%" fig-alt="A diagram of RNA-Seq analysis steps." .lightbox}

::: footer
:::

## From reads to counts: overview

This is what the Adobe Firefly AI came up with when I tried to get
it to help me with producing the diagram on the previous slide üòµ‚Äçüí´

![](../img/firefly_diagram.png){fig-align="center" width="45%" fig-alt="Adobe Firefly's image when asking it create a diagram of RNA-Seq analysis steps." .lightbox}

::: footer
:::

## From reads to counts: overview

It specifically involves, at a minimum:

1. **Read preprocessing**

2. **Aligning reads to a reference genome**

3. **Quantifying expression levels**

<br>

. . .

::: {.callout-note appearance="simple"}
This part is "bioinformatics"-heavy, with large files, high computing needs,
and using command-line tools ran in the Unix shell.

If this is a problem:
the process if fairly standardized and suitable to be outsourced. 
:::

## Read pre-processing

Read pre-processing includes:

<hr style="height:1pt; visibility:hidden;" />

- **Checking the quantity and quality of your reads** (e.g. with `FastQC`)

<hr style="height:1pt; visibility:hidden;" />

- **Removing unwanted sequences**, such as:
    - Adapters, low-quality bases, and very short reads
    - rRNA-derived reads (*optional*)
    - Contaminant sequences (*optional*)

## Read alignment to a reference genome

<details><summary>
Consider what you know about how mRNAs are produced.
Does an RNA transcript always correspond to a contiguous stretch of the genome?
If not, what implications does this have for read alignment?
&nbsp; <span style="color:darkolivegreen;">{{< fa comments >}}</span>
</summary>

</details>

## Side note: Splicing

![](../img/splicing_biorender2a.png){fig-align="center" width="85%"}

:::legend2
From [BioRender](https://biorender.com/)
:::

## Side note: Splicing

![](../img/splicing_biorender2.png){fig-align="center" width="85%"}

:::legend2
From [BioRender](https://biorender.com/)
:::

## Read alignment to a reference genome

So, the alignment of reads to a reference genome needs to be "_splice-aware_":

<hr style="height:1pt; visibility:hidden;" />

![](../img/alignment_ed2.png){fig-align="center" width="70%"}

:::legend2
@vandenberge2019
:::

:::footer
:::

## Read alignment to a reference genome

Alternatively, you can align to the *transcriptome* (i.e., all mature transcripts):

![](../img/alignment_ed1.jpg){fig-align="center" width="70%"}

:::legend2
@vandenberge2019
:::

:::footer
:::

## Read alignment to a reference genome

<details><summary>Why are there multiple bars in panel **b**? What do these represent?
&nbsp; <span style="color:darkolivegreen;">{{< fa comments >}}</span>
</summary>
These represent different transcripts originating from the same gene due to
***alternative splicing***.
These will produce different proteins, which are called ***isoforms***.

Most short-read RNA-Seq studies do not attempt to distinguish between isoforms,\
but rather quantify expression at the gene level.

If distinguishing between isoforms is important,
performing RNA-Seq with a **long-read HTS** platform is a better option.
</details>

![](../img/alignment_ed1.jpg){fig-align="center" width="50%"}

:::legend2
@vandenberge2019
:::

:::footer
:::

## Gene-wise quantification

In essence, a simple counting exercise once you have the alignments in hand:\
for each sample, how many reads map to each gene?

![](../img/alignment_counting.png){fig-align="center" width="72%" fig-alt="A diagram of RNA-Seq analysis steps." .lightbox}

. . .

::: {.callout-note appearance="simple"}
Though in practice, a bit more complicated than this, due to e.g.:

- Reads that map to multiple genes ("multi-mapping reads")
- Sequencing biases and RNA fragmentation
:::

## A best-practice pipeline to produce counts

The "nf-core" initiative (<https://nf-co.re>, @ewels2020)
aims to produce best-practice and automated bioinformatics pipelines,
like for RNA-Seq (<https://nf-co.re/rnaseq>):

![](../img/nfcore_rnaseq.png){fig-align="center" width="100%"}

# Stage IV: From counts to conclusions {background-color="silver"}

## Count table analysis: overview

The second part of RNA-Seq data analysis involves analyzing the count table.

In contrast to the first part,
this can be done on a laptop and instead is heavier on\
**statistics**, **data visualization** and **biological interpretation**.

<br>

. . .

It is typically done with the R language, and common aspects include:

- **Principal Component Analysis (PCA)**\
  Assessing overall sample "clustering" (similarity) patterns

- **Differential Expression (DE) analysis**\
  Finding genes that differ in expression level between sample groups (DEGs)

- **Functional enrichment analysis**\
  See whether certain gene function categories are overrepresented among DEGs

## Principal Component Analysis (PCA)

PCA visualizes overall patterns of similarity among samples,\
such as whether groups of interest form distinct clusters:

<hr style="height:1pt; visibility:hidden;" />

![](../img/PCA.jpeg){fig-align="center" width="100%"}

:::legend2
Fig. 1 from @garrig√≥s2025
:::

## Differential expression (DE) analysis

Differential Expression (DE) analysis allows you to test,
separately for every expressed gene in your dataset,
whether it significantly differs in expression level between groups.

. . .

Typically, this is done with pairwise comparisons between groups:

::::: columns
::: column
![](../img/geneplot.png)
:::

::: column
:::
:::::

## Differential expression (DE) analysis

Differential Expression (DE) analysis allows you to test,
separately for every expressed gene in your dataset,
whether it significantly differs in expression level between groups.

Typically, this is done with pairwise comparisons between groups:

::::: columns
::: column
![](../img/geneplot.png)
:::

::: column
![](../img/volcano.png)
:::
:::::

## Statistical considerations for the DE analysis

1. Gene count normalization
2. Probability distribution of the gene count data
3. Multiple-testing correction

::: {.callout-note appearance="simple"}
We will talk about these during tomorrow's lab.

While not necessarily easy conceptually,
this is all fairly straightforward in practice:
specialized R packages like `DESeq2` take care of the details.
:::

## Functional enrichment: introduction

Lists of DEGs can be long: not always easy to make biological sense of.\
Functional **enrichment analyses** help with this.

Functional enrichment analyses check whether certain
functional **categories** of genes are *statistically overrepresented*
among up- and/or downregulated genes.

<br>

. . .

Several databases group genes into functional categories ---\
the two main ones used for enrichment analysis are:

- **Gene Ontology (GO)**
- **Kyoto Encyclopedia of Genes and Genomes (KEGG)**

## Functional enrichment: conceptually

A GO term called "**photosynthesis**" is
associated with a set of genes involved in photosynthesis.

- Say that 300 genes out of 30,000 genes in the genome have been annotated to
  this term = 1%

- Say that among your 50 out of your 500 DEGs
  are annotated to this term = 5%

. . .

The larger the difference between these percentages,
the stronger the indication that the function in question --here, photosynthesis--
is overrepresented among your DEGs.

## Functional enrichment: GO

![](../img/garrigo_go.jpeg){fig-align="center" width="60%"}

::: legend2
Fig. 4 from @garrig√≥s2025
:::

::: {.callout-note appearance="simple"}
GO terms have a *hierarchical* structure with more specific terms grouping into
more general terms
:::

:::footer
:::

## Functional enrichment: KEGG {.nostretch}

KEGG focuses on **pathways** for cellular and organismal functions whose genes can be drawn and connected in maps.

. . .

<br>

![](../img/kegg_example.jpg){fig-align="center" width="90%"}

::: legend2
@rodriguez2020

> *KEGG representation of up-regulated genes related to jasmonic acid (JA) signal*
> *transduction pathways (ko04075) in banana cv. Calcutta 4 after inoculation with*
> *Pseudocercospora fijiensis*.
> *Genes or chemicals up-regulated at any time point were highlighted in green.*

:::

# Questions? {background-color="silver"}

## Read alignment QC

- **Alignment rates**\
  What percentage of reads was successfully aligned? _(Should be \>80%)_

. . .

- **Alignment targets**\
  What percentages of aligned reads mapped to
  *exons* vs. *introns* vs. *intergenic regions*?

. . .

<hr style="height:1pt; visibility:hidden;" />

<details><summary>What might cause high intronic mapping rates?
&nbsp; <span style="color:darkolivegreen;">{{< fa comments >}}</span>
</summary>
An abundance of pre-mRNA versus mature-mRNA.
</details>

. . .

<details><summary>What might cause high intergenic mapping rates?
&nbsp; <span style="color:darkolivegreen;">{{< fa comments >}}</span>
</summary>
DNA contamination or poor genome assembly/annotation quality
</details>

## References

:::footer
:::
