---
title: "High-throughput sequencing and genomes"
pagetitle: "High-throughput sequencing"
author: Jelmer Poelstra
date: 2026-01-29
institute: "CFAES Bioinformatics Core, Ohio State University"
format:
  revealjs:
    auto-stretch: false
    theme: [default, ../theme_mod.scss]
    css: ../slides.css
    smaller: true
    scrollable: false
    progress: true
    incremental: false
    slide-number: true
    slide-level: 2
    highlight-style: pygments
    #embed-resources: true
    chalkboard:
      theme: whiteboard  # Need to turn off embed-resources for this
    toc: true
    toc-depth: 1
    toc-title: Overview
    footer: "[~ ENTMLGY6703-SP26 ~](https://mcic-osu.github.io/ENTMLGY6703-SP26)"
    logo: ../img/CFAES-Horiz-RGBHEX.svg
    menu:
      width: wide
      numbers: true
editor_options: 
  chunk_output_type: console
bibliography: ../references.bib
---

# Introduction to sequencing technologies {background-color="silver"}

## What do we mean by *sequencing*?

- Sequencing generally refers to
  **determining the sequence of DNA, RNA, or protein fragments**

<hr style="height:1pt; visibility:hidden;" />

- Most commonly, especially in the context of "high-throughput" sequencing,\
  it specifically refers to **DNA** sequencing

<hr style="height:1pt; visibility:hidden;" />

. . .

- Here, we'll **only focus on DNA sequencing**, keeping in mind that:

  - Protein sequencing involves completely different technologies

  - RNA can be, and usually is, sequenced as DNA as well

    <details><summary> How is that done and why?
    &nbsp; <span style="color:darkolivegreen;">{{< fa comments >}}</span>
    </summary>
    RNA is usually reverse transcribed to DNA (cDNA) prior to sequencing.
    
    While it is becoming more feasible to directly sequence RNA molecules,\
    RNA is an unstable molecule that is easily degraded and harder to sequence.
    </details>

## Overview of sequencing technologies

::: incremental

- **Sanger sequencing [*(since ~1985)*]{style="color:darkolivegreen"}** \
  Sequences a single, typically PCR-amplified, DNA fragment at a time

<hr style="height:1pt; visibility:hidden;" />

- **High-throughput sequencing (HTS)**\
  Sequences 10^5^-10^9^, often randomly selected,
  DNA fragments at a time â€” two types:

  - Short-read HTS: More accurate, shorter reads
    [***(since 2005)***]{style="color:darkolivegreen"}
  
  - Long-read HTS: Less accurate, longer reads
    [***(since 2011)***]{style="color:darkolivegreen"}

:::

. . .

<hr style="height:1pt; visibility:hidden;" />

::: {.callout-tip appearance=simple}
These sequenced fragments of DNA are usually called **reads**
:::

## Sanger sequencing

- Sanger sequencing almost always starts with **PCR amplification** of the
  target DNA region ---\
  as illustrated by Dr. Popp last week:

![](../img/popp_primers.png){fig-align="center" width="90%"}

. . .

- Therefore, to design primers,
  you must know something about the target sequence in advance ---
  this can be highly limiting
  
- The sequenced fragment can be up to about 800-1,000 bp

## Sanger sequencing

- Sequencing itself is performed by **synthesizing a new DNA strand** with
  fluorescently-labeled nucleotides, using a different color for each base
  (`A`, `C`, `G`, `T`)

- The final result is a chromatogram that can be "base-called":

![](../img/good.chromatogram.png){fig-alt="https://dnacore.mgh.harvard.edu/new-cgi-bin/site/pages/sequencing_pages/seq_troubleshooting.jsp" fig-align="center" width="100%"}

. . .

<hr style="height:1pt; visibility:hidden;" />

::: callout-warning
## The entire human genome was sequenced with Sanger technology!
How many basepairs is that? Want to guess how much it cost to do this?
:::

## Sequencing cost through time

![](../img/2022_Sequencing_cost_per_Human_Genome.jpg){fig-align="center" width="92%" fig-alt="A graph showing the cost of sequencing a human genome through time." .lightbox}

::: legend2
<https://www.genome.gov/about-genomics/fact-sheets/Sequencing-Human-Genome-cost>
:::

:::footer
:::

## Present-day Sanger applications

- Because HTS has _much_ higher throughput and is _much_ cheaper per base,\
  Sanger sequencing is now less widely used

- But it is not obsolete,
  in part because high throughput isn't always needed

##

![](../img/peanut-butter.jpg){fig-align="center" width="62%" fig-alt="An AI-generated image showing a giant peanut butter jar." .lightbox}

:::legend2
Image generated by Adobe Firefly
:::

::: footer
:::

## Present-day Sanger applications

Some present-day uses of Sanger sequencing include...

- Taxonomic identification of samples

- Examining genetic variation among individuals/populations

...where using one or few marker loci or candidate genes can be sufficient

# High-throughput sequencing (HTS) {background-color="silver"}

## Omics

Let's start with the big picture --
HTS data underlies several of these main "omics" approaches:

![](../img/fisher-omics-crop.png){fig-align="center" width="50%" fig-alt="A diagram showing the main omics data types." .lightbox}

::: legend2
Copyright [ThermoFisher](https://www.thermofisher.com/us/en/home/brands/thermo-scientific/molecular-biology/molecular-biology-learning-center/molecular-biology-resource-library/spotlight-articles/supporting-multi-omics-approaches.html)
:::

## The main omics data types

| Omics type       | Molecule type      |               |
|------------------|--------------------|-----------------------------|
| Genomics         | DNA                | |
| Epigenomics      | DNA modifications  | [**High-throughput sequencing (HTS)**]{style="color:white"} |
| Transcriptomics  | RNA                | |
| Proteomics       | Proteins           | |
| Metabolomics     | Metabolites        | |

: {.striped}

. . .

<br>

<details><summary>
What does the -omics suffix mean?
&nbsp; <i class="fa-solid fa-comments" style="color:darkolivegreen;"></i>
</summary>

The "omics" suffix indicates the involvement of large-scale datasets ---\
for example, "genomics" data typically spans much or all of the genome
</details>

## The main omics data types

| Omics type       | Molecule type      | Data mainly produced by              |
|------------------|--------------------|-----------------------------|
| Genomics         | DNA                | [**High-throughput sequencing (HTS)**]{style="color:#00A087FF"}  |
| Epigenomics      | DNA modifications  | [**High-throughput sequencing (HTS)**]{style="color:#00A087FF"}  |
| Transcriptomics  | RNA                | [**High-throughput sequencing (HTS)**]{style="color:#00A087FF"}  |
| Proteomics       | Proteins           | [**Mass Spectrometry**]{style="color:#3C5488FF"}       |
| Metabolomics     | Metabolites        | [**Mass Spectrometry**]{style="color:#3C5488FF"}       |

: {.striped}

## Examples of HTS applications

- **Whole-genome assembly** --- for producing reference genomes

. . .

- **Typing of SNPs and other sequence variants**
  --- for population genetics, GWAS, etc.

. . .

- **Metabarcoding** --- for microbial community characterization

. . .

- **RNA-Seq** --- for large-scale gene expression analysis

. . .

<br>

<details><summary>
***Does any of you work on, or is planning to work on, projects like these?***
&nbsp; <i class="fa-solid fa-comments" style="color:darkolivegreen;"></i>
</summary></details>

## The main HTS technologies

| | Short-read HTS   | Long-read HTS
|---|-------|-------|
[**Main companies**]{style="color:white"}  | [Illumina]{style="color:white"}                | [Oxford Nanopore Technologies (ONT) & Pacific Biosciences (PacBio)]{style="color:white"}

## The main HTS technologies

| | Short-read HTS   | Long-read HTS
|---|-------|-------|
**Usage**           | More                    | Less (but increasing)
**Main companies**  | Illumina                | Oxford Nanopore Technologies (ONT) & Pacific Biosciences (PacBio)
**Timeline**        | Since 2005 â€” [technology fairly stable]{style="color:crimson"} | Since 2011 â€” [still rapid development]{style="color:green"}

: {.striped}

## The main HTS technologies

|                   | Short-read HTS          | Long-read HTS
|---|-------|-------|
**Usage**           | More                    | Less (but increasing)
**Main companies**  | Illumina                | Oxford Nanopore Technologies (ONT) & Pacific Biosciences (PacBio)
**Timeline**        | Since 2005 â€” [technology fairly stable]{style="color:crimson"} | Since 2011 â€” [still rapid development]{style="color:green"}
**Read lengths**    | [50-300 bp]{style="color:crimson"}               | [10-100+ kbp]{style="color:green"}
**Error rates**     | [Mostly <0.1%]{style="color:green"}            | [1-10% (ONT) / <0.1-10% (PacBio)]{style="color:crimson"}

: {.striped}

## The main HTS technologies

|                   | Short-read HTS          | Long-read HTS
|---|-------|-------|
**Usage**           | More                    | Less (but increasing)
**Main companies**  | Illumina                | Oxford Nanopore Technologies (ONT) & Pacific Biosciences (PacBio)
**Timeline**        | Since 2005 â€” [technology fairly stable]{style="color:crimson"} | Since 2011 â€” [still rapid development]{style="color:green"}
**Read lengths**    | [50-300 bp]{style="color:crimson"}               | [10-100+ kbp]{style="color:green"}
**Error rates**     | [Mostly <0.1%]{style="color:green"}            | [1-10% (ONT) / <0.1-10% (PacBio)]{style="color:crimson"}
**Throughput**    | [Higher]{style="color:green"} | [Lower]{style="color:crimson"}
**Cost per base**    | [Lower]{style="color:green"} | [Higher]{style="color:crimson"}

: {.striped}

## Read lengths

<details><summary>
**Can you think of applications where long reads are useful?**
&nbsp; <i class="fa-solid fa-comments" style="color:darkolivegreen;"></i>
</summary>
For example:

- Genome assembly
- Read-based taxonomic identification
</details>

. . .

<hr style="height:1pt; visibility:hidden;" />

<details><summary>
**Can you think of applications where read length may not matter much?**
&nbsp; <i class="fa-solid fa-comments" style="color:darkolivegreen;"></i>
</summary>
For example:

- (SNP) variant analysis
- "Counting applications" such as RNA-Seq

For these, genomic locations (variant analysis) or gene identities (RNA-Seq)
can be reliably inferred from as little as 25 bp
  
</details>

## Error rates

A read's sequence may differ from the actual DNA sequence it originated from:

- The read can have base-calling errors, missing bases, or extra bases

. . .

- When the base calling software is not confident,
  it can also return `N`s (= undetermined)

![](../img/noisy.sequence.png){fig-align="center" width="85%" fig-alt="A chromatogram with several uncalled bases." .lightbox}

. . .

::: {.callout-note appearance="simple"}
When you receive HTS reads, base calls have typically been made already.\
Every base call is accompanied by a **quality score**,
representing the estimated error probability.
:::

## Correcting sequencing errors

To overcome sequencing errors, every base can be sequenced multiple times --\
i.e., obtaining a "**depth of coverage**" greater than 1:

![](../img/coverage2.jpg){fig-align="center" width="60%" fig-alt="A diagram illustraing the concept of depth of coverage." .lightbox}

. . .

<details><summary>
**Which natural phenomenon might complicate this effort?**
&nbsp; <i class="fa-solid fa-comments" style="color:darkolivegreen;"></i>
</summary>
**Genetic variation** among and (for diploid organisms) within individuals
</details>

<hr style="height:1pt; visibility:hidden;" />

. . .

::: {.callout-tip appearance="simple"}
_Typical depths of coverage: ~50-100x for genome assembly & 10-30x for variant typing (!)_
:::

# Short-read HTS {background-color="silver"}

## Libraries and library prep

- In a HTS context,
  a "**library**" is a collection of DNA fragments ready for sequencing

. . .

- These fragments can number in the **millions or billions** and are often
  randomly generated from input like genomic DNA:

::: {style="margin-bottom: 0px;"}
![](../img/library-prep_cropped.png){fig-align="center" fig-alt="A diagram showing the main Illumina library preparation steps." .lightbox}
:::

::: legend2
An overview of the **library prep** procedure.
This is typically done for you by a sequencing facility or company.
:::

## Libraries and library prep

- After library prep,
  each DNA fragment is flanked by several types of short sequences that
  together make up the "**adapters**":

<br>

![](../img/adapters2.png){fig-align="center"}

. . .

<hr style="height:1pt; visibility:hidden;" />

::: callout-note
#### Multiplexing!
Adapters can include "indices" or "barcodes" to identify individual samples,
so many samples can be combined (_multiplexed_) into a single library
::::::

## Paired-end vs. single-end sequencing

- Fragments can be sequenced from both ends as shown below â€”\
  this is called **"paired-end" (PE) sequencing**:

![](../img/pairedend.png){fig-align="center" fig-alt="A diagram showing forward and reverse reads in paired-end sequencing." .lightbox}

. . .

- When sequencing is instead **single-end (SE)**, no reverse read is produced:

![](../img/singleend.png){fig-align="center"}

## Fragment size variation

- DNA fragment size varies --
  by design and because of limited precision in size selection

![](../img/pairedend.png){width="75%" fig-align="center" fig-alt="A diagram showing forward and reverse reads in paired-end sequencing." .lightbox}

<details><summary>
**What happens when a fragment is shorter than the length of a single F or R read?**
&nbsp; <i class="fa-solid fa-comments" style="color:darkolivegreen;"></i>
</summary>

"**Adapter read-through**": the final bases in the resulting reads will consist of
adapter sequence (these should be removed before downstream analysis)

![](../img/insert_size_adapter-read-through.png){fig-align="center" width="52%" fig-alt="A diagram illustrating the scenario when the DNA fragment is shorter than the single read length" .lightbox}

</details>

::: footer
:::

## Fragment size variation

- DNA fragment size varies --
  by design and because of limited precision in size selection

![](../img/pairedend.png){width="75%" fig-align="center" fig-alt="A diagram showing forward and reverse reads in paired-end sequencing." .lightbox}

<details><summary>
**What happens when a fragment is shorter than the combined F + R read length?**
&nbsp; <i class="fa-solid fa-comments" style="color:darkolivegreen;"></i>
</summary>

**Overlapping reads** (this can be useful!):

![](../img/insert_size_overlap.png){fig-align="center" width="60%" fig-alt="A diagram illustrating the scenario when the DNA fragment is shorter than the combined read length" .lightbox}

</details>

::: footer
:::

## How Illumina sequencing works

We start with fragments that have been attached to a flow cell ---\
this image shows just two, but millions are present simultaneously

![](../img/illumina_cycle0.png){width="100%" fig-align="center"}

## How Illumina sequencing works

Sequencing is performed by synthesizing a new strand using
fluorescently-labeled bases,\
and taking a picture each time a new nucleotide is incorporated:

![](../img/illumina_cycle1.png){width="100%" fig-align="center"}

## How Illumina sequencing works

Sequencing is performed by synthesizing a new strand using
fluorescently-labeled bases,\
and taking a picture each time a new nucleotide is incorporated:

![](../img/illumina_cycle2.png){width="100%" fig-align="center"}

## How Illumina sequencing works

Sequencing is performed by synthesizing a new strand using
fluorescently-labeled bases,\
and taking a picture each time a new nucleotide is incorporated:

![](../img/illumina_cycle5.png){width="100%" fig-align="center"}

## The scale of Illumina sequencing

![](../img/illumina_scale1.png){width="95%" fig-align="center"}

## The scale of Illumina sequencing

![](../img/illumina_scale2.png){width="95%" fig-align="center"}

## The scale of Illumina sequencing

![](../img/illumina_scale3.png){width="95%" fig-align="center"}

## The scale of Illumina sequencing

![](../img/illumina_scale4.png){width="95%" fig-align="center"}

## Video of Illumina technology

{{< video https://www.youtube.com/watch?v=fCd6B5HRaZ8 width="95%" height="81%" >}}

# Reference genomes

## Reference genomes

Many HTS applications either **require a "reference genome"** or
**involve its production**:

. . .

![](../img/reads2genome.png){fig-align="center" width="100%" fig-alt="A diagram illustrating how sequencing reads are aligned to a reference genome." .lightbox}

## Reference genomes

Many HTS applications either **require a "reference genome"** or
**involve its production:**

![](../img/reads2genome_2.png){fig-align="center" width="100%" fig-alt="A diagram illustrating how sequencing reads are aligned to a reference genome." .lightbox}

## Reference genomes

What exactly does reference genome refer to? It usually includes:

. . .

- **An assembly**\
  A representation of most or all of the genome DNA sequence: the genome assembly

- **An annotation**\
  Provides e.g. *locations of genes and other genomic "features"*
  in the corresponding genome assembly, and functional information for these features

. . .

<hr style="height:1pt; visibility:hidden;" />

:::callout-tip
#### Taxonomic identity

Reference genomes are typically applicable at the **species level**.
For example, if you work with maize, you want a _Zea mays_ reference genome. But:

- If needed, it's often possible to work with genomes of closely related species
- Conversely, different subspecies/lines may have their own reference genomes
:::

## There is enormous variation in genome size

[![](../img/Genome_Sizes.png){fig-align="center" width="73%"}](https://en.wikipedia.org/wiki/Genome_size)

:::legend2
<https://en.wikipedia.org/wiki/Genome_size>
:::

:::footer
:::

## And enormous growth of genomes in databases

![](../img/genomes_genbank.png){width="83%" fig-align="center"}

::: legend2
<br>
@konkel2023
:::

## Genome assemblies

- Besides being produced at higher rates as shown on the previous slide,\
  assemblies keep getting better with increasing usage & quality of long-read HTS

. . .

- Still, many assemblies instead consist of --often 1000s of-- fragments
  (contigs and scaffolds)

<hr style="height:1pt; visibility:hidden;" />

. . .

::: callout-note
## How is this data stored?
Both genome assemblies and annotations are typically saved in a single text file each â€”
we'll explore these in tomorrow's lab
:::

## Recap

You've learned...

1. That high-throughput sequencing (HTS) enables DNA sequencing at _much_ larger
   scales than was previously possible
   
2. That HTS underlies several major "omics" fields

2. How short-read and long-read HTS have different strengths and weaknesses

3. About libraries and the technology underlying short-read sequencing

. . .

4. That reference genomes are essential for many HTS applications

# Looking forward {background-color="silver"}

## The GarrigÃ³s et al. 2025 dataset

The labs this and next week are organized around the data set from @garrigÃ³s2025:

![](../img/garrigos_paper.png){fig-align="center" width="75%" fig-alt="A screenshot of the paper's front matter." .lightbox}

This paper uses **paired-end Illumina RNA-Seq data** to study gene expression in
*Culex pipiens* mosquitos infected with two different malaria-causing *Plasmodium*
protozoans.

## Tomorrow's lab

- You will first learn about what kind of computing environment is commonly
  used for analyzing high-throughput sequencing (HTS) data

. . .

- Then, you'll work in this environment to explore an HTS dataset,
  checking out HTS read and reference genome files,
  and performing read quality control

## Next week's content

- In the lecture,
  you will learn how RNA-Seq is used to study gene expression at scale

. . .

- In the lab, you will perform differential gene expression analysis
  on the *Culex pipiens* RNA-Seq dataset

# Bonus: More on HTS applications

## Examples of HTS applications

We talked about these HTS applications:

- **Whole-genome assembly** --- for producing reference genomes
- **Typing of SNPs and other sequence variants**
  --- for population genetics, GWAS, etc.
- **Metabarcoding** --- for microbial community characterization
- **RNA-Seq** --- for large-scale gene expression analysis

<hr style="height:1pt; visibility:hidden;" />

. . .

In each of of these applications:

- What part(s) of the genome are you sequencing?
- Who are you sequencing (how many individuals/species)?
- How are the sequences used?

## Examples of HTS applications

::: {style="font-size: 85%;"}

| Appl. | What| Who | How
|---|-----|-----|-----|
| **Whole-genome assembly** | The whole genome! ðŸ˜ƒ | A single individual | "_Overlap_" reads into larger fragments |
| **Variant typing** |  | | | |
| **16S Metabarcoding**  | | | |
| **RNA-Seq**  | | | |

: {.striped}
:::

## Examples of HTS applications

::: {style="font-size: 85%;"}

| Appl. | What| Who | How
|---|-----|-----|-----|
| **Whole-genome assembly** | Whole genome ðŸ˜ƒ | A single individual | "_Overlap_" reads into larger fragments |
| **Variant typing** | Whole genome <br>or specific (e.g. exome)<br> or random (GBS/RAD) regions | Multiple individuals | For each site & individual, determine the _variant state_ (allele)
| **16S Metabarcoding**  | | | |
| **RNA-Seq**  | | | |

: {.striped}
:::

## Examples of HTS applications

::: {style="font-size: 85%;"}

| Appl. | What| Who | How
|---|-----|-----|-----|
| **Whole-genome assembly** | Whole genome ðŸ˜ƒ | A single individual | "_Overlap_" reads into larger fragments |
| **Variant typing** | Whole genome <br>or specific (e.g. exome)<br> or random (GBS/RAD) regions | Multiple individuals | For each site & individual, determine the _variant state_ (allele)
| **16S Metabarcoding** | A specific locus (e.g. 16S) | Multi-species samples ---<br>soil, gut contents, etc. | Assign reads a _taxonomic_ identity and ***count*** |
| **RNA-Seq**  | | | |

: {.striped}
:::

## Examples of HTS applications

::: {style="font-size: 85%;"}

| Appl. | What| Who | How
|---|-----|-----|-----|
| **Whole-genome assembly** | Whole genome ðŸ˜ƒ | A single individual | "_Overlap_" reads into larger fragments |
| **Variant typing** | Whole genome <br>or specific (e.g. exome)<br> or random (GBS/RAD) regions | Multiple individuals | For each site & individual, determine the _variant state_ (allele)
| **16S Metabarcoding** | A specific locus (e.g. 16S) | Multi-species samples ---<br>soil, gut contents, etc. | Assign reads a _taxonomic_ identity and ***count*** |
| **RNA-Seq** | Whole transcriptome | Multiple indivuals<br>\(& multiple conditions) | Assign reads a _gene_ identity and ***count*** |

: {.striped}
:::

## References
