<!DOCTYPE html>
<html lang="en"><head>
<link href="../img/favicon.png" rel="icon" type="image/png">
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-html/tabby.min.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/light-border.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting-1092c56c6cadf2eb47b1bc8063ab382a.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<link href="../site_libs/quarto-contrib/fontawesome6-6.7.2/all.min.css" rel="stylesheet">
<link href="../site_libs/quarto-contrib/fontawesome6-6.7.2/latex-fontsize.css" rel="stylesheet">
<script src="../site_libs/quarto-contrib/glightbox/glightbox.min.js"></script>
<link href="../site_libs/quarto-contrib/glightbox/glightbox.min.css" rel="stylesheet">
<link href="../site_libs/quarto-contrib/glightbox/lightbox.css" rel="stylesheet"><meta charset="utf-8">
  <meta name="generator" content="quarto-1.6.40">

  <meta name="author" content="Jelmer Poelstra">
  <meta name="dcterms.date" content="2026-01-29">
  <title>ENTMLGY6703-SP26 - Jelmer's lectures ‚Äì High-throughput sequencing</title>
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no, minimal-ui">
  <link rel="stylesheet" href="../site_libs/revealjs/dist/reset.css">
  <link rel="stylesheet" href="../site_libs/revealjs/dist/reveal.css">
  <style>
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    div.columns{display: flex; gap: min(4vw, 1.5em);}
    div.column{flex: auto; overflow-x: auto;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
    ul.task-list li input[type="checkbox"] {
      width: 0.8em;
      margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
      vertical-align: middle;
    }
    /* CSS for citations */
    div.csl-bib-body { }
    div.csl-entry {
      clear: both;
      margin-bottom: 0em;
    }
    .hanging-indent div.csl-entry {
      margin-left:2em;
      text-indent:-2em;
    }
    div.csl-left-margin {
      min-width:2em;
      float:left;
    }
    div.csl-right-inline {
      margin-left:2em;
      padding-left:1em;
    }
    div.csl-indent {
      margin-left: 2em;
    }  </style>
  <link rel="stylesheet" href="../site_libs/revealjs/dist/theme/quarto-fb88a47fb37db585264132e6c4198fc7.css">
  <link rel="stylesheet" href="../slides.css">
  <link href="../site_libs/revealjs/plugin/quarto-line-highlight/line-highlight.css" rel="stylesheet">
  <link href="../site_libs/revealjs/plugin/reveal-menu/menu.css" rel="stylesheet">
  <link href="../site_libs/revealjs/plugin/reveal-menu/quarto-menu.css" rel="stylesheet">
  <link href="../site_libs/revealjs/plugin/reveal-chalkboard/font-awesome/css/all.css" rel="stylesheet">
  <link href="../site_libs/revealjs/plugin/reveal-chalkboard/style.css" rel="stylesheet">
  <link href="../site_libs/revealjs/plugin/quarto-support/footer.css" rel="stylesheet">
  <style type="text/css">
    .reveal div.sourceCode {
      margin: 0;
      overflow: auto;
    }
    .reveal div.hanging-indent {
      margin-left: 1em;
      text-indent: -1em;
    }
    .reveal .slide:not(.center) {
      height: 100%;
    }
    .reveal .slide.scrollable {
      overflow-y: auto;
    }
    .reveal .footnotes {
      height: 100%;
      overflow-y: auto;
    }
    .reveal .slide .absolute {
      position: absolute;
      display: block;
    }
    .reveal .footnotes ol {
      counter-reset: ol;
      list-style-type: none; 
      margin-left: 0;
    }
    .reveal .footnotes ol li:before {
      counter-increment: ol;
      content: counter(ol) ". "; 
    }
    .reveal .footnotes ol li > p:first-child {
      display: inline-block;
    }
    .reveal .slide ul,
    .reveal .slide ol {
      margin-bottom: 0.5em;
    }
    .reveal .slide ul li,
    .reveal .slide ol li {
      margin-top: 0.4em;
      margin-bottom: 0.2em;
    }
    .reveal .slide ul[role="tablist"] li {
      margin-bottom: 0;
    }
    .reveal .slide ul li > *:first-child,
    .reveal .slide ol li > *:first-child {
      margin-block-start: 0;
    }
    .reveal .slide ul li > *:last-child,
    .reveal .slide ol li > *:last-child {
      margin-block-end: 0;
    }
    .reveal .slide .columns:nth-child(3) {
      margin-block-start: 0.8em;
    }
    .reveal blockquote {
      box-shadow: none;
    }
    .reveal .tippy-content>* {
      margin-top: 0.2em;
      margin-bottom: 0.7em;
    }
    .reveal .tippy-content>*:last-child {
      margin-bottom: 0.2em;
    }
    .reveal .slide > img.stretch.quarto-figure-center,
    .reveal .slide > img.r-stretch.quarto-figure-center {
      display: block;
      margin-left: auto;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-left,
    .reveal .slide > img.r-stretch.quarto-figure-left  {
      display: block;
      margin-left: 0;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-right,
    .reveal .slide > img.r-stretch.quarto-figure-right  {
      display: block;
      margin-left: auto;
      margin-right: 0; 
    }
  </style>
</head>
<body class="quarto-light">
  <div class="reveal">
    <div class="slides">

<section id="title-slide" class="quarto-title-block center">
  <h1 class="title">High-throughput sequencing and genomes</h1>

<div class="quarto-title-authors">
<div class="quarto-title-author">
<div class="quarto-title-author-name">
Jelmer Poelstra 
</div>
        <p class="quarto-title-affiliation">
            CFAES Bioinformatics Core, Ohio State University
          </p>
    </div>
</div>

  <p class="date">2026-01-29</p>
</section><section id="TOC">
<nav role="doc-toc"> 
<h2 id="toc-title">Overview</h2>
<ul>
<li><a href="#/introduction-to-sequencing-technologies" id="/toc-introduction-to-sequencing-technologies">Introduction to sequencing technologies</a></li>
<li><a href="#/high-throughput-sequencing-hts" id="/toc-high-throughput-sequencing-hts">High-throughput sequencing (HTS)</a></li>
<li><a href="#/short-read-hts" id="/toc-short-read-hts">Short-read HTS</a></li>
<li><a href="#/reference-genomes" id="/toc-reference-genomes">Reference genomes</a></li>
<li><a href="#/looking-forward" id="/toc-looking-forward">Looking forward</a></li>
<li><a href="#/bonus-more-on-hts-applications" id="/toc-bonus-more-on-hts-applications">Bonus: More on HTS applications</a></li>
</ul>
</nav>
</section>
<section>
<section id="introduction-to-sequencing-technologies" class="title-slide slide level1 center" data-background-color="silver">
<h1>Introduction to sequencing technologies</h1>

</section>
<section id="what-do-we-mean-by-sequencing" class="slide level2">
<h2>What do we mean by <em>sequencing</em>?</h2>
<ul>
<li>Sequencing generally refers to <strong>determining the sequence of DNA, RNA, or protein fragments</strong></li>
</ul>
<hr style="height:1pt; visibility:hidden;">
<ul>
<li>Most commonly, especially in the context of ‚Äúhigh-throughput‚Äù sequencing,<br>
it specifically refers to <strong>DNA</strong> sequencing</li>
</ul>
<hr style="height:1pt; visibility:hidden;">
<div class="fragment">
<ul>
<li><p>Here, we‚Äôll <strong>only focus on DNA sequencing</strong>, keeping in mind that:</p>
<ul>
<li><p>Protein sequencing involves completely different technologies</p></li>
<li><p>RNA can be, and usually is, sequenced as DNA as well</p>
<details>
<summary>
<p>How is that done and why? &nbsp; <span style="color:darkolivegreen;"><i class="fa-solid fa-comments" aria-label="comments"></i></span></p>
</summary>
<p>RNA is usually reverse transcribed to DNA (cDNA) prior to sequencing.</p>
<p>While it is becoming more feasible to directly sequence RNA molecules,<br>
RNA is an unstable molecule that is easily degraded and harder to sequence.</p>
</details></li>
</ul></li>
</ul>
</div>
</section>
<section id="overview-of-sequencing-technologies" class="slide level2">
<h2>Overview of sequencing technologies</h2>
<div>
<ul>
<li class="fragment"><strong>Sanger sequencing <span style="color:darkolivegreen"><em>(since ~1985)</em></span></strong><br>
Sequences a single, typically PCR-amplified, DNA fragment at a time</li>
</ul>
<hr style="height:1pt; visibility:hidden;">
<ul>
<li class="fragment"><p><strong>High-throughput sequencing (HTS)</strong><br>
Sequences 10<sup>5</sup>-10<sup>9</sup>, often randomly selected, DNA fragments at a time ‚Äî two types:</p>
<ul>
<li class="fragment"><p>Short-read HTS: More accurate, shorter reads <span style="color:darkolivegreen"><strong><em>(since 2005)</em></strong></span></p></li>
<li class="fragment"><p>Long-read HTS: Less accurate, longer reads <span style="color:darkolivegreen"><strong><em>(since 2011)</em></strong></span></p></li>
</ul></li>
</ul>
</div>
<div class="fragment">
<hr style="height:1pt; visibility:hidden;">
<div class="callout callout-tip callout-style-simple">
<div class="callout-body">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-content">
<p>These sequenced fragments of DNA are usually called <strong>reads</strong></p>
</div>
</div>
</div>
</div>
</section>
<section id="sanger-sequencing" class="slide level2">
<h2>Sanger sequencing</h2>
<ul>
<li>Sanger sequencing almost always starts with <strong>PCR amplification</strong> of the target DNA region ‚Äî<br>
as illustrated by Dr.&nbsp;Popp last week:</li>
</ul>
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="../img/popp_primers.png" class="quarto-figure quarto-figure-center" style="width:90.0%"></p>
</figure>
</div>
<div class="fragment">
<ul>
<li><p>Therefore, to design primers, you must know something about the target sequence in advance ‚Äî this can be highly limiting</p></li>
<li><p>The sequenced fragment can be up to about 800-1,000 bp</p></li>
</ul>
</div>
</section>
<section id="sanger-sequencing-1" class="slide level2">
<h2>Sanger sequencing</h2>
<ul>
<li><p>Sequencing itself is performed by <strong>synthesizing a new DNA strand</strong> with fluorescently-labeled nucleotides, using a different color for each base (<code>A</code>, <code>C</code>, <code>G</code>, <code>T</code>)</p></li>
<li><p>The final result is a chromatogram that can be ‚Äúbase-called‚Äù:</p></li>
</ul>
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="../img/good.chromatogram.png" class="quarto-figure quarto-figure-center" style="width:100.0%" alt="https://dnacore.mgh.harvard.edu/new-cgi-bin/site/pages/sequencing_pages/seq_troubleshooting.jsp"></p>
</figure>
</div>
<div class="fragment">
<hr style="height:1pt; visibility:hidden;">
<div class="callout callout-warning callout-titled callout-style-default">
<div class="callout-body">
<div class="callout-title">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<p><strong>The entire human genome was sequenced with Sanger technology!</strong></p>
</div>
<div class="callout-content">
<p>How many basepairs is that? Want to guess how much it cost to do this?</p>
</div>
</div>
</div>
</div>
</section>
<section id="sequencing-cost-through-time" class="slide level2">
<h2>Sequencing cost through time</h2>
<div class="quarto-figure quarto-figure-center">
<figure>
<p><a href="../img/2022_Sequencing_cost_per_Human_Genome.jpg" class="lightbox" data-gallery="quarto-lightbox-gallery-1"><img data-src="../img/2022_Sequencing_cost_per_Human_Genome.jpg" class="quarto-figure quarto-figure-center" style="width:92.0%" alt="A graph showing the cost of sequencing a human genome through time."></a></p>
</figure>
</div>
<div class="legend2">
<p><a href="https://www.genome.gov/about-genomics/fact-sheets/Sequencing-Human-Genome-cost" class="uri">https://www.genome.gov/about-genomics/fact-sheets/Sequencing-Human-Genome-cost</a></p>
</div>
<div class="footer">

</div>
</section>
<section id="present-day-sanger-applications" class="slide level2">
<h2>Present-day Sanger applications</h2>
<ul>
<li><p>Because HTS has <em>much</em> higher throughput and is <em>much</em> cheaper per base,<br>
Sanger sequencing is now less widely used</p></li>
<li><p>But it is not obsolete, in part because high throughput isn‚Äôt always needed</p></li>
</ul>
</section>
<section id="section" class="slide level2">
<h2></h2>
<div class="quarto-figure quarto-figure-center">
<figure>
<p><a href="../img/peanut-butter.jpg" class="lightbox" data-gallery="quarto-lightbox-gallery-2"><img data-src="../img/peanut-butter.jpg" class="quarto-figure quarto-figure-center" style="width:62.0%" alt="An AI-generated image showing a giant peanut butter jar."></a></p>
</figure>
</div>
<div class="legend2">
<p>Image generated by Adobe Firefly</p>
</div>
<div class="footer">

</div>
</section>
<section id="present-day-sanger-applications-1" class="slide level2">
<h2>Present-day Sanger applications</h2>
<p>Some present-day uses of Sanger sequencing include‚Ä¶</p>
<ul>
<li><p>Taxonomic identification of samples</p></li>
<li><p>Examining genetic variation among individuals/populations</p></li>
</ul>
<p>‚Ä¶where using one or few marker loci or candidate genes can be sufficient</p>
</section></section>
<section>
<section id="high-throughput-sequencing-hts" class="title-slide slide level1 center" data-background-color="silver">
<h1>High-throughput sequencing (HTS)</h1>

</section>
<section id="omics" class="slide level2">
<h2>Omics</h2>
<p>Let‚Äôs start with the big picture ‚Äì HTS data underlies several of these main ‚Äúomics‚Äù approaches:</p>
<div class="quarto-figure quarto-figure-center">
<figure>
<p><a href="../img/fisher-omics-crop.png" class="lightbox" data-gallery="quarto-lightbox-gallery-3"><img data-src="../img/fisher-omics-crop.png" class="quarto-figure quarto-figure-center" style="width:50.0%" alt="A diagram showing the main omics data types."></a></p>
</figure>
</div>
<div class="legend2">
<p>Copyright <a href="https://www.thermofisher.com/us/en/home/brands/thermo-scientific/molecular-biology/molecular-biology-learning-center/molecular-biology-resource-library/spotlight-articles/supporting-multi-omics-approaches.html">ThermoFisher</a></p>
</div>
</section>
<section id="the-main-omics-data-types" class="slide level2">
<h2>The main omics data types</h2>
<table class="table-striped caption-top">
<colgroup>
<col style="width: 26%">
<col style="width: 29%">
<col style="width: 43%">
</colgroup>
<thead>
<tr class="header">
<th>Omics type</th>
<th>Molecule type</th>
<th></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Genomics</td>
<td>DNA</td>
<td></td>
</tr>
<tr class="even">
<td>Epigenomics</td>
<td>DNA modifications</td>
<td><span style="color:white"><strong>High-throughput sequencing (HTS)</strong></span></td>
</tr>
<tr class="odd">
<td>Transcriptomics</td>
<td>RNA</td>
<td></td>
</tr>
<tr class="even">
<td>Proteomics</td>
<td>Proteins</td>
<td></td>
</tr>
<tr class="odd">
<td>Metabolomics</td>
<td>Metabolites</td>
<td></td>
</tr>
</tbody>
</table>
<div class="fragment">
<p><br></p>
<details>
<summary>
What does the -omics suffix mean? &nbsp; <i class="fa-solid fa-comments" style="color:darkolivegreen;"></i>
</summary>
The ‚Äúomics‚Äù suffix indicates the involvement of large-scale datasets ‚Äî<br>
for example, ‚Äúgenomics‚Äù data typically spans much or all of the genome
</details>
</div>
</section>
<section id="the-main-omics-data-types-1" class="slide level2">
<h2>The main omics data types</h2>
<table class="table-striped caption-top">
<colgroup>
<col style="width: 26%">
<col style="width: 29%">
<col style="width: 43%">
</colgroup>
<thead>
<tr class="header">
<th>Omics type</th>
<th>Molecule type</th>
<th>Data mainly produced by</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Genomics</td>
<td>DNA</td>
<td><span style="color:#00A087FF"><strong>High-throughput sequencing (HTS)</strong></span></td>
</tr>
<tr class="even">
<td>Epigenomics</td>
<td>DNA modifications</td>
<td><span style="color:#00A087FF"><strong>High-throughput sequencing (HTS)</strong></span></td>
</tr>
<tr class="odd">
<td>Transcriptomics</td>
<td>RNA</td>
<td><span style="color:#00A087FF"><strong>High-throughput sequencing (HTS)</strong></span></td>
</tr>
<tr class="even">
<td>Proteomics</td>
<td>Proteins</td>
<td><span style="color:#3C5488FF"><strong>Mass Spectrometry</strong></span></td>
</tr>
<tr class="odd">
<td>Metabolomics</td>
<td>Metabolites</td>
<td><span style="color:#3C5488FF"><strong>Mass Spectrometry</strong></span></td>
</tr>
</tbody>
</table>
</section>
<section id="examples-of-hts-applications" class="slide level2">
<h2>Examples of HTS applications</h2>
<ul>
<li><strong>Whole-genome assembly</strong> ‚Äî for producing reference genomes</li>
</ul>
<div class="fragment">
<ul>
<li><strong>Typing of SNPs and other sequence variants</strong> ‚Äî for population genetics, GWAS, etc.</li>
</ul>
</div>
<div class="fragment">
<ul>
<li><strong>Metabarcoding</strong> ‚Äî for microbial community characterization</li>
</ul>
</div>
<div class="fragment">
<ul>
<li><strong>RNA-Seq</strong> ‚Äî for large-scale gene expression analysis</li>
</ul>
</div>
<div class="fragment">
<p><br></p>
<details>
<summary>
<strong><em>Does any of you work on, or is planning to work on, projects like these?</em></strong> &nbsp; <i class="fa-solid fa-comments" style="color:darkolivegreen;"></i>
</summary>
</details>
</div>
</section>
<section id="the-main-hts-technologies" class="slide level2">
<h2>The main HTS technologies</h2>
<table class="caption-top">
<colgroup>
<col style="width: 17%">
<col style="width: 41%">
<col style="width: 41%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th>Short-read HTS</th>
<th>Long-read HTS</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><span style="color:white"><strong>Main companies</strong></span></td>
<td><span style="color:white">Illumina</span></td>
<td><span style="color:white">Oxford Nanopore Technologies (ONT) &amp; Pacific Biosciences (PacBio)</span></td>
</tr>
</tbody>
</table>
</section>
<section id="the-main-hts-technologies-1" class="slide level2">
<h2>The main HTS technologies</h2>
<table class="table-striped caption-top">
<colgroup>
<col style="width: 17%">
<col style="width: 41%">
<col style="width: 41%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th>Short-read HTS</th>
<th>Long-read HTS</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><strong>Usage</strong></td>
<td>More</td>
<td>Less (but increasing)</td>
</tr>
<tr class="even">
<td><strong>Main companies</strong></td>
<td>Illumina</td>
<td>Oxford Nanopore Technologies (ONT) &amp; Pacific Biosciences (PacBio)</td>
</tr>
<tr class="odd">
<td><strong>Timeline</strong></td>
<td>Since 2005 ‚Äî <span style="color:crimson">technology fairly stable</span></td>
<td>Since 2011 ‚Äî <span style="color:green">still rapid development</span></td>
</tr>
</tbody>
</table>
</section>
<section id="the-main-hts-technologies-2" class="slide level2">
<h2>The main HTS technologies</h2>
<table class="table-striped caption-top">
<colgroup>
<col style="width: 17%">
<col style="width: 41%">
<col style="width: 41%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th>Short-read HTS</th>
<th>Long-read HTS</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><strong>Usage</strong></td>
<td>More</td>
<td>Less (but increasing)</td>
</tr>
<tr class="even">
<td><strong>Main companies</strong></td>
<td>Illumina</td>
<td>Oxford Nanopore Technologies (ONT) &amp; Pacific Biosciences (PacBio)</td>
</tr>
<tr class="odd">
<td><strong>Timeline</strong></td>
<td>Since 2005 ‚Äî <span style="color:crimson">technology fairly stable</span></td>
<td>Since 2011 ‚Äî <span style="color:green">still rapid development</span></td>
</tr>
<tr class="even">
<td><strong>Read lengths</strong></td>
<td><span style="color:crimson">50-300 bp</span></td>
<td><span style="color:green">10-100+ kbp</span></td>
</tr>
<tr class="odd">
<td><strong>Error rates</strong></td>
<td><span style="color:green">Mostly &lt;0.1%</span></td>
<td><span style="color:crimson">1-10% (ONT) / &lt;0.1-10% (PacBio)</span></td>
</tr>
</tbody>
</table>
</section>
<section id="the-main-hts-technologies-3" class="slide level2">
<h2>The main HTS technologies</h2>
<table class="table-striped caption-top">
<colgroup>
<col style="width: 17%">
<col style="width: 41%">
<col style="width: 41%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th>Short-read HTS</th>
<th>Long-read HTS</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><strong>Usage</strong></td>
<td>More</td>
<td>Less (but increasing)</td>
</tr>
<tr class="even">
<td><strong>Main companies</strong></td>
<td>Illumina</td>
<td>Oxford Nanopore Technologies (ONT) &amp; Pacific Biosciences (PacBio)</td>
</tr>
<tr class="odd">
<td><strong>Timeline</strong></td>
<td>Since 2005 ‚Äî <span style="color:crimson">technology fairly stable</span></td>
<td>Since 2011 ‚Äî <span style="color:green">still rapid development</span></td>
</tr>
<tr class="even">
<td><strong>Read lengths</strong></td>
<td><span style="color:crimson">50-300 bp</span></td>
<td><span style="color:green">10-100+ kbp</span></td>
</tr>
<tr class="odd">
<td><strong>Error rates</strong></td>
<td><span style="color:green">Mostly &lt;0.1%</span></td>
<td><span style="color:crimson">1-10% (ONT) / &lt;0.1-10% (PacBio)</span></td>
</tr>
<tr class="even">
<td><strong>Throughput</strong></td>
<td><span style="color:green">Higher</span></td>
<td><span style="color:crimson">Lower</span></td>
</tr>
<tr class="odd">
<td><strong>Cost per base</strong></td>
<td><span style="color:green">Lower</span></td>
<td><span style="color:crimson">Higher</span></td>
</tr>
</tbody>
</table>
</section>
<section id="read-lengths" class="slide level2">
<h2>Read lengths</h2>
<details>
<summary>
<strong>Can you think of applications where long reads are useful?</strong> &nbsp; <i class="fa-solid fa-comments" style="color:darkolivegreen;"></i>
</summary>
<p>For example:</p>
<ul>
<li>Genome assembly</li>
<li>Read-based taxonomic identification</li>
</ul>
</details>
<div class="fragment">
<hr style="height:1pt; visibility:hidden;">
<details>
<summary>
<strong>Can you think of applications where read length may not matter much?</strong> &nbsp; <i class="fa-solid fa-comments" style="color:darkolivegreen;"></i>
</summary>
<p>For example:</p>
<ul>
<li>(SNP) variant analysis</li>
<li>‚ÄúCounting applications‚Äù such as RNA-Seq</li>
</ul>
<p>For these, genomic locations (variant analysis) or gene identities (RNA-Seq) can be reliably inferred from as little as 25 bp</p>
</details>
</div>
</section>
<section id="error-rates" class="slide level2">
<h2>Error rates</h2>
<p>A read‚Äôs sequence may differ from the actual DNA sequence it originated from:</p>
<ul>
<li>The read can have base-calling errors, missing bases, or extra bases</li>
</ul>
<div class="fragment">
<ul>
<li>When the base calling software is not confident, it can also return <code>N</code>s (= undetermined)</li>
</ul>
<div class="quarto-figure quarto-figure-center">
<figure>
<p><a href="../img/noisy.sequence.png" class="lightbox" data-gallery="quarto-lightbox-gallery-4"><img data-src="../img/noisy.sequence.png" class="quarto-figure quarto-figure-center" style="width:85.0%" alt="A chromatogram with several uncalled bases."></a></p>
</figure>
</div>
</div>
<div class="fragment">
<div class="callout callout-note callout-style-simple">
<div class="callout-body">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-content">
<p>When you receive HTS reads, base calls have typically been made already.<br>
Every base call is accompanied by a <strong>quality score</strong>, representing the estimated error probability.</p>
</div>
</div>
</div>
</div>
</section>
<section id="correcting-sequencing-errors" class="slide level2">
<h2>Correcting sequencing errors</h2>
<p>To overcome sequencing errors, every base can be sequenced multiple times ‚Äì<br>
i.e., obtaining a ‚Äú<strong>depth of coverage</strong>‚Äù greater than 1:</p>
<div class="quarto-figure quarto-figure-center">
<figure>
<p><a href="../img/coverage2.jpg" class="lightbox" data-gallery="quarto-lightbox-gallery-5"><img data-src="../img/coverage2.jpg" class="quarto-figure quarto-figure-center" style="width:60.0%" alt="A diagram illustraing the concept of depth of coverage."></a></p>
</figure>
</div>
<div class="fragment">
<details>
<summary>
<strong>Which natural phenomenon might complicate this effort?</strong> &nbsp; <i class="fa-solid fa-comments" style="color:darkolivegreen;"></i>
</summary>
<strong>Genetic variation</strong> among and (for diploid organisms) within individuals
</details>
<hr style="height:1pt; visibility:hidden;">
</div>
<div class="fragment">
<div class="callout callout-tip callout-style-simple">
<div class="callout-body">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-content">
<p><em>Typical depths of coverage: ~50-100x for genome assembly &amp; 10-30x for variant typing (!)</em></p>
</div>
</div>
</div>
</div>
</section></section>
<section>
<section id="short-read-hts" class="title-slide slide level1 center" data-background-color="silver">
<h1>Short-read HTS</h1>

</section>
<section id="libraries-and-library-prep" class="slide level2">
<h2>Libraries and library prep</h2>
<ul>
<li>In a HTS context, a ‚Äú<strong>library</strong>‚Äù is a collection of DNA fragments ready for sequencing</li>
</ul>
<div class="fragment">
<ul>
<li>These fragments can number in the <strong>millions or billions</strong> and are often randomly generated from input like genomic DNA:</li>
</ul>
<div style="margin-bottom: 0px;">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><a href="../img/library-prep_cropped.png" class="lightbox" data-gallery="quarto-lightbox-gallery-6"><img data-src="../img/library-prep_cropped.png" class="quarto-figure quarto-figure-center" alt="A diagram showing the main Illumina library preparation steps."></a></p>
</figure>
</div>
</div>
<div class="legend2">
<p>An overview of the <strong>library prep</strong> procedure. This is typically done for you by a sequencing facility or company.</p>
</div>
</div>
</section>
<section id="libraries-and-library-prep-1" class="slide level2">
<h2>Libraries and library prep</h2>
<ul>
<li>After library prep, each DNA fragment is flanked by several types of short sequences that together make up the ‚Äú<strong>adapters</strong>‚Äù:</li>
</ul>
<p><br></p>
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="../img/adapters2.png" class="quarto-figure quarto-figure-center"></p>
</figure>
</div>
<div class="fragment">
<hr style="height:1pt; visibility:hidden;">
<div class="callout callout-note callout-titled callout-style-default">
<div class="callout-body">
<div class="callout-title">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<p><strong>Multiplexing!</strong></p>
</div>
<div class="callout-content">
<p>Adapters can include ‚Äúindices‚Äù or ‚Äúbarcodes‚Äù to identify individual samples, so many samples can be combined (<em>multiplexed</em>) into a single library</p>
</div>
</div>
</div>
</div>
</section>
<section id="paired-end-vs.-single-end-sequencing" class="slide level2">
<h2>Paired-end vs.&nbsp;single-end sequencing</h2>
<ul>
<li>Fragments can be sequenced from both ends as shown below ‚Äî<br>
this is called <strong>‚Äúpaired-end‚Äù (PE) sequencing</strong>:</li>
</ul>
<div class="quarto-figure quarto-figure-center">
<figure>
<p><a href="../img/pairedend.png" class="lightbox" data-gallery="quarto-lightbox-gallery-7"><img data-src="../img/pairedend.png" class="quarto-figure quarto-figure-center" alt="A diagram showing forward and reverse reads in paired-end sequencing."></a></p>
</figure>
</div>
<div class="fragment">
<ul>
<li>When sequencing is instead <strong>single-end (SE)</strong>, no reverse read is produced:</li>
</ul>
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="../img/singleend.png" class="quarto-figure quarto-figure-center"></p>
</figure>
</div>
</div>
</section>
<section id="fragment-size-variation" class="slide level2">
<h2>Fragment size variation</h2>
<ul>
<li>DNA fragment size varies ‚Äì by design and because of limited precision in size selection</li>
</ul>
<div class="quarto-figure quarto-figure-center">
<figure>
<p><a href="../img/pairedend.png" class="lightbox" data-gallery="quarto-lightbox-gallery-8"><img data-src="../img/pairedend.png" class="quarto-figure quarto-figure-center" style="width:75.0%" alt="A diagram showing forward and reverse reads in paired-end sequencing."></a></p>
</figure>
</div>
<details>
<summary>
<strong>What happens when a fragment is shorter than the length of a single F or R read?</strong> &nbsp; <i class="fa-solid fa-comments" style="color:darkolivegreen;"></i>
</summary>
<p>‚Äú<strong>Adapter read-through</strong>‚Äù: the final bases in the resulting reads will consist of adapter sequence (these should be removed before downstream analysis)</p>
<div class="quarto-figure quarto-figure-center">
<figure>
<p><a href="../img/insert_size_adapter-read-through.png" class="lightbox" data-gallery="quarto-lightbox-gallery-9"><img data-src="../img/insert_size_adapter-read-through.png" class="quarto-figure quarto-figure-center" style="width:52.0%" alt="A diagram illustrating the scenario when the DNA fragment is shorter than the single read length"></a></p>
</figure>
</div>
</details>
<div class="footer">

</div>
</section>
<section id="fragment-size-variation-1" class="slide level2">
<h2>Fragment size variation</h2>
<ul>
<li>DNA fragment size varies ‚Äì by design and because of limited precision in size selection</li>
</ul>
<div class="quarto-figure quarto-figure-center">
<figure>
<p><a href="../img/pairedend.png" class="lightbox" data-gallery="quarto-lightbox-gallery-10"><img data-src="../img/pairedend.png" class="quarto-figure quarto-figure-center" style="width:75.0%" alt="A diagram showing forward and reverse reads in paired-end sequencing."></a></p>
</figure>
</div>
<details>
<summary>
<strong>What happens when a fragment is shorter than the combined F + R read length?</strong> &nbsp; <i class="fa-solid fa-comments" style="color:darkolivegreen;"></i>
</summary>
<p><strong>Overlapping reads</strong> (this can be useful!):</p>
<div class="quarto-figure quarto-figure-center">
<figure>
<p><a href="../img/insert_size_overlap.png" class="lightbox" data-gallery="quarto-lightbox-gallery-11"><img data-src="../img/insert_size_overlap.png" class="quarto-figure quarto-figure-center" style="width:60.0%" alt="A diagram illustrating the scenario when the DNA fragment is shorter than the combined read length"></a></p>
</figure>
</div>
</details>
<div class="footer">

</div>
</section>
<section id="how-illumina-sequencing-works" class="slide level2">
<h2>How Illumina sequencing works</h2>
<p>We start with fragments that have been attached to a flow cell ‚Äî<br>
this image shows just two, but millions are present simultaneously</p>
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="../img/illumina_cycle0.png" class="quarto-figure quarto-figure-center" style="width:100.0%"></p>
</figure>
</div>
</section>
<section id="how-illumina-sequencing-works-1" class="slide level2">
<h2>How Illumina sequencing works</h2>
<p>Sequencing is performed by synthesizing a new strand using fluorescently-labeled bases,<br>
and taking a picture each time a new nucleotide is incorporated:</p>
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="../img/illumina_cycle1.png" class="quarto-figure quarto-figure-center" style="width:100.0%"></p>
</figure>
</div>
</section>
<section id="how-illumina-sequencing-works-2" class="slide level2">
<h2>How Illumina sequencing works</h2>
<p>Sequencing is performed by synthesizing a new strand using fluorescently-labeled bases,<br>
and taking a picture each time a new nucleotide is incorporated:</p>
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="../img/illumina_cycle2.png" class="quarto-figure quarto-figure-center" style="width:100.0%"></p>
</figure>
</div>
</section>
<section id="how-illumina-sequencing-works-3" class="slide level2">
<h2>How Illumina sequencing works</h2>
<p>Sequencing is performed by synthesizing a new strand using fluorescently-labeled bases,<br>
and taking a picture each time a new nucleotide is incorporated:</p>
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="../img/illumina_cycle5.png" class="quarto-figure quarto-figure-center" style="width:100.0%"></p>
</figure>
</div>
</section>
<section id="the-scale-of-illumina-sequencing" class="slide level2">
<h2>The scale of Illumina sequencing</h2>
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="../img/illumina_scale1.png" class="quarto-figure quarto-figure-center" style="width:95.0%"></p>
</figure>
</div>
</section>
<section id="the-scale-of-illumina-sequencing-1" class="slide level2">
<h2>The scale of Illumina sequencing</h2>
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="../img/illumina_scale2.png" class="quarto-figure quarto-figure-center" style="width:95.0%"></p>
</figure>
</div>
</section>
<section id="the-scale-of-illumina-sequencing-2" class="slide level2">
<h2>The scale of Illumina sequencing</h2>
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="../img/illumina_scale3.png" class="quarto-figure quarto-figure-center" style="width:95.0%"></p>
</figure>
</div>
</section>
<section id="the-scale-of-illumina-sequencing-3" class="slide level2">
<h2>The scale of Illumina sequencing</h2>
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="../img/illumina_scale4.png" class="quarto-figure quarto-figure-center" style="width:95.0%"></p>
</figure>
</div>
</section>
<section id="video-of-illumina-technology" class="slide level2">
<h2>Video of Illumina technology</h2>
<iframe data-external="1" src="https://www.youtube.com/embed/fCd6B5HRaZ8" width="95%" height="81%" title="" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen=""></iframe>
</section></section>
<section>
<section id="reference-genomes" class="title-slide slide level1 center">
<h1>Reference genomes</h1>

</section>
<section id="reference-genomes-1" class="slide level2">
<h2>Reference genomes</h2>
<p>Many HTS applications either <strong>require a ‚Äúreference genome‚Äù</strong> or <strong>involve its production</strong>:</p>
<div class="fragment">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><a href="../img/reads2genome.png" class="lightbox" data-gallery="quarto-lightbox-gallery-12"><img data-src="../img/reads2genome.png" class="quarto-figure quarto-figure-center" style="width:100.0%" alt="A diagram illustrating how sequencing reads are aligned to a reference genome."></a></p>
</figure>
</div>
</div>
</section>
<section id="reference-genomes-2" class="slide level2">
<h2>Reference genomes</h2>
<p>Many HTS applications either <strong>require a ‚Äúreference genome‚Äù</strong> or <strong>involve its production:</strong></p>
<div class="quarto-figure quarto-figure-center">
<figure>
<p><a href="../img/reads2genome_2.png" class="lightbox" data-gallery="quarto-lightbox-gallery-13"><img data-src="../img/reads2genome_2.png" class="quarto-figure quarto-figure-center" style="width:100.0%" alt="A diagram illustrating how sequencing reads are aligned to a reference genome."></a></p>
</figure>
</div>
</section>
<section id="reference-genomes-3" class="slide level2">
<h2>Reference genomes</h2>
<p>What exactly does reference genome refer to? It usually includes:</p>
<div class="fragment">
<ul>
<li><p><strong>An assembly</strong><br>
A representation of most or all of the genome DNA sequence: the genome assembly</p></li>
<li><p><strong>An annotation</strong><br>
Provides e.g.&nbsp;<em>locations of genes and other genomic ‚Äúfeatures‚Äù</em> in the corresponding genome assembly, and functional information for these features</p></li>
</ul>
</div>
<div class="fragment">
<hr style="height:1pt; visibility:hidden;">
<div class="callout callout-tip callout-titled callout-style-default">
<div class="callout-body">
<div class="callout-title">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<p><strong>Taxonomic identity</strong></p>
</div>
<div class="callout-content">
<p>Reference genomes are typically applicable at the <strong>species level</strong>. For example, if you work with maize, you want a <em>Zea mays</em> reference genome. But:</p>
<ul>
<li>If needed, it‚Äôs often possible to work with genomes of closely related species</li>
<li>Conversely, different subspecies/lines may have their own reference genomes</li>
</ul>
</div>
</div>
</div>
</div>
</section>
<section id="there-is-enormous-variation-in-genome-size" class="slide level2">
<h2>There is enormous variation in genome size</h2>
<div class="quarto-figure quarto-figure-center">
<figure>
<p><a href="https://en.wikipedia.org/wiki/Genome_size"><img data-src="../img/Genome_Sizes.png" class="quarto-figure quarto-figure-center" style="width:73.0%"></a></p>
</figure>
</div>
<div class="legend2">
<p><a href="https://en.wikipedia.org/wiki/Genome_size" class="uri">https://en.wikipedia.org/wiki/Genome_size</a></p>
</div>
<div class="footer">

</div>
</section>
<section id="and-enormous-growth-of-genomes-in-databases" class="slide level2">
<h2>And enormous growth of genomes in databases</h2>
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="../img/genomes_genbank.png" class="quarto-figure quarto-figure-center" style="width:83.0%"></p>
</figure>
</div>
<div class="legend2">
<p><br> <span class="citation" data-cites="konkel2023">Konkel and Slot (<a href="#/references" role="doc-biblioref" onclick="">2023</a>)</span></p>
</div>
</section>
<section id="genome-assemblies" class="slide level2">
<h2>Genome assemblies</h2>
<ul>
<li>Besides being produced at higher rates as shown on the previous slide,<br>
assemblies keep getting better with increasing usage &amp; quality of long-read HTS</li>
</ul>
<div class="fragment">
<ul>
<li>Still, many assemblies instead consist of ‚Äìoften 1000s of‚Äì fragments (contigs and scaffolds)</li>
</ul>
<hr style="height:1pt; visibility:hidden;">
</div>
<div class="fragment">
<div class="callout callout-note callout-titled callout-style-default">
<div class="callout-body">
<div class="callout-title">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<p><strong>How is this data stored?</strong></p>
</div>
<div class="callout-content">
<p>Both genome assemblies and annotations are typically saved in a single text file each ‚Äî we‚Äôll explore these in tomorrow‚Äôs lab</p>
</div>
</div>
</div>
</div>
</section>
<section id="recap" class="slide level2">
<h2>Recap</h2>
<p>You‚Äôve learned‚Ä¶</p>
<ol type="1">
<li><p>That high-throughput sequencing (HTS) enables DNA sequencing at <em>much</em> larger scales than was previously possible</p></li>
<li><p>That HTS underlies several major ‚Äúomics‚Äù fields</p></li>
<li><p>How short-read and long-read HTS have different strengths and weaknesses</p></li>
<li><p>About libraries and the technology underlying short-read sequencing</p></li>
</ol>
<div class="fragment">
<ol start="4" type="1">
<li>That reference genomes are essential for many HTS applications</li>
</ol>
</div>
</section></section>
<section>
<section id="looking-forward" class="title-slide slide level1 center" data-background-color="silver">
<h1>Looking forward</h1>

</section>
<section id="the-garrig√≥s-et-al.-2025-dataset" class="slide level2">
<h2>The Garrig√≥s et al.&nbsp;2025 dataset</h2>
<p>The labs this and next week are organized around the data set from <span class="citation" data-cites="garrig√≥s2025">Garrig√≥s et al. (<a href="#/references" role="doc-biblioref" onclick="">2025</a>)</span>:</p>
<div class="quarto-figure quarto-figure-center">
<figure>
<p><a href="../img/garrigos_paper.png" class="lightbox" data-gallery="quarto-lightbox-gallery-14"><img data-src="../img/garrigos_paper.png" class="quarto-figure quarto-figure-center" style="width:75.0%" alt="A screenshot of the paper's front matter."></a></p>
</figure>
</div>
<p>This paper uses <strong>paired-end Illumina RNA-Seq data</strong> to study gene expression in <em>Culex pipiens</em> mosquitos infected with two different malaria-causing <em>Plasmodium</em> protozoans.</p>
</section>
<section id="tomorrows-lab" class="slide level2">
<h2>Tomorrow‚Äôs lab</h2>
<ul>
<li>You will first learn about what kind of computing environment is commonly used for analyzing high-throughput sequencing (HTS) data</li>
</ul>
<div class="fragment">
<ul>
<li>Then, you‚Äôll work in this environment to explore an HTS dataset, checking out HTS read and reference genome files, and performing read quality control</li>
</ul>
</div>
</section>
<section id="next-weeks-content" class="slide level2">
<h2>Next week‚Äôs content</h2>
<ul>
<li>In the lecture, you will learn how RNA-Seq is used to study gene expression at scale</li>
</ul>
<div class="fragment">
<ul>
<li>In the lab, you will perform differential gene expression analysis on the <em>Culex pipiens</em> RNA-Seq dataset</li>
</ul>
</div>
</section></section>
<section>
<section id="bonus-more-on-hts-applications" class="title-slide slide level1 center">
<h1>Bonus: More on HTS applications</h1>

</section>
<section id="examples-of-hts-applications-1" class="slide level2">
<h2>Examples of HTS applications</h2>
<p>We talked about these HTS applications:</p>
<ul>
<li><strong>Whole-genome assembly</strong> ‚Äî for producing reference genomes</li>
<li><strong>Typing of SNPs and other sequence variants</strong> ‚Äî for population genetics, GWAS, etc.</li>
<li><strong>Metabarcoding</strong> ‚Äî for microbial community characterization</li>
<li><strong>RNA-Seq</strong> ‚Äî for large-scale gene expression analysis</li>
</ul>
<hr style="height:1pt; visibility:hidden;">
<div class="fragment">
<p>In each of of these applications:</p>
<ul>
<li>What part(s) of the genome are you sequencing?</li>
<li>Who are you sequencing (how many individuals/species)?</li>
<li>How are the sequences used?</li>
</ul>
</div>
</section>
<section id="examples-of-hts-applications-2" class="slide level2">
<h2>Examples of HTS applications</h2>
<div style="font-size: 85%;">
<table class="table-striped caption-top">
<colgroup>
<col style="width: 16%">
<col style="width: 27%">
<col style="width: 27%">
<col style="width: 27%">
</colgroup>
<thead>
<tr class="header">
<th>Appl.</th>
<th>What</th>
<th>Who</th>
<th>How</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><strong>Whole-genome assembly</strong></td>
<td>The whole genome! üòÉ</td>
<td>A single individual</td>
<td>‚Äú<em>Overlap</em>‚Äù reads into larger fragments</td>
</tr>
<tr class="even">
<td><strong>Variant typing</strong></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr class="odd">
<td><strong>16S Metabarcoding</strong></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr class="even">
<td><strong>RNA-Seq</strong></td>
<td></td>
<td></td>
<td></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="examples-of-hts-applications-3" class="slide level2">
<h2>Examples of HTS applications</h2>
<div style="font-size: 85%;">
<table class="table-striped caption-top">
<colgroup>
<col style="width: 16%">
<col style="width: 27%">
<col style="width: 27%">
<col style="width: 27%">
</colgroup>
<thead>
<tr class="header">
<th>Appl.</th>
<th>What</th>
<th>Who</th>
<th>How</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><strong>Whole-genome assembly</strong></td>
<td>Whole genome üòÉ</td>
<td>A single individual</td>
<td>‚Äú<em>Overlap</em>‚Äù reads into larger fragments</td>
</tr>
<tr class="even">
<td><strong>Variant typing</strong></td>
<td>Whole genome <br>or specific (e.g.&nbsp;exome)<br> or random (GBS/RAD) regions</td>
<td>Multiple individuals</td>
<td>For each site &amp; individual, determine the <em>variant state</em> (allele)</td>
</tr>
<tr class="odd">
<td><strong>16S Metabarcoding</strong></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr class="even">
<td><strong>RNA-Seq</strong></td>
<td></td>
<td></td>
<td></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="examples-of-hts-applications-4" class="slide level2">
<h2>Examples of HTS applications</h2>
<div style="font-size: 85%;">
<table class="table-striped caption-top">
<colgroup>
<col style="width: 16%">
<col style="width: 27%">
<col style="width: 27%">
<col style="width: 27%">
</colgroup>
<thead>
<tr class="header">
<th>Appl.</th>
<th>What</th>
<th>Who</th>
<th>How</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><strong>Whole-genome assembly</strong></td>
<td>Whole genome üòÉ</td>
<td>A single individual</td>
<td>‚Äú<em>Overlap</em>‚Äù reads into larger fragments</td>
</tr>
<tr class="even">
<td><strong>Variant typing</strong></td>
<td>Whole genome <br>or specific (e.g.&nbsp;exome)<br> or random (GBS/RAD) regions</td>
<td>Multiple individuals</td>
<td>For each site &amp; individual, determine the <em>variant state</em> (allele)</td>
</tr>
<tr class="odd">
<td><strong>16S Metabarcoding</strong></td>
<td>A specific locus (e.g.&nbsp;16S)</td>
<td>Multi-species samples ‚Äî<br>soil, gut contents, etc.</td>
<td>Assign reads a <em>taxonomic</em> identity and <strong><em>count</em></strong></td>
</tr>
<tr class="even">
<td><strong>RNA-Seq</strong></td>
<td></td>
<td></td>
<td></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="examples-of-hts-applications-5" class="slide level2">
<h2>Examples of HTS applications</h2>
<div style="font-size: 85%;">
<table class="table-striped caption-top">
<colgroup>
<col style="width: 16%">
<col style="width: 27%">
<col style="width: 27%">
<col style="width: 27%">
</colgroup>
<thead>
<tr class="header">
<th>Appl.</th>
<th>What</th>
<th>Who</th>
<th>How</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><strong>Whole-genome assembly</strong></td>
<td>Whole genome üòÉ</td>
<td>A single individual</td>
<td>‚Äú<em>Overlap</em>‚Äù reads into larger fragments</td>
</tr>
<tr class="even">
<td><strong>Variant typing</strong></td>
<td>Whole genome <br>or specific (e.g.&nbsp;exome)<br> or random (GBS/RAD) regions</td>
<td>Multiple individuals</td>
<td>For each site &amp; individual, determine the <em>variant state</em> (allele)</td>
</tr>
<tr class="odd">
<td><strong>16S Metabarcoding</strong></td>
<td>A specific locus (e.g.&nbsp;16S)</td>
<td>Multi-species samples ‚Äî<br>soil, gut contents, etc.</td>
<td>Assign reads a <em>taxonomic</em> identity and <strong><em>count</em></strong></td>
</tr>
<tr class="even">
<td><strong>RNA-Seq</strong></td>
<td>Whole transcriptome</td>
<td>Multiple indivuals<br>(&amp; multiple conditions)</td>
<td>Assign reads a <em>gene</em> identity and <strong><em>count</em></strong></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="references" class="slide level2 smaller scrollable">
<h2>References</h2>


<div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" role="list">
<div id="ref-garrig√≥s2025" class="csl-entry" role="listitem">
Garrig√≥s, Marta, Guillem Ylla, Josu√© Mart√≠nez-de la Puente, Jordi Figuerola, and Mar√≠a Jos√© Ruiz-L√≥pez. 2025. <span>‚ÄúTwo Avian <em>Plasmodium</em> Species Trigger Different Transcriptional Responses on Their Vector <em>Culex <span>p</span>ipiens</em>.‚Äù</span> <em>Molecular Ecology</em> 34 (15): e17240. <a href="https://doi.org/10.1111/mec.17240">https://doi.org/10.1111/mec.17240</a>.
</div>
<div id="ref-konkel2023" class="csl-entry" role="listitem">
Konkel, Zachary, and Jason C. Slot. 2023. <span>‚ÄúMycotools: An Automated and Scalable Platform for Comparative Genomics.‚Äù</span> <em>BioRxiv</em>. <a href="https://doi.org/10.1101/2023.09.08.556886">https://doi.org/10.1101/2023.09.08.556886</a>.
</div>
</div>
</section></section>
    </div>
  <div class="quarto-auto-generated-content" style="display: none;">
<p><img src="../img/CFAES-Horiz-RGBHEX.svg" class="slide-logo"></p>
<div class="footer footer-default">
<p><a href="https://mcic-osu.github.io/ENTMLGY6703-SP26">~ ENTMLGY6703-SP26 ~</a></p>
</div>
</div></div>

  <script>window.backupDefine = window.define; window.define = undefined;</script>
  <script src="../site_libs/revealjs/dist/reveal.js"></script>
  <!-- reveal.js plugins -->
  <script src="../site_libs/revealjs/plugin/quarto-line-highlight/line-highlight.js"></script>
  <script src="../site_libs/revealjs/plugin/pdf-export/pdfexport.js"></script>
  <script src="../site_libs/revealjs/plugin/reveal-menu/menu.js"></script>
  <script src="../site_libs/revealjs/plugin/reveal-menu/quarto-menu.js"></script>
  <script src="../site_libs/revealjs/plugin/reveal-chalkboard/plugin.js"></script>
  <script src="../site_libs/revealjs/plugin/quarto-support/support.js"></script>
  

  <script src="../site_libs/revealjs/plugin/notes/notes.js"></script>
  <script src="../site_libs/revealjs/plugin/search/search.js"></script>
  <script src="../site_libs/revealjs/plugin/zoom/zoom.js"></script>
  <script src="../site_libs/revealjs/plugin/math/math.js"></script>
  <script>window.define = window.backupDefine; window.backupDefine = undefined;</script>

  <script>

      // Full list of configuration options available at:
      // https://revealjs.com/config/
      Reveal.initialize({
'controlsAuto': true,
'previewLinksAuto': false,
'pdfSeparateFragments': false,
'autoAnimateEasing': "ease",
'autoAnimateDuration': 1,
'autoAnimateUnmatched': true,
'jumpToSlide': true,
'menu': {"side":"left","useTextContentForMissingTitles":true,"markers":false,"loadIcons":false,"custom":[{"title":"Tools","icon":"<i class=\"fas fa-gear\"></i>","content":"<ul class=\"slide-menu-items\">\n<li class=\"slide-tool-item active\" data-item=\"0\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.fullscreen(event)\"><kbd>f</kbd> Fullscreen</a></li>\n<li class=\"slide-tool-item\" data-item=\"1\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.speakerMode(event)\"><kbd>s</kbd> Speaker View</a></li>\n<li class=\"slide-tool-item\" data-item=\"2\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.overview(event)\"><kbd>o</kbd> Slide Overview</a></li>\n<li class=\"slide-tool-item\" data-item=\"3\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.togglePdfExport(event)\"><kbd>e</kbd> PDF Export Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"4\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleScrollView(event)\"><kbd>r</kbd> Scroll View Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"5\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleChalkboard(event)\"><kbd>b</kbd> Toggle Chalkboard</a></li>\n<li class=\"slide-tool-item\" data-item=\"6\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleNotesCanvas(event)\"><kbd>c</kbd> Toggle Notes Canvas</a></li>\n<li class=\"slide-tool-item\" data-item=\"7\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.downloadDrawings(event)\"><kbd>d</kbd> Download Drawings</a></li>\n<li class=\"slide-tool-item\" data-item=\"8\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.keyboardHelp(event)\"><kbd>?</kbd> Keyboard Help</a></li>\n</ul>"}],"openButton":true,"width":"wide","numbers":true},
'chalkboard': {"buttons":true,"theme":"whiteboard"},
'smaller': true,
 
        // Display controls in the bottom right corner
        controls: false,

        // Help the user learn the controls by providing hints, for example by
        // bouncing the down arrow when they first encounter a vertical slide
        controlsTutorial: false,

        // Determines where controls appear, "edges" or "bottom-right"
        controlsLayout: 'edges',

        // Visibility rule for backwards navigation arrows; "faded", "hidden"
        // or "visible"
        controlsBackArrows: 'faded',

        // Display a presentation progress bar
        progress: true,

        // Display the page number of the current slide
        slideNumber: 'c/t',

        // 'all', 'print', or 'speaker'
        showSlideNumber: 'all',

        // Add the current slide number to the URL hash so that reloading the
        // page/copying the URL will return you to the same slide
        hash: true,

        // Start with 1 for the hash rather than 0
        hashOneBasedIndex: false,

        // Flags if we should monitor the hash and change slides accordingly
        respondToHashChanges: true,

        // Push each slide change to the browser history
        history: true,

        // Enable keyboard shortcuts for navigation
        keyboard: true,

        // Enable the slide overview mode
        overview: true,

        // Disables the default reveal.js slide layout (scaling and centering)
        // so that you can use custom CSS layout
        disableLayout: false,

        // Vertical centering of slides
        center: false,

        // Enables touch navigation on devices with touch input
        touch: true,

        // Loop the presentation
        loop: false,

        // Change the presentation direction to be RTL
        rtl: false,

        // see https://revealjs.com/vertical-slides/#navigation-mode
        navigationMode: 'linear',

        // Randomizes the order of slides each time the presentation loads
        shuffle: false,

        // Turns fragments on and off globally
        fragments: true,

        // Flags whether to include the current fragment in the URL,
        // so that reloading brings you to the same fragment position
        fragmentInURL: false,

        // Flags if the presentation is running in an embedded mode,
        // i.e. contained within a limited portion of the screen
        embedded: false,

        // Flags if we should show a help overlay when the questionmark
        // key is pressed
        help: true,

        // Flags if it should be possible to pause the presentation (blackout)
        pause: true,

        // Flags if speaker notes should be visible to all viewers
        showNotes: false,

        // Global override for autoplaying embedded media (null/true/false)
        autoPlayMedia: null,

        // Global override for preloading lazy-loaded iframes (null/true/false)
        preloadIframes: null,

        // Number of milliseconds between automatically proceeding to the
        // next slide, disabled when set to 0, this value can be overwritten
        // by using a data-autoslide attribute on your slides
        autoSlide: 0,

        // Stop auto-sliding after user input
        autoSlideStoppable: true,

        // Use this method for navigation when auto-sliding
        autoSlideMethod: null,

        // Specify the average time in seconds that you think you will spend
        // presenting each slide. This is used to show a pacing timer in the
        // speaker view
        defaultTiming: null,

        // Enable slide navigation via mouse wheel
        mouseWheel: false,

        // The display mode that will be used to show slides
        display: 'block',

        // Hide cursor if inactive
        hideInactiveCursor: true,

        // Time before the cursor is hidden (in ms)
        hideCursorTime: 5000,

        // Opens links in an iframe preview overlay
        previewLinks: false,

        // Transition style (none/fade/slide/convex/concave/zoom)
        transition: 'none',

        // Transition speed (default/fast/slow)
        transitionSpeed: 'default',

        // Transition style for full page slide backgrounds
        // (none/fade/slide/convex/concave/zoom)
        backgroundTransition: 'none',

        // Number of slides away from the current that are visible
        viewDistance: 3,

        // Number of slides away from the current that are visible on mobile
        // devices. It is advisable to set this to a lower number than
        // viewDistance in order to save resources.
        mobileViewDistance: 2,

        // The "normal" size of the presentation, aspect ratio will be preserved
        // when the presentation is scaled to fit different resolutions. Can be
        // specified using percentage units.
        width: 1050,

        height: 700,

        // Factor of the display size that should remain empty around the content
        margin: 0.1,

        math: {
          mathjax: 'https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/MathJax.js',
          config: 'TeX-AMS_HTML-full',
          tex2jax: {
            inlineMath: [['\\(','\\)']],
            displayMath: [['\\[','\\]']],
            balanceBraces: true,
            processEscapes: false,
            processRefs: true,
            processEnvironments: true,
            preview: 'TeX',
            skipTags: ['script','noscript','style','textarea','pre','code'],
            ignoreClass: 'tex2jax_ignore',
            processClass: 'tex2jax_process'
          },
        },

        // reveal.js plugins
        plugins: [QuartoLineHighlight, PdfExport, RevealMenu, RevealChalkboard, QuartoSupport,

          RevealMath,
          RevealNotes,
          RevealSearch,
          RevealZoom
        ]
      });
    </script>
    <script id="quarto-html-after-body" type="application/javascript">
    window.document.addEventListener("DOMContentLoaded", function (event) {
      const toggleBodyColorMode = (bsSheetEl) => {
        const mode = bsSheetEl.getAttribute("data-mode");
        const bodyEl = window.document.querySelector("body");
        if (mode === "dark") {
          bodyEl.classList.add("quarto-dark");
          bodyEl.classList.remove("quarto-light");
        } else {
          bodyEl.classList.add("quarto-light");
          bodyEl.classList.remove("quarto-dark");
        }
      }
      const toggleBodyColorPrimary = () => {
        const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
        if (bsSheetEl) {
          toggleBodyColorMode(bsSheetEl);
        }
      }
      toggleBodyColorPrimary();  
      const tabsets =  window.document.querySelectorAll(".panel-tabset-tabby")
      tabsets.forEach(function(tabset) {
        const tabby = new Tabby('#' + tabset.id);
      });
      const isCodeAnnotation = (el) => {
        for (const clz of el.classList) {
          if (clz.startsWith('code-annotation-')) {                     
            return true;
          }
        }
        return false;
      }
      const onCopySuccess = function(e) {
        // button target
        const button = e.trigger;
        // don't keep focus
        button.blur();
        // flash "checked"
        button.classList.add('code-copy-button-checked');
        var currentTitle = button.getAttribute("title");
        button.setAttribute("title", "Copied!");
        let tooltip;
        if (window.bootstrap) {
          button.setAttribute("data-bs-toggle", "tooltip");
          button.setAttribute("data-bs-placement", "left");
          button.setAttribute("data-bs-title", "Copied!");
          tooltip = new bootstrap.Tooltip(button, 
            { trigger: "manual", 
              customClass: "code-copy-button-tooltip",
              offset: [0, -8]});
          tooltip.show();    
        }
        setTimeout(function() {
          if (tooltip) {
            tooltip.hide();
            button.removeAttribute("data-bs-title");
            button.removeAttribute("data-bs-toggle");
            button.removeAttribute("data-bs-placement");
          }
          button.setAttribute("title", currentTitle);
          button.classList.remove('code-copy-button-checked');
        }, 1000);
        // clear code selection
        e.clearSelection();
      }
      const getTextToCopy = function(trigger) {
          const codeEl = trigger.previousElementSibling.cloneNode(true);
          for (const childEl of codeEl.children) {
            if (isCodeAnnotation(childEl)) {
              childEl.remove();
            }
          }
          return codeEl.innerText;
      }
      const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
        text: getTextToCopy
      });
      clipboard.on('success', onCopySuccess);
      if (window.document.getElementById('quarto-embedded-source-code-modal')) {
        const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
          text: getTextToCopy,
          container: window.document.getElementById('quarto-embedded-source-code-modal')
        });
        clipboardModal.on('success', onCopySuccess);
      }
        var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
        var mailtoRegex = new RegExp(/^mailto:/);
          var filterRegex = new RegExp("https:\/\/jelmerp\.github\.io\/ENTMLGY6703-SP26");
        var isInternal = (href) => {
            return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
        }
        // Inspect non-navigation links and adorn them if external
     	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
        for (var i=0; i<links.length; i++) {
          const link = links[i];
          if (!isInternal(link.href)) {
            // undo the damage that might have been done by quarto-nav.js in the case of
            // links that we want to consider external
            if (link.dataset.originalHref !== undefined) {
              link.href = link.dataset.originalHref;
            }
          }
        }
      function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
        const config = {
          allowHTML: true,
          maxWidth: 500,
          delay: 100,
          arrow: false,
          appendTo: function(el) {
              return el.closest('section.slide') || el.parentElement;
          },
          interactive: true,
          interactiveBorder: 10,
          theme: 'light-border',
          placement: 'bottom-start',
        };
        if (contentFn) {
          config.content = contentFn;
        }
        if (onTriggerFn) {
          config.onTrigger = onTriggerFn;
        }
        if (onUntriggerFn) {
          config.onUntrigger = onUntriggerFn;
        }
          config['offset'] = [0,0];
          config['maxWidth'] = 700;
        window.tippy(el, config); 
      }
      const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
      for (var i=0; i<noterefs.length; i++) {
        const ref = noterefs[i];
        tippyHover(ref, function() {
          // use id or data attribute instead here
          let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
          try { href = new URL(href).hash; } catch {}
          const id = href.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note) {
            return note.innerHTML;
          } else {
            return "";
          }
        });
      }
      const findCites = (el) => {
        const parentEl = el.parentElement;
        if (parentEl) {
          const cites = parentEl.dataset.cites;
          if (cites) {
            return {
              el,
              cites: cites.split(' ')
            };
          } else {
            return findCites(el.parentElement)
          }
        } else {
          return undefined;
        }
      };
      var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
      for (var i=0; i<bibliorefs.length; i++) {
        const ref = bibliorefs[i];
        const citeInfo = findCites(ref);
        if (citeInfo) {
          tippyHover(citeInfo.el, function() {
            var popup = window.document.createElement('div');
            citeInfo.cites.forEach(function(cite) {
              var citeDiv = window.document.createElement('div');
              citeDiv.classList.add('hanging-indent');
              citeDiv.classList.add('csl-entry');
              var biblioDiv = window.document.getElementById('ref-' + cite);
              if (biblioDiv) {
                citeDiv.innerHTML = biblioDiv.innerHTML;
              }
              popup.appendChild(citeDiv);
            });
            return popup.innerHTML;
          });
        }
      }
    });
    </script>
    <script>var lightboxQuarto = GLightbox({"closeEffect":"zoom","descPosition":"bottom","loop":false,"openEffect":"zoom","selector":".lightbox"});
    (function() {
      let previousOnload = window.onload;
      window.onload = () => {
        if (previousOnload) {
          previousOnload();
        }
        lightboxQuarto.on('slide_before_load', (data) => {
          const { slideIndex, slideNode, slideConfig, player, trigger } = data;
          const href = trigger.getAttribute('href');
          if (href !== null) {
            const imgEl = window.document.querySelector(`a[href="${href}"] img`);
            if (imgEl !== null) {
              const srcAttr = imgEl.getAttribute("src");
              if (srcAttr && srcAttr.startsWith("data:")) {
                slideConfig.href = srcAttr;
              }
            }
          } 
        });
      
        lightboxQuarto.on('slide_after_load', (data) => {
          const { slideIndex, slideNode, slideConfig, player, trigger } = data;
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(slideNode);
          }
        });
      
      };
      
    })();
              </script>
    

</body></html>