---
title: "High-throughput sequencing and genomes"
pagetitle: "High-throughput sequencing"
author: Jelmer Poelstra
date: 2026-01-29
institute: "CFAES Bioinformatics Core, Ohio State University"
format:
  revealjs:
    theme: [default, ../theme_mod.scss]
    css: ../slides.css
    smaller: true
    scrollable: false
    progress: true
    incremental: false
    slide-number: true
    slide-level: 2
    highlight-style: pygments
    embed-resources: true
    #chalkboard:
    #  theme: whiteboard  # Need to turn off embed-resources for this
    toc: true
    toc-depth: 1
    toc-title: Overview
    footer: "[~ ENTMLGY6703-SP26 ~](https://mcic-osu.github.io/ENTMLGY6703-SP26)"
    logo: ../img/CFAES-Horiz-RGBHEX.svg
    menu:
      width: wide
      numbers: true
editor_options: 
  chunk_output_type: console
bibliography: ../references.bib
---

# Introduction to <br> sequencing technologies {background-color="silver"}

## What do we mean by *sequencing*?

<details><summary>
What does the term sequencing, like in "high-throughput sequencing",
generally refer to?
</summary>
Determining the nucleotide sequence of DNA fragments:
the order of A, Cs, Gs, and Ts.
</details>

. . .

<br>

<details><summary>And what about RNA?</summary>
RNA is usually reverse transcribed to DNA (cDNA) prior to sequencing,
as in nearly all "RNA-Seq".
</details>

## Sequencing technologies: overview

- **Sanger sequencing** \
  Sequences a single, typically PCR-amplified,
  short-ish (≤900 bp) DNA fragment at a time

. . .

<hr style="height:1pt; visibility:hidden;" />

- **High-throughput sequencing (HTS)**\
  Sequences 10^5^-10^9^, usually randomly selected,
  DNA fragments ("reads") at a time — two types:
  
  - Short-read HTS: Cheaper, more accurate, but shorter reads
  - Long-read HTS: More expensive, less accurate, but longer reads

## Sequencing technology development timeline

![](../img/seq_timeline2a.png){fig-align="center" width="72%"}

::: legend2
<br>
Modified after @pereira2020
:::

## Sequencing technology development timeline

![](../img/seq_timeline2.png){fig-align="center" width="66%"}

::: legend2
<br>
Modified after @pereira2020
:::

## Sanger sequencing

Sequences a single, typically PCR-amplified,
short-ish (≤900 bp) DNA fragment at a time.

Sequencing is performed by **synthesizing a new DNA strand** with
fluorescently-labeled nucleotides, using a different color for each base
(`A`, `C`, `G`, `T`).

. . .

<hr style="height:1pt; visibility:hidden;" />

The final result is a **chromatogram** that can be "base-called":

![](../img/good.chromatogram.png){fig-alt="https://dnacore.mgh.harvard.edu/new-cgi-bin/site/pages/sequencing_pages/seq_troubleshooting.jsp" fig-align="center" width="100%"}

. . .

<hr style="height:1pt; visibility:hidden;" />

::: callout-warning
## The entire human genome (3 Gbp) was sequenced with Sanger technology!
Anyone want to guess how much this may have cost?
:::

## Sequencing cost through time

![https://www.genome.gov/about-genomics/fact-sheets/Sequencing-Human-Genome-cost](../img/2022_Sequencing_cost_per_Human_Genome.jpg)

## Present-day Sanger applications

With HTS, DNA can be sequenced much more efficiently and cheaply,
and Sanger sequencing has become less widely used.

<hr style="height:1pt; visibility:hidden;" />

But it is not obsolete, in part because high throughput isn't always needed --\
some present-day uses of Sanger sequencing:

- Examining variation among individuals or populations in one or more
  *candidate or marker genes*
  (for population genetics, phylogenetics, functional inferences, etc.)

- Taxonomic identification of samples

# High-throughput sequencing (HTS) {background-color="silver"}

## Omics

Looking at the bigger picture first,
HTS produces data that underlies several of these main "omics" approaches:

![](../img/fisher-omics-crop.png){fig-align="center" width="50%" fig-alt="A diagram showing the main omics data types." .lightbox}

::: legend2
Copyright [ThermoFisher](https://www.thermofisher.com/us/en/home/brands/thermo-scientific/molecular-biology/molecular-biology-learning-center/molecular-biology-resource-library/spotlight-articles/supporting-multi-omics-approaches.html)
:::

## The main omics data types

| Omics type       | Molecule type      |               |
|------------------|--------------------|-----------------------------|
| Genomics         | DNA                | |
| Epigenomics      | DNA modifications  | [**High-throughput sequencing (HTS)**]{style="color:white"} |
| Transcriptomics  | RNA                | |
| Proteomics       | Proteins           | |
| Metabolomics     | Metabolites        | |

: {.striped}

. . .

<br>

<details><summary>What does the -omics suffix mean {{< fa question >}}</summary>

The "omics" suffix indicates the involvement of large-scale datasets ---
in the sense that, for example, "genomics" data typically spans much or all 
of the genome.

While the boundaries can be fuzzy,
sequencing a single gene in a single organism is not genomics,
and running qPCR for a handful of genes is not transcriptomics.
</details>

## The main omics data types (cont.)

| Omics type       | Molecule type      | Data mainly produced by              |
|------------------|--------------------|-----------------------------|
| Genomics         | DNA                | [**High-throughput sequencing (HTS)**]{style="color:#00A087FF"}  |
| Epigenomics      | DNA modifications  | [**High-throughput sequencing (HTS)**]{style="color:#00A087FF"}  |
| Transcriptomics  | RNA                | [**High-throughput sequencing (HTS)**]{style="color:#00A087FF"}  |
| Proteomics       | Proteins           | [**Mass Spectometry**]{style="color:#3C5488FF"}       |
| Metabolomics     | Metabolites        | [**Mass Spectometry**]{style="color:#3C5488FF"}       |

: {.striped}

## Examples of HTS applications

- **Whole-genome assembly**

. . .

- **Analysis of SNPs and other sequence variants**\
  (For population genetics/genomics, GWAS, etc. --
  often referred to as "resequencing")

. . .

- **RNA-Seq**
  (transcriptome analysis)

. . .

- **Microbial community characterization**
  - Metabarcoding
  - Shotgun metagenomics

## Illustration of HTS

[ILLUSTRATION OF A TANGLED MASS OF READS - Recall that "reads" are sequenced fragments of DNA]

- Assemble ("build back") into a single sequence that can be used e.g. as a reference

- Compare specific sequence variants across multiple samples

- Count the number of reads originating from distinct units,
  such as genes (RNA-Seq) or organisms (microbial community characterization)

## Two key variables in HTS

- Read lengths
- Error rates

## Read lengths

HTS read lengths vary from 300 bp and shorter (short-read HTS) up to tens of
thousands of base pairs (long-read HTS).

<hr style="height:1pt; visibility:hidden;" />

<details><summary>**Can you think of applications where long reads are useful?**</summary>
For example:

- Genome assembly
- Taxonomic identification of single reads (microbial metabarcoding)
</details>

. . .

<hr style="height:1pt; visibility:hidden;" />

<details><summary>**Can you think of applications where read length may not matter much?**</summary>
For example:

- (SNP) variant analysis
- Counting application such as RNA-seq, .... TBA
  
</details>

## Error rates

Currently, no sequencing technology is error-free:
the sequenced read may differ from the actual DNA sequence it came from.

- The read can have base-calling errors, missing bases, or extra bases

. . .

- When the base calling software is not confident,
  it can also return `N`s (= undetermined)

![](../img/noisy.sequence.png){fig-align="center" width="85%" fig-alt="A chromatogram with several uncalled bases." .lightbox}

. . .

::: {.callout-note appearance="simple"}
When you receive HTS reads, base calls have typically been made already.\
Every base call is accompanied by a **quality score**,
representing the estimated error probability.
:::

## Correcting sequencing errors

To overcome sequencing errors, every base can be sequenced multiple times --\
i.e., obtaining a "**depth of coverage**" greater than 1:

![](../img/coverage2.jpg){fig-align="center" width="60%" fig-alt="A diagram illustraing the concept of depth of coverage." .lightbox}

::: {.callout-tip appearance="simple"}
_Typical depths of coverage are ~50-100x for genome assembly and 10-30x for "resequencing" (!)_
:::

. . .

<hr style="height:1pt; visibility:hidden;" />

<details><summary>Which natural phenomenon might complicate this effort?</summary>
**Genetic variation** among and (for diploid organisms) within individuals
</details>

## The main HTS technologies

| | Short-read HTS   | Long-read HTS
|---|-------|-------|
[**Main companies**]{style="color:white"}  | [Illumina]{style="color:white"}                | [Oxford Nanopore Technologies (ONT) & Pacific Biosciences (PacBio)]{style="color:white"}

## The main HTS technologies

| | Short-read HTS   | Long-read HTS
|---|-------|-------|
**Usage**           | More                    | Less (but increasing)
**Main companies**  | Illumina                | Oxford Nanopore Technologies (ONT) & Pacific Biosciences (PacBio)
**Timeline**        | Since 2005 — [technology fairly stable]{style="color:crimson"} | Since 2011 — [still rapid development]{style="color:green"}

: {.striped}

## The main HTS technologies

|                   | Short-read HTS          | Long-read HTS
|---|-------|-------|
**Usage**           | More                    | Less (but increasing)
**Main companies**  | Illumina                | Oxford Nanopore Technologies (ONT) & Pacific Biosciences (PacBio)
**Timeline**        | Since 2005 — [technology fairly stable]{style="color:crimson"} | Since 2011 — [still rapid development]{style="color:green"}
**Read lengths**    | [50-300 bp]{style="color:crimson"}               | [10-100+ kbp]{style="color:green"}
**Error rates**     | [Mostly <0.1%]{style="color:green"}            | [1-10% (ONT) / <0.1-10% (PacBio)]{style="color:crimson"}
**Throughput**    | [Higher]{style="color:green"} | [Lower]{style="color:crimson"}
**Cost per base**    | [Lower]{style="color:green"} | [Higher]{style="color:crimson"}

: {.striped}

## The main HTS technologies

|                   | Short-read HTS          | Long-read HTS
|---|-------|-------|
**Usage**           | More                    | Less (but increasing)
**Main companies**  | Illumina                | Oxford Nanopore Technologies (ONT) & Pacific Biosciences (PacBio)
**Timeline**        | Since 2005 — [technology fairly stable]{style="color:crimson"} | Since 2011 — [still rapid development]{style="color:green"}
**Read lengths**    | [50-300 bp]{style="color:crimson"}               | [10-100+ kbp]{style="color:green"}
**Error rates**     | [Mostly <0.1%]{style="color:green"}            | [1-10% (ONT) / <0.1-10% (PacBio)]{style="color:crimson"}
**Throughput**    | [Higher]{style="color:green"} | [Lower]{style="color:crimson"}
**Cost per base**    | [Lower]{style="color:green"} | [Higher]{style="color:crimson"}
**AKA**             | Next-Generation Sequencing (NGS) | Third-generation sequencing

: {.striped}

# Illumina HTS {background-color="silver"}

## Illumina (short-read HTS / NGS)

- 100-300 bp reads with 0.1-0.2% error rates

- More reads, lower per-base cost,
  and generally lower error rates than long-read sequencing.

. . .

- Machines differ in throughput, read length, cost per Gb:

![](../img/illumina_machines3.png){fig-align="center"}

## Libraries and library prep

In a HTS context,
a "**library**" is a collection of DNA fragments ready for sequencing.

<hr style="height:1pt; visibility:hidden;" />

. . .

In Illumina libraries,
these fragments number in the **millions or billions** and are often
simply _randomly_ generated from input such as genomic DNA:

::: {style="margin-bottom: 0px;"}
![](../img/library-prep_cropped.png){fig-align="center" fig-alt="A diagram showing the main Illumina library preparation steps." .lightbox}
:::

::: legend2
An overview of the **library prep** procedure.
This is typically done for you by a sequencing facility or company.
:::

## Libraries and library prep (cont.)

After library prep,
each DNA fragment is flanked by several types of short sequences that
together make up the "**adapters**":

<br>

![](../img/adapters2.png){fig-align="center"}

. . .

<hr style="height:1pt; visibility:hidden;" />

::: callout-note
#### Multiplexing!
Adapters can include so-called "indices" or "barcodes" that identify individual
samples.
That way, up to 96 samples can be combined (_multiplexed_) into a single library,\
i.e. into a single tube.
::::::

## Paired-end vs. single-end sequencing

DNA fragments can be sequenced from both ends as shown below —\
this is called **"paired-end" (PE) sequencing**:

![](../img/pairedend.png){fig-align="center" fig-alt="A diagram showing forward and reverse reads in paired-end sequencing." .lightbox}

. . .

<hr style="height:1pt; visibility:hidden;" />

When sequencing is instead **single-end (SE)**, no reverse read is produced:

![](../img/singleend.png){fig-align="center"}

## Insert size

- The total size of the biological DNA fragment (without adapters) is often
  called the **insert size:**

![](../img/insert_size.png){fig-align="center"}

## Insert size variation

The insert size can vary --
by design, but also because of limited precision in size selection.
In some cases, it is:

<hr style="height:1pt; visibility:hidden;" />

<details><summary>Shorter than the *combined* read length, which leads to?</summary>

**Overlapping reads** (this can be useful!):

![](../img/insert_size_overlap.png){fig-align="center" width="70%" fig-alt="A diagram illustrating the scenario when the DNA fragment is shorter than the combined read length" .lightbox}

</details>

. . .

<hr style="height:1pt; visibility:hidden;" />

<details><summary>Shorter than the *single* read length, which leads to?</summary>

"**Adapter read-through**": the final bases in the resulting reads will consist of
adapter sequence, which should be removed before downstream analysis

![](../img/insert_size_adapter-read-through.png){fig-align="center" width="58%" fig-alt="A diagram illustrating the scenario when the DNA fragment is shorter than the single read length" .lightbox}

</details>

## How Illumina sequencing works

First, library fragments **bind to a surface** thanks to the adapters,
and the DNA templates are then
**PCR-amplified to form "clusters"** of identical fragments:

![](../img/illumina_seq1.png){fig-align="center" width="10%"}

. . .

::: callout-note
## In the diagram above, for illustrative purposes:

- Only a few nucleotides are shown (1 block = 1 nucleotide) — in reality, fragments are much longer
- Only two templates and clusters are shown — in reality, there are millions
:::

## How Illumina sequencing works (cont.)

Then, sequencing is performed by synthesizing a new strand using
fluorescently-labeled bases and taking a picture each time a new nucleotide is incorporated:

. . .

![](../img/illumina_seq2.png){fig-align="center"}

## How Illumina sequencing works (cont.)

![](../img/illumina_seq3_2.png){fig-align="center"}

## Video of Illumina technology

{{< video https://www.youtube.com/watch?v=fCd6B5HRaZ8 width="95%" height="81%" >}}

## How errors come about in Illumina

::::: columns
::: {.column width="50%"}
- The different templates within a cluster get out of sync because occasionally:
  - They miss a base incorporation
  - They incorporate two bases at once

<hr style="height:1pt; visibility:hidden;" />

- Base incorporation may also terminate before the end of the template is reached
:::

::: {.column width="50%"}
![](../img/illumina_errors.png){fig-align="center" width="72%"}
:::
:::::

. . .

::: callout-tip
## This error profile is why, for Illumina:

- There are hard limits on read lengths
- Base quality scores typically decrease towards the end of the reads
:::

## How Illumina sequencing works: Zooming out

![](../img/flowcell1.png){fig-align="center"}

## How Illumina sequencing works: Zooming out

![](../img/flowcell2.png){fig-align="center"}

# Long-read HTS {background-color="silver"}

## Long-read HTS

The technologies underlying the two main long-read HTS technologies are very different,
but have some commonalities beyond long reads — they:

. . .

<hr style="height:1pt; visibility:hidden;" />

- Perform **"single-molecule" sequencing** (no PCR amplification of library fragments)
- Require higher quality & quantity of DNA (because of the lack of PCR)
- Can detect some **base modifications**, like methyl groups

. . .

<hr style="height:1pt; visibility:hidden;" />

::: callout-important
## Error rates are changing

I mentioned earlier that long-read HTS has higher error rate than short-read (Illumina) HTS.

However, error rates in one type of PacBio sequencing where individual fragments
are sequenced multiple times ("HiFi") *are now lower than in Illumina*.
:::

## Nanopore sequencing

**A single strand of DNA passes through a nanopore** —\
the *electrical current* is measured, which depends on the combination of bases passes in the pore:

[![](../img/Nanopore-sequencing_crop.jpg){fig-align="center" width="73%"}](https://www.genome.gov/genetics-glossary/Nanopore-DNA-Sequencing)

::: legend2
<https://www.genome.gov/genetics-glossary/Nanopore-DNA-Sequencing>
:::

## Video of Oxford Nanopore technology

{{< video https://www.youtube.com/watch?v=RcP85JHLmnI width="93%" height="79%" >}}

## ONT (Nanopore) sequencers

![](../img/ont_sequencers.png){fig-align="center" width="70%"}

. . .

::: callout-caution
## Under development!

ONT constantly releases new flow cells with updated technology,
which have led to large decreases in error rates over the past decade ---
and even over the past two or so years.
:::

# (Reference) Genomes

## Reference genomes

Many HTS applications either **require a "reference genome"** _or_
**involve its production.**
What exactly does reference genome refer to? It usually includes:

. . .

- **An assembly**\
  A representation of most or all of the genome DNA sequence: the genome assembly

- **An annotation**\
  Provides e.g. *locations of genes and other genomic "features"*
  in the corresponding genome assembly, and functional information for these features

. . .

<hr style="height:1pt; visibility:hidden;" />

:::callout-tip
#### Taxonomic identity

Reference genomes are typically applicable at the **species level**.
For example, if you work with maize, you want a _Zea mays_ reference genome. But:

- If needed, it's often possible to work with genomes of closely related species
- Conversely, different subspecies/lines may have their own reference genomes
:::

## Genome size variation

[![https://en.wikipedia.org/wiki/Genome_size](../img/Genome_Sizes.png){fig-align="center" width="70%"}](https://en.wikipedia.org/wiki/Genome_size)

## Genome structure

:::::: columns
:::: {.column width="75%"}
![](../img/DNA_human_male_chromosomes.gif){fig-align="center" width="85%"}

::: legend2
https://en.wikipedi.org/wiki/Karyotype
:::
::::

::: {.column width="25%"}
<br> <br> <br> <br>

Key features:

- Number of distinct chromosomes
- Ploidy
:::
::::::

## Growth of genome databases

![](../img/genomes_genbank.png){fig-align="center" width="70%"}

::: legend2
<br>
@konkel2023
:::

## Genome assemblies

::: incremental
- With increasing usage & quality of long-read HTS,
  assemblies are getting better and better

- For chromosome-level assemblies,
  i.e. with one contiguous sequence for each chromosome,
  additional technologies than sequencing are often needed
  (e.g. Hi-C, optical mapping)

- Many assemblies are not "chromosome-level",
  but consist of --often 1000s of-- fragments (contigs and scaffolds).
  Even chromosome-level assemblies are not 100% complete.
:::

<hr style="height:1pt; visibility:hidden;" />

. . .

<details><summary>Question: Contigs vs. scaffolds?</summary>

**Contigs** are contiguous, known stretches of DNA created by the assembly process,
basically by overlapping reads.

Often, the order and orientation of two or more contigs is known,
but there is a gap of unknown size between them.
Such contigs are connected into **scaffolds** with a stretch of `N`s in between.
</details>

. . .

<hr style="height:1pt; visibility:hidden;" />

::: callout-note
## How is this data stored?
Both genome assemblies and annotations are typically saved in a single text file each —
we'll explore some of these files in tomorrow's lab.
:::

## Take-home messages

TBA

# Up next {background-color="silver"}

## The Garrigós et al. 2025 dataset

The labs this and next week are organized around the data set from @garrigós2025:

![](../img/garrigos_paper.png){fig-align="center" width="75%" fig-alt="A screenshot of the paper's front matter." .lightbox}

This paper uses **paired-end Illumina RNA-Seq data** to study gene expression in
*Culex pipiens* mosquitos infected with two different malaria-causing *Plasmodium*
protozoans.

## Tomorrow's lab

## Next week's content
